{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f98c1ecb-cb9d-4146-bf32-4df26333da2e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "35e9ec1f-88e2-42b4-82ea-f710b0f2513c",
   "metadata": {},
   "source": [
    "## Plotting Functions by MP and LO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "71f32d01-9164-47bd-a77e-846e2aecc2c4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘ArchR’ was built under R version 4.3.3”\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.3\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : devtools v2.4.5\n",
      "\n",
      "Warning message:\n",
      "“package ‘devtools’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘usethis’ was built under R version 4.3.3”\n",
      "\tLoading Package : grid v4.3.2\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "Warning message:\n",
      "“package ‘gridExtra’ was built under R version 4.3.3”\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "Warning message:\n",
      "“package ‘gtools’ was built under R version 4.3.3”\n",
      "\tLoading Package : gtable v0.3.6\n",
      "\n",
      "Warning message:\n",
      "“package ‘gtable’ was built under R version 4.3.3”\n",
      "\tLoading Package : ggplot2 v3.4.4\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "Warning message:\n",
      "“package ‘magrittr’ was built under R version 4.3.3”\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "Warning message:\n",
      "“package ‘plyr’ was built under R version 4.3.3”\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "Warning message:\n",
      "“package ‘stringr’ was built under R version 4.3.3”\n",
      "\tLoading Package : data.table v1.15.4\n",
      "\n",
      "Warning message:\n",
      "“package ‘data.table’ was built under R version 4.3.3”\n",
      "\tLoading Package : matrixStats v1.4.1\n",
      "\n",
      "Warning message:\n",
      "“package ‘matrixStats’ was built under R version 4.3.3”\n",
      "\tLoading Package : sparseMatrixStats v1.14.0\n",
      "\n",
      "Warning message:\n",
      "“package ‘sparseMatrixStats’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘MatrixGenerics’ was built under R version 4.3.3”\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "Warning message:\n",
      "“package ‘S4Vectors’ was built under R version 4.3.3”\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomicRanges’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.3.3”\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "Warning message:\n",
      "“package ‘Matrix’ was built under R version 4.3.3”\n",
      "\tLoading Package : Rcpp v1.0.13.1\n",
      "\n",
      "Warning message:\n",
      "“package ‘Rcpp’ was built under R version 4.3.3”\n",
      "\tLoading Package : RcppArmadillo v14.2.0.1\n",
      "\n",
      "Warning message:\n",
      "“package ‘RcppArmadillo’ was built under R version 4.3.3”\n",
      "\tLoading Package : SummarizedExperiment v1.32.0\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.3.3”\n",
      "\tLoading Package : rhdf5 v2.46.1\n",
      "\n",
      "Warning message:\n",
      "“package ‘rhdf5’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘dplyr’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘tidyverse’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘tibble’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘tidyr’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘readr’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘purrr’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘forcats’ was built under R version 4.3.3”\n",
      "Warning message:\n",
      "“package ‘lubridate’ was built under R version 4.3.3”\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mBiocGenerics\u001b[39m::\u001b[32mPosition()\u001b[39m  masks \u001b[34mggplot2\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32marrange()\u001b[39m          masks \u001b[34mplyr\u001b[39m::arrange()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m          masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m          masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine(), \u001b[34mgridExtra\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m          masks \u001b[34mplyr\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m            masks \u001b[34mmatrixStats\u001b[39m::count(), \u001b[34mplyr\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m             masks \u001b[34mIRanges\u001b[39m::desc(), \u001b[34mplyr\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m           masks \u001b[34mMatrix\u001b[39m::expand(), \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mextract()\u001b[39m          masks \u001b[34mmagrittr\u001b[39m::extract()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m         masks \u001b[34mplyr\u001b[39m::failwith()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m           masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m            masks \u001b[34mS4Vectors\u001b[39m::first(), \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mhour()\u001b[39m         masks \u001b[34mdata.table\u001b[39m::hour()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mid()\u001b[39m               masks \u001b[34mplyr\u001b[39m::id()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32misoweek()\u001b[39m      masks \u001b[34mdata.table\u001b[39m::isoweek()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m              masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m             masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mmday()\u001b[39m         masks \u001b[34mdata.table\u001b[39m::mday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mminute()\u001b[39m       masks \u001b[34mdata.table\u001b[39m::minute()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mmonth()\u001b[39m        masks \u001b[34mdata.table\u001b[39m::month()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mmutate()\u001b[39m           masks \u001b[34mplyr\u001b[39m::mutate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mpack()\u001b[39m             masks \u001b[34mMatrix\u001b[39m::pack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mquarter()\u001b[39m      masks \u001b[34mdata.table\u001b[39m::quarter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m           masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m           masks \u001b[34mS4Vectors\u001b[39m::rename(), \u001b[34mplyr\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::second(), \u001b[34mdata.table\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m     masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mset_names()\u001b[39m        masks \u001b[34mmagrittr\u001b[39m::set_names()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m            masks \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mGenomicRanges\u001b[39m::\u001b[32msubtract()\u001b[39m masks \u001b[34mmagrittr\u001b[39m::subtract()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarise()\u001b[39m        masks \u001b[34mplyr\u001b[39m::summarise()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m        masks \u001b[34mplyr\u001b[39m::summarize()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m        masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32munpack()\u001b[39m           masks \u001b[34mMatrix\u001b[39m::unpack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mwday()\u001b[39m         masks \u001b[34mdata.table\u001b[39m::wday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mweek()\u001b[39m         masks \u001b[34mdata.table\u001b[39m::week()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32myday()\u001b[39m         masks \u001b[34mdata.table\u001b[39m::yday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32myear()\u001b[39m         masks \u001b[34mdata.table\u001b[39m::year()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "Attaching package: ‘plyranges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, n, n_distinct\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    between\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, mutate, summarise, summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyranges’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘doParallel’ was built under R version 4.3.3”\n",
      "Loading required package: foreach\n",
      "\n",
      "Warning message:\n",
      "“package ‘foreach’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘foreach’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    accumulate, when\n",
      "\n",
      "\n",
      "Loading required package: iterators\n",
      "\n",
      "Warning message:\n",
      "“package ‘iterators’ was built under R version 4.3.3”\n",
      "Loading required package: parallel\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### These functions are meant for exporting ATAC data from an ArchR Project, whether into plots or various Bed/BedGraph/BigWig files\n",
    "\n",
    "library(ArchR)\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(plyranges)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.refGene)\n",
    "library(doParallel)\n",
    "\n",
    "#########################  Section 1: Exporting ATAC data into Coverage/Insertions file at single basepair resolution ----\n",
    "\n",
    "\n",
    "## This will allow for cell-count normalized insertion and coverage file creation.\n",
    "## Export options\n",
    "\n",
    "### Here's the pipeline:\n",
    "\n",
    "### Step 1: Extract fragments by populations from an ArchR Project: getPopFrags()\n",
    "### Step 2: Export to the file type that you desire\n",
    "################  FragsToCoverageFiles: Export fragments to BigWig/BedGraph files that contain coverage counts per 1000 cells.\n",
    "################  FragsToInsertionFiles: Export fragments to BigWig/BedGraph files that contain insertion counts per 1000 cells.\n",
    "\n",
    "\n",
    "#' Extract fragments by populations from an ArchR Project\n",
    "#'\n",
    "#' \\code{getPopFrags} returns a list of fragments per cell subset as a GRanges.\n",
    "#'\n",
    "#' @param ArchRProj The ArchR Project.\n",
    "#' @param metaColumn The name of metadata column that contains the populations\n",
    "#'   of cells you want to merge and export.\n",
    "#' @param cellSubsets Default is 'ALL'. If you want to export only some groups,\n",
    "#'   then give it a list of group names. This needs to be unique - no duplicated\n",
    "#'   names. This list of group names must be identical to names that appear in\n",
    "#'   the metadata column of the ArchR Project (e.g. metaColumn).\n",
    "#' @param region Optional parameter. Set this if you only want to extract\n",
    "#'   fragments from particular regions of the genome. Format should be as a\n",
    "#'   string (e.g. 'chr1:1000-2000'), or a GRanges object.\n",
    "#' @param numCores Number of cores to use.\n",
    "#' @param sampleSpecific Set to TRUE to further subset cells by sample\n",
    "#' @param NormMethod Normalization method. Can be either \"nFrags\",\"nCells\", or\n",
    "#'   \"Median\".\n",
    "#' @param blackList Blacklisted region to filter out. Default is to not filter\n",
    "#'   out anything (i.e. NULL). Input should be provided as a GRanges object. Any\n",
    "#'   fragments with more than a certain overlap will be thrown out.\n",
    "#' @param overlapList The minimum overlap necessary for a fragment marked as\n",
    "#'   overlapping with the blacklist region and thus thrown out.\n",
    "#'\n",
    "#' @return A list of GRanges containing fragments. Each GRanges corresponds to a\n",
    "#'   population defined by cellSubsets (and sample, if\n",
    "#'   \\code{sampleSpecific=TRUE})\n",
    "#'\n",
    "getPopFrags <- function(ArchRProj, metaColumn, cellSubsets = \"ALL\", region = NULL, numCores = 1, sampleSpecific = TRUE,\n",
    "                        NormMethod = \"nfrags\", blackList = NULL, overlapList = 50) {\n",
    "  # Extract metadata\n",
    "  metadf <- getCellColData(ArchRProj)\n",
    "\n",
    "  if (!any(colnames(metadf) %in% metaColumn)) {\n",
    "    stop(paste(\n",
    "      str_interp(\"Provided metaColumn ${metaColumn} does not exist in the cellColData of your ArchRProj.\"),\n",
    "      str_interp(\"Available columns are: ${colnames(metadf)}\")\n",
    "    ))\n",
    "  }\n",
    "\n",
    "  # Get cell counts for all cell populations\n",
    "  tmp <- metadf[, metaColumn]\n",
    "  cellCounts <- table(tmp[!is.na(tmp)])\n",
    "\n",
    "  if (all(cellSubsets == \"ALL\")) {\n",
    "    cellPopulations <- names(cellCounts)\n",
    "  } else {\n",
    "\n",
    "    # Take our cell populations from given cellSubsets\n",
    "    cellPopulations <- unique(cellSubsets[!is.na(cellSubsets)][order(cellSubsets)])\n",
    "\n",
    "    # Filter cellCounts to selected cell subsets (populations)\n",
    "    cellCounts <- cellCounts[cellPopulations]\n",
    "    if (any(is.na(cellCounts))) {\n",
    "      stop(paste(\n",
    "        \"Some cell populations have NA cell counts.\",\n",
    "        \"Please verify that all given cellSubsets exist for the given metaColumn.\",\n",
    "        str_interp(\"cellSubsets with NA cell counts: ${cellPopulations[is.na(names(cellCounts))]}\")\n",
    "      ))\n",
    "    }\n",
    "  }\n",
    "\n",
    "  if (length(cellCounts) == 0) {\n",
    "    stop(\"No cells were found for the given cellSubsets and/or metaColumn.\")\n",
    "  }\n",
    "\n",
    "\n",
    "  #### Up to here, the above is only specific to sampleSpecific=False\n",
    "  allArrows <- getArrowFiles(ArchRProj)\n",
    "  # Get sample names from our populations of interest\n",
    "  samplesToExtract <- paste(\n",
    "    unique(metadf$Sample[which(metadf[, metaColumn] %in% cellPopulations)]),\n",
    "    collapse = \"|\"\n",
    "  )\n",
    "  # Get arrows containing the samples we want\n",
    "  arrows <- allArrows[grepl(samplesToExtract, allArrows)]\n",
    "\n",
    "  if (length(arrows) == 0) {\n",
    "    stop(paste(\n",
    "      \"Found no arrows containing samples from your selected cell subset(s).\",\n",
    "      \"Check cellSubsets or the input ArchR Project.\"\n",
    "    ))\n",
    "  }\n",
    "\n",
    "  if (is.null(region)) {\n",
    "\n",
    "    # Extract fragments from all available regions\n",
    "    fragsList <- mclapply(seq_along(arrows), function(x) {\n",
    "      getFragmentsFromArrow(\n",
    "        ArrowFile = arrows[x],\n",
    "        cellNames = NULL,\n",
    "        verbose = FALSE\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  } else {\n",
    "\n",
    "    # Extract fragments from a given region only\n",
    "    # Validate region and interpret as string or GRanges\n",
    "    if (validRegionString(region) & tolower(NormMethod) == \"raw\") {\n",
    "      regionGRanges <- StringsToGRanges(region)\n",
    "    } else if (class(region)[1] == \"GRanges\" & tolower(NormMethod) == \"raw\") {\n",
    "      regionGRanges <- region\n",
    "    } else if (tolower(NormMethod) != \"raw\") {\n",
    "      stop(paste(\n",
    "        \"Wrong NormMethod set. Please set NormMethod = 'Raw'\",\n",
    "        \"if you wish to extract fragments from a specific region.\",\n",
    "        \"\\n Alternatively, you can extract the entire genome and then subset.\"\n",
    "      ))\n",
    "    } else {\n",
    "      stop(paste(\n",
    "        \"Invalid region input.\",\n",
    "        \"Region must either be a string matching the format\",\n",
    "        \"'seqname:start-end', or a GRanges object.\"\n",
    "      ))\n",
    "    }\n",
    "\n",
    "    chrom <- regionGRanges %>%\n",
    "      as.data.frame() %>%\n",
    "      dplyr::select(seqnames)\n",
    "    fragsList <- mclapply(seq_along(arrows), function(x) {\n",
    "      fragsGRanges <- getFragmentsFromArrow(\n",
    "        ArrowFile = arrows[x],\n",
    "        cellNames = NULL,\n",
    "        chr = as.character(chrom[, 1]),\n",
    "        verbose = FALSE\n",
    "      ) %>% plyranges::join_overlap_intersect(regionGRanges)\n",
    "\n",
    "      # Filter according to provided blacklist\n",
    "      if (is.null(blackList)) {\n",
    "        fragsGRanges\n",
    "      } else if (class(blackList)[1] == \"GRanges\") {\n",
    "        fragsGRanges %>% plyranges::filter_by_non_overlaps(blackList, minoverlap = overlapList)\n",
    "      } else {\n",
    "        stop(\"Error: Wrong format for blackList region. Please provide GRanges\")\n",
    "      }\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "\n",
    "  # From scMACS - sorts cell barcodes by population\n",
    "  barcodesByCellPop <- lapply(cellPopulations, function(x) {\n",
    "    row.names(metadf)[which(metadf[, metaColumn] == x)]\n",
    "  })\n",
    "\n",
    "  # Add normalization factor.\n",
    "  if (tolower(NormMethod) == \"raw\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), 1, sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"ncells\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), cellCounts / 1000, sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"nfrags\") {\n",
    "    # Calculate the total nFrags for our cell populations\n",
    "    nFragsNorm <- as.data.frame(metadf[, c(metaColumn, \"nFrags\")]) %>%\n",
    "      dplyr::filter(!is.na(get(metaColumn))) %>%\n",
    "      dplyr::filter(get(metaColumn) %in% cellPopulations) %>%\n",
    "      dplyr::group_by(get(metaColumn)) %>%\n",
    "      dplyr::summarize(nFrags = sum(nFrags))\n",
    "\n",
    "    # Verify expected cellPopulations have corresponding nFrags\n",
    "    if (!all(nFragsNorm[, 1] == cellPopulations)) {\n",
    "      print(nFragsNorm[, 1])\n",
    "      stop(\"Names of nFrags don't match the cell populations.\")\n",
    "    }\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), nFragsNorm$nFrags / 10^6, sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"median\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), \"Median\", sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"medmax\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), \"MedianMax\", sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"TotalSampleByNCells\") {\n",
    "    # This calculates total fragments per sample.\n",
    "    totalsampleFrags <- unlist(fragsList, length)\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), totalsampleFrags / 10^6 * cellCounts / 1000, sep = \"__\")\n",
    "  } else {\n",
    "    stop(\"Error: Incorrect NormMethod given.\")\n",
    "  }\n",
    "\n",
    "  # From scMACS - Function to sort fragments by populations based on cell barcode lists\n",
    "  subset_Frag <- function(cellNames, fragsGRanges) {\n",
    "    fragsTable <- as.data.table(fragsGRanges)\n",
    "    idx <- which(fragsTable$RG %in% cellNames)\n",
    "    fragsGRanges[idx]\n",
    "  }\n",
    "\n",
    "  ## Identify which subset of arrows the population can be found it.\n",
    "  ## Speeds up sample-specific population extraction\n",
    "  fragsListIndex <- lapply(cellPopulations, function(x) {\n",
    "    names(arrows) %in% unique(metadf$Sample[which(metadf[, metaColumn] == x)])\n",
    "  })\n",
    "\n",
    "  # Sort fragments into a list by cell population\n",
    "\n",
    "  popFrags <- lapply(seq_along(barcodesByCellPop), function(x) {\n",
    "    print(paste(\"Sorting \", names(barcodesByCellPop)[x], sep = \"\"))\n",
    "    if (sum(fragsListIndex[[x]]) > 1) {\n",
    "      tmp <- mclapply(which(fragsListIndex[[x]]), function(y) {\n",
    "        subset_Frag(barcodesByCellPop[[x]], fragsList[[y]])\n",
    "      }, mc.cores = 20)\n",
    "    } else {\n",
    "      tmp <- list(subset_Frag(\n",
    "        barcodesByCellPop[[x]],\n",
    "        fragsList[[which(fragsListIndex[[x]])]]\n",
    "      ))\n",
    "    }\n",
    "\n",
    "    # For this population, get sample-specific normalization factors\n",
    "    # and rename the fragment GRanges\n",
    "    if (sampleSpecific) {\n",
    "      subSet_tmp <- gsub(\"__.*\", \"\", names(barcodesByCellPop)[x])\n",
    "\n",
    "      if (tolower(NormMethod) == \"ncells\") {\n",
    "        tmp_cellCount <- unlist(lapply(tmp, function(y) {\n",
    "          length(unique(y$RG))\n",
    "        }))\n",
    "\n",
    "        names(tmp) <- paste(\n",
    "          subSet_tmp,\n",
    "          \"#\",\n",
    "          names(arrows)[unlist(fragsListIndex[[x]])],\n",
    "          \"__\",\n",
    "          tmp_cellCount / 1000,\n",
    "          sep = \"\"\n",
    "        )\n",
    "      } else if (tolower(NormMethod) == \"nfrags\") {\n",
    "        names(tmp) <- paste(\n",
    "          subSet_tmp,\n",
    "          \"#\",\n",
    "          names(arrows)[unlist(fragsListIndex[[x]])],\n",
    "          \"__\",\n",
    "          unlist(lapply(tmp, length)) / 10^6,\n",
    "          sep = \"\"\n",
    "        )\n",
    "      } else {\n",
    "        subSet_tmp <- gsub(\"__.*\", \"\", names(barcodesByCellPop)[x])\n",
    "        norm_tmp <- gsub(\".*__\", \"\", names(barcodesByCellPop)[x])\n",
    "\n",
    "        names(tmp) <- paste(\n",
    "          subSet_tmp,\n",
    "          \"#\",\n",
    "          names(arrows)[unlist(fragsListIndex[[x]])], \"__\", norm_tmp,\n",
    "          sep = \"\"\n",
    "        )\n",
    "      }\n",
    "\n",
    "      tmp\n",
    "    } else {\n",
    "      stack(as(tmp, \"GRangesList\"))\n",
    "    }\n",
    "  })\n",
    "\n",
    "  if (sampleSpecific) {\n",
    "    popFrags <- unlist(popFrags, recursive = FALSE)\n",
    "  } else {\n",
    "    names(popFrags) <- names(barcodesByCellPop)\n",
    "  }\n",
    "  return(popFrags)\n",
    "}\n",
    "\n",
    "\n",
    "###### This function take the output from getPopFrags and figures out how to normalize before exporting to the desired file formats.\n",
    "## popFrags = output from getPopFrags, as a list for the sake of parallelization. The name for each population is populatioName_Norm\n",
    "## metadata = ArchR metadata\n",
    "## metaColumn = column in metadata with popFrags\n",
    "## outDir = the directory to output the resultant fragment files. Optional - default is the current working directory.\n",
    "FragsToCoverageFiles <- function(popFrags, files = \"BigWig\", genome = \"hg38\", sampleAvg = TRUE, fname = NULL, outDir = NULL, numCores = 1) {\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  allNorm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(allNorm) <- groups\n",
    "\n",
    "  # Make bed file names for each group, either with names from popFrags or from the fname list.\n",
    "  if (is.null(fname) & !sampleAvg) {\n",
    "    bedFiles <- paste(gsub(\"-\", \"_\", gsub(\"/\", \"_\", gsub(\" \", \"_\", groups, fixed = TRUE))), \".bed\", sep = \"\")\n",
    "  } else if (length(fname) == length(popFrags) & !sampleAvg) {\n",
    "\n",
    "    # Check if fname is the same length as the popFrags and we're not doing sample Avg\n",
    "    bedFiles <- paste(fname, \".bed\", sep = \"\")\n",
    "  } else if (is.null(fname) & sampleAvg & all(grepl(\"#\", names(popFrags)))) {\n",
    "\n",
    "    # If fname is null and sampleAvg is true, then the popFrags needs to be formated for sample-specific coverage.\n",
    "\n",
    "    uniqueGroups <- unique(gsub(\"#.*\", \"#\", names(popFrags)))\n",
    "    bedFiles <- paste(gsub(\"-\", \"_\", gsub(\"/\", \"_\", gsub(\" \", \"_\", groups, fixed = TRUE))), \".bed\", sep = \"\")\n",
    "  } else if (length(fname) == length(unique(gsub(\"#.*\", \"#\", names(popFrags)))) & sampleAvg & all(grepl(\"#\", names(popFrags)))) {\n",
    "\n",
    "    # If we want to rename the sample average, then fname needs to be the same length as the number of unique groups.\n",
    "\n",
    "    uniqueGroups <- unique(gsub(\"#.*\", \"#\", names(popFrags)))\n",
    "    bedFiles <- paste(fname, \".bed\", sep = \"\")\n",
    "  } else if (sampleAvg == TRUE & !all(grepl(\"#\", names(popFrags)))) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists.\n",
    "\t\tPlease re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  } else {\n",
    "    stop(\"Error: fname parameter is not the same length as the number of populations you want to export.\")\n",
    "  }\n",
    "  # Prepend output directory (if given) to file names\n",
    "  if (!is.null(outDir)) {\n",
    "    dir.create(outDir, recursive = TRUE)\n",
    "    bedFiles <- file.path(outDir, bedFiles)\n",
    "  }\n",
    "\n",
    "  # Test export options for bed and bam files\n",
    "  if (all(tolower(files) == \"all\" & !sampleAvg)) {\n",
    "\n",
    "    # Export to BAM\n",
    "    writeBAMHelper(popFrags, bedFiles, genome, numCores)\n",
    "    # Export data from GRanges to bed\n",
    "    mclapply(seq_along(popFrags), function(x) {\n",
    "      # Export to bed\n",
    "      plyranges::write_bed(popFrags[[x]], bedFiles[x])\n",
    "    }, mc.cores = numCores)\n",
    "  } else if (any(tolower(files) == \"bed\" & !sampleAvg)) {\n",
    "\n",
    "    # Export data from GRanges to bed files, and end the function.\n",
    "    mclapply(seq_along(popFrags), function(x) {\n",
    "      plyranges::write_bed(popFrags[[x]], bedFiles[x])\n",
    "    }, mc.cores = numCores)\n",
    "    return()\n",
    "  } else if (any(tolower(files) == \"bam\" & !sampleAvg)) {\n",
    "    # Export to BAM, and end the function\n",
    "    writeBAMHelper(popFrags, bedFiles, genome, numCores)\n",
    "    return()\n",
    "  }\n",
    "\n",
    "  ### Let's generate coverage files within R,\n",
    "\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popFrags)))) {\n",
    "    mclapply(seq_along(uniqueGroups), function(x) {\n",
    "      print(paste(\"Summarizing\", uniqueGroups[x], sep = \"\"))\n",
    "\n",
    "      # Iterate along all samples for this population, calculating the coverage, and then turn the GRanges into a data.table\n",
    "      # After which, we bind those data.tables together, and turn it back into a GRanges object.\n",
    "      # This might seem ineffecient, but computationally, it's faster and less prone to error as plyranges is actively being developed.\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "\n",
    "        # Iterate along all the samples of this cell group.\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popFrags)))), function(y) {\n",
    "          tmp_Cov <- plyranges::compute_coverage(popFrags[[y]])\n",
    "\n",
    "          medCov <- tmp_Cov %>%\n",
    "            plyranges::filter(score != 0) %>%\n",
    "            plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "          maxScore <- max(tmp_Cov$score[medCov$partition])\n",
    "          medScore <- median(tmp_Cov$score[medCov$partition])\n",
    "          medMax <- max(tmp_Cov$score[medCov$partition] / medScore)\n",
    "\n",
    "          tmp_Cov %>%\n",
    "            plyranges::mutate(Norm = rep(NormVal, n = plyranges::n())) %>%\n",
    "            plyranges::mutate(score = case_when(\n",
    "              grepl(\"Median\", Norm) ~ score / medScore,\n",
    "              grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "              grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "              TRUE ~ score / as.numeric(Norm)\n",
    "            )) %>% # Redundant/safety measure. Should not be triggered.\n",
    "            as.data.table()\n",
    "\n",
    "\n",
    "          tmp_Cov\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      } else {\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popFrags)))), function(y) {\n",
    "\n",
    "          # Compute coverage for each sample, and normalize by the normalization factor.\n",
    "\n",
    "          NormVal <- unique(gsub(\".*__\", \"\", names(popFrags)[y]))\n",
    "          plyranges::compute_coverage(popFrags[[y]]) %>%\n",
    "            plyranges::mutate(name = names(popFrags)[y]) %>%\n",
    "            plyranges::mutate(score = score / as.numeric(NormVal)) %>%\n",
    "            as.data.table()\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      }\n",
    "\n",
    "      sampleNum <- sum(grepl(uniqueGroups[x], names(popFrags)))\n",
    "      popCov <- tmp1 %>% disjoin_ranges(score = sum(score) / sampleNum)\n",
    "      # return(popCov)\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\"#\", \"\", uniqueGroups[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  } else {\n",
    "    mclapply(seq_along(popFrags), function(x) {\n",
    "      print(paste(\"Summarizing \", gsub(\"__.*\", \"\", bedFiles[x]), sep = \"\"))\n",
    "\n",
    "      tmp1 <- plyranges::compute_coverage(popFrags[[x]])\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "        medCov <- tmp1 %>%\n",
    "          plyranges::filter(score != 0) %>%\n",
    "          plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "        maxScore <- max(tmp1$score[medCov$partition])\n",
    "        medScore <- median(tmp1$score[medCov$partition])\n",
    "        medMax <- max(tmp1$score[medCov$partition] / medScore)\n",
    "\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = case_when(\n",
    "            grepl(\"Median\", Norm) ~ score / medScore,\n",
    "            grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "            grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "            TRUE ~ score / as.numeric(Norm)\n",
    "          )) %>%\n",
    "          plyranges::mutate(score = ifelse(grepl(\"MedMax\", Norm), score / medMax))\n",
    "      } else {\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = score / as.numeric(Norm))\n",
    "      }\n",
    "\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\".bed\", \"\", bedFiles[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "### Similar function to above, except will export insertion counts.\n",
    "FragsToInsertionFiles <- function(popFrags, files = \"bw\", genome = \"hg38\", fname = NULL, outDir = NULL, numCores = 1) {\n",
    "\n",
    "  # Double check inputs: Are the files that we need actually present?\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  allNorm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(allNorm) <- groups\n",
    "\n",
    "  ## Let's extract Insertion sites\n",
    "  popIns <- mclapply(seq_along(popFrags), function(x) {\n",
    "    cutstart <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = start(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    cutend <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = end(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    plyranges::bind_ranges(cutstart, cutend)\n",
    "  }, mc.cores = numCores)\n",
    "  names(popIns) <- names(popFrags)\n",
    "\n",
    "\n",
    "  # Make bed file names for each group, either with names from popFrags or from the fname list.\n",
    "  if (is.null(fname)) {\n",
    "    bedFiles2 <- paste(gsub(\"-\", \"_\", gsub(\"/\", \"_\", gsub(\" \", \"_\", groups, fixed = TRUE))), \"_insertions.bed\", sep = \"\")\n",
    "  } else if (length(fname) == length(popFrags)) {\n",
    "    bedFiles2 <- paste(fname, \"_insertions.bed\", sep = \"\")\n",
    "  } else {\n",
    "    stop(\"Error: fname parameter is not the same length as popFrags\")\n",
    "  }\n",
    "  # Prepend output directory (if given) to file names\n",
    "  if (!is.null(outDir)) {\n",
    "    dir.create(outDir, recursive = TRUE)\n",
    "    bedFiles2 <- file.path(outDir, bedFiles2)\n",
    "  }\n",
    "\n",
    "  if (all(tolower(files) == \"all\")) {\n",
    "\n",
    "\n",
    "    # Export to BAM\n",
    "    writeBAMHelper(popIns, bedFiles2, genome, numCores)\n",
    "    # Export data from GRanges to bed files\n",
    "    # Then move onto export the rest\n",
    "    mclapply(seq_along(popIns), function(x) {\n",
    "      plyranges::write_bed(popIns[[x]], bedFiles2[x])\n",
    "    }, mc.cores = numCores)\n",
    "  } else if (any(tolower(files) == \"bed\")) {\n",
    "\n",
    "    # Export data from GRanges to bed files, and end the function.\n",
    "    mclapply(seq_along(popIns), function(x) {\n",
    "      plyranges::write_bed(popIns[[x]], bedFiles2[x])\n",
    "    }, mc.cores = numCores)\n",
    "    return()\n",
    "  } else if (any(tolower(files) == \"bam\")) {\n",
    "\n",
    "    # Export to BAM, and end the function\n",
    "    writeBAMHelper(popIns, bedFiles2, genome, numCores)\n",
    "    return()\n",
    "  }\n",
    "\n",
    "\n",
    "  # Export to bigwigs or bed graphs.\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popIns)))) {\n",
    "    uniqueGroups <- unique(gsub(\"#.*\", \"#\", names(popIns)))\n",
    "\n",
    "\n",
    "    mclapply(seq_along(uniqueGroups), function(x) {\n",
    "      print(paste(\"Summarizing\", uniqueGroups[x], sep = \"\"))\n",
    "\n",
    "      # Iterate along all samples for this population, calculating the coverage, and then turn the GRanges into a data.table\n",
    "      # After which, we bind those data.tables together, and turn it back into a GRanges object.\n",
    "      # This might seem ineffecient, but computationally, it's faster and less prone to error as plyranges is actively being developed.\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "\n",
    "        # Iterate along all the samples of this cell group.\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popIns)))), function(y) {\n",
    "          tmp_Cov <- plyranges::compute_coverage(popIns[[y]])\n",
    "\n",
    "          medCov <- tmp_Cov %>%\n",
    "            plyranges::filter(score != 0) %>%\n",
    "            plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "          maxScore <- max(tmp_Cov$score[medCov$partition])\n",
    "          medScore <- median(tmp_Cov$score[medCov$partition])\n",
    "          medMax <- max(tmp_Cov$score[medCov$partition] / medScore)\n",
    "\n",
    "          tmp_Cov %>%\n",
    "            plyranges::mutate(Norm = rep(NormVal, n = plyranges::n())) %>%\n",
    "            plyranges::mutate(score = case_when(\n",
    "              grepl(\"Median\", Norm) ~ score / medScore,\n",
    "              grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "              grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "              TRUE ~ score / as.numeric(Norm)\n",
    "            )) %>% # Redundant/safety measure. Should not be triggered.\n",
    "            as.data.table()\n",
    "\n",
    "\n",
    "          tmp_Cov\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      } else {\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popIns)))), function(y) {\n",
    "\n",
    "          # Compute coverage for each sample, and normalize by the normalization factor.\n",
    "\n",
    "          NormVal <- unique(gsub(\".*__\", \"\", names(popIns)[y]))\n",
    "          plyranges::compute_coverage(popIns[[y]]) %>%\n",
    "            plyranges::mutate(name = names(popIns)[y]) %>%\n",
    "            plyranges::mutate(score = score / as.numeric(NormVal)) %>%\n",
    "            as.data.table()\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      }\n",
    "\n",
    "      popCov <- tmp1 %>% disjoin_ranges(score = sum(score) / sum(grepl(uniqueGroups[x], names(popIns))))\n",
    "\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\"#\", \"\", uniqueGroups[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  } else if (sampleAvg == TRUE) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists. Please re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  } else {\n",
    "    mclapply(seq_along(popIns), function(x) {\n",
    "      print(paste(\"Exporting \", gsub(\".bed\", \"\", bedFiles2[x]), sep = \"\"))\n",
    "      tmp1 <- plyranges::compute_coverage(popIns[[x]])\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "        medCov <- tmp1 %>%\n",
    "          plyranges::filter(score != 0) %>%\n",
    "          plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "        maxScore <- max(tmp1$score[medCov$partition])\n",
    "        medScore <- median(tmp1$score[medCov$partition])\n",
    "        medMax <- max(tmp1$score[medCov$partition] / medScore)\n",
    "\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = case_when(\n",
    "            grepl(\"Median\", Norm) ~ score / medScore,\n",
    "            grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "            grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "            TRUE ~ score / as.numeric(Norm)\n",
    "          )) %>%\n",
    "          plyranges::mutate(score = ifelse(grepl(\"MedMax\", Norm), score / medMax))\n",
    "      } else {\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = score / as.numeric(Norm))\n",
    "      }\n",
    "\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\".bed\", \"_Insertions\", bedFiles2[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "#' Get basepair-resolution fragments coverage or insertion counts\n",
    "#'\n",
    "#' \\code{getbpCounts} Returns a GRanges or data frame containing single-basepair\n",
    "#' resolution scores on the given fragments.\n",
    "#'\n",
    "#' @param regionString The region from which to extract basepair counts. Format\n",
    "#'   should be as a string (e.g. 'chr1:1000-2000')\n",
    "#' @param popFrags Fragments from \\code{getPopFrags}\n",
    "#' @param numCores Number of cores to use.\n",
    "#' @param returnGRanges Set TRUE to return coverage as a GRanges, else default\n",
    "#'   behavior is to return a data frame\n",
    "#' @param sampleAvg Set to TRUE to take the sample average across a group. This\n",
    "#'   only works if \\code{popFrags} was generated with \\code{sampleSpecific=TRUE}\n",
    "#' @param approxLimit Max size of region in basepairs before binning is used.\n",
    "#' @param binSize Size of bins to use when binning\n",
    "#'\n",
    "#' @return A GRanges or Data Frame containing basepair resolution fragment or\n",
    "#'   insertion counts. If \\code{sampleAvg=TRUE} these counts will be averaged\n",
    "#'   over all samples within a population. If the region given by regionString\n",
    "#'   exceeds the size of approxLimit, counts will be aggregated into bins of\n",
    "#'   size binSize bp.\n",
    "getbpCounts <- function(regionString, popFrags, numCores,\n",
    "                        returnGRanges = FALSE, sampleAvg = TRUE,\n",
    "                        approxLimit = 100000, binSize = 250) {\n",
    "\n",
    "  # Parse through region\n",
    "  regionString <- gsub(\",\", \"\", regionString, fixed = TRUE)\n",
    "  chrom <- gsub(\":.*\", \"\", regionString)\n",
    "  startSite <- gsub(\".*:\", \"\", regionString) %>%\n",
    "    gsub(\"-.*\", \"\", .) %>%\n",
    "    as.numeric()\n",
    "  endSite <- gsub(\".*-\", \"\", regionString) %>% as.numeric()\n",
    "\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  Norm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(Norm) <- groups\n",
    "\n",
    "  # Determine if binning is needed to simplify things\n",
    "  if (endSite - startSite > approxLimit) {\n",
    "    print(str_interp(\"Size of region exceeds ${approxLimit}bp. Binning data over ${binSize}bp windows.\"))\n",
    "    binnedData <- regionGRanges %>%\n",
    "      tile_ranges(., binSize) %>%\n",
    "      mutate(idx = c(1:length(.)))\n",
    "  }\n",
    "\n",
    "  # Summarize the coverage over the region window at a single basepair resolution\n",
    "  tmpCounts <- mclapply(seq_along(groups), function(x) {\n",
    "    print(paste(\"Counting\", names(popFrags)[x], sep = \" \"))\n",
    "\n",
    "    Num <- as.numeric(Norm[groups[x]])\n",
    "    tmpGR <- plyranges::compute_coverage(popFrags[[x]]) %>%\n",
    "      plyranges::mutate(score = score / Num) %>%\n",
    "      plyranges::filter_by_overlaps(regionGRanges)\n",
    "\n",
    "    # If we are binning, average scores over each bin\n",
    "    if (endSite - startSite > approxLimit) {\n",
    "      tmpGR <- plyranges::join_overlap_intersect(tmpGR, binnedData) %>%\n",
    "        plyranges::group_by(idx) %>%\n",
    "        plyranges::reduce_ranges(score = mean(score)) %>%\n",
    "        ungroup()\n",
    "    }\n",
    "\n",
    "    tmpGR\n",
    "  }, mc.cores = numCores)\n",
    "  names(tmpCounts) <- groups\n",
    "\n",
    "\n",
    "  if (all(lapply(seq_along(tmpCounts), function(x) {\n",
    "    sum(tmpCounts[[x]]$score) == 0\n",
    "  }))) {\n",
    "    stop(\"Error: No fragments detected in this region\")\n",
    "  }\n",
    "\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popFrags)))) {\n",
    "    groupCounts <- stack(as(lapply(tmpCounts, setSeqInfo), \"GRangesList\")) %>%\n",
    "      plyranges::mutate(Group = gsub(\"#.*\", \"\", name)) %>%\n",
    "      S4Vectors::split(., .$Group)\n",
    "\n",
    "    tmpCounts <- lapply(groupCounts, function(x) {\n",
    "      sampleCount <- length(unique(x$name))\n",
    "      x %>%\n",
    "        plyranges::compute_coverage(weight = .$score / sampleCount) %>%\n",
    "        plyranges::filter_by_overlaps(regionGRanges)\n",
    "    })\n",
    "    names(tmpCounts) <- names(groupCounts)\n",
    "  } else if (sampleAvg == TRUE) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists.\n",
    "\t\tPlease re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  }\n",
    "\n",
    "  if (returnGRanges) {\n",
    "    return(tmpCounts)\n",
    "  }\n",
    "\n",
    "  groups <- names(tmpCounts)\n",
    "\n",
    "\n",
    "  ## Generate a data.frame for export.\n",
    "  allTmp <- mclapply(seq_along(tmpCounts), function(x) {\n",
    "    tmpdf <- as.data.frame(tmpCounts[[x]])\n",
    "    if (dim(tmpdf)[1] < 2 & endSite - startSite <= approxLimit) {\n",
    "      data.frame(\n",
    "        chr = rep(chrom, (endSite - startSite + 1)),\n",
    "        Locus = c(startSite:endSite),\n",
    "        Counts = rep(0, (endSite - startSite + 1)),\n",
    "        CellType = rep(groups[x], (endSite - startSite + 1))\n",
    "      )\n",
    "\n",
    "      print(paste(groups[x], \" has no coverage for this region\", sep = \"\"))\n",
    "    } else if (dim(tmpdf)[1] < 2 & endSite - startSite > approxLimit) {\n",
    "\n",
    "      # If you bin data, then you need to set the Locus as the middle of the GRanges.\n",
    "      # This is only an issue if you aren't doing bp-specific coverage.\n",
    "      # In this case, if a given group has no coverage, then we create an empty\n",
    "      # data.frame for that group, using the midpoint of the whole region.\n",
    "      as_tibble(binnedData) %>%\n",
    "        dplyr::mutate(Groups = groups[x]) %>%\n",
    "        mutate(\n",
    "          Locus = as.integer((start + end) / 2),\n",
    "          Counts = 0,\n",
    "          chr = seqnames\n",
    "        ) %>%\n",
    "        dplyr::select(chr, Locus, Counts, Groups) %>%\n",
    "        dplyr::filter(Locus >= startSite & Locus <= endSite)\n",
    "\n",
    "      print(paste(groups[x], \" has no coverage for this region\", sep = \"\"))\n",
    "    } else if (endSite - startSite > approxLimit) {\n",
    "\n",
    "      # If you bin data, then you need to set the Locus as the middle of the GRanges.\n",
    "      # This is only an issue if you aren't doing bp-specific coverage.\n",
    "\n",
    "      as_tibble(tmpdf) %>%\n",
    "        dplyr::mutate(Groups = groups[x]) %>%\n",
    "        mutate(\n",
    "          Locus = as.integer((start + end) / 2),\n",
    "          Counts = score,\n",
    "          chr = seqnames\n",
    "        ) %>%\n",
    "        dplyr::select(chr, Locus, Counts, Groups) %>%\n",
    "        dplyr::filter(Locus >= startSite & Locus <= endSite)\n",
    "    } else {\n",
    "      tmp <- tmpCounts[[x]] %>% plyranges::tile_ranges(width = 1)\n",
    "      tmp$score <- tmpCounts[[x]]$score[tmp$partition]\n",
    "      tmp$Groups <- rep(groups[x], length(tmp))\n",
    "\n",
    "      covdf <- as.data.frame(tmp) %>% dplyr::select(seqnames, start, score, Groups)\n",
    "      colnames(covdf) <- c(\"chr\", \"Locus\", \"Counts\", \"Groups\")\n",
    "\n",
    "      covdf %>% dplyr::filter(Locus >= startSite & Locus <= endSite)\n",
    "    }\n",
    "  }, mc.cores = numCores)\n",
    "\n",
    "  tmpCountsdf <- as.data.frame(data.table::rbindlist(allTmp))\n",
    "\n",
    "  return(tmpCountsdf)\n",
    "}\n",
    "\n",
    "# same function as above but for insertions\n",
    "# This calculates bp counts over a region, scaled per 1000 cells\n",
    "# regionString = region that you want to plot, e.g. chr1:10000-20000\n",
    "# frags = list of fragments by group, derived from getPopFrags\n",
    "# numCores for multithreading\n",
    "\n",
    "getbpInserts <- function(regionString, popFrags, numCores = 1, sampleAvg = TRUE) {\n",
    "\n",
    "  # Parse through region\n",
    "  chrom <- gsub(\":.*\", \"\", regionString)\n",
    "  startSite <- gsub(\".*:\", \"\", regionString) %>%\n",
    "    gsub(\"-.*\", \"\", .) %>%\n",
    "    as.numeric()\n",
    "  endSite <- gsub(\".*-\", \"\", regionString) %>% as.numeric()\n",
    "\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  Norm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(Norm) <- groups\n",
    "\n",
    "  tmpCounts <- lapply(c(1:length(popFrags)), function(x) {\n",
    "    print(paste(\"Counting insertions sites for \", names(popFrags)[x], sep = \" \"))\n",
    "    # Pull out insertion sites - start and end\n",
    "    cutstart <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = start(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    cutend <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = end(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    tmp_inserts <- plyranges::bind_ranges(cutstart, cutend)\n",
    "\n",
    "    # now iterate over region, looking at overlap with insertion sites\n",
    "    tmp <- unlist(mclapply(c(1:(endSite - startSite + 1)), function(i) {\n",
    "      singlebp <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite + i, end = startSite + i), strand = \"*\")\n",
    "      bpRegion <- tmp_inserts %>% plyranges::join_overlap_intersect(singlebp)\n",
    "      length(bpRegion)\n",
    "    }, mc.cores = numCores))\n",
    "\n",
    "    Num <- as.numeric(Norm[groups[x]])\n",
    "    tmp / Norm\n",
    "  })\n",
    "  names(tmpCounts) <- groups\n",
    "  tmpdf <- as.data.frame(tmpCounts)\n",
    "  tmpdf$Locus <- c(startSite:endSite)\n",
    "\n",
    "  # Create a table of counts data by Celltype, and basepair, with a column for the Chromosome name.\n",
    "  tmpCountsdf <- tmpdf %>%\n",
    "    tidyr::pivot_longer(cols = colnames(.)[colnames(.) != \"Locus\"], names_to = \"Groups\", values_to = \"Counts\") %>%\n",
    "    dplyr::mutate(chr = rep(chrom, dim(.)[1]))\n",
    "\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popFrags)))) {\n",
    "    AvgCountsdf <- tmpCountsdf %>%\n",
    "      dplyr::mutate(Groups = gsub(\"#.*\", \"\", Groups)) %>%\n",
    "      group_by(Groups, chr, Locus) %>%\n",
    "      dplyr::summarize(Counts = mean(Counts))\n",
    "\n",
    "    return(AvgCountsdf)\n",
    "  } else if (sampleAvg == TRUE) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists.\n",
    "\t\tPlease re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  } else {\n",
    "    return(tmpCountsdf)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "### Quick support functions for section 1 ---\n",
    "\n",
    "### Function to set seqinfo for GRanges objects.\n",
    "### This makes them easier to merge and export to BWs by setting the standard\n",
    "### chromosome lengths and names to a standard, as defined by TxDb.\n",
    "\n",
    "setSeqInfo <- function(GRangesObj, TxDb = TxDb.Hsapiens.UCSC.hg38.refGene) {\n",
    "  seqinfo(GRangesObj) <- seqinfo(TxDb)[seqnames(seqinfo(GRangesObj))]\n",
    "  return(GRangesObj)\n",
    "}\n",
    "\n",
    "\n",
    "## Export plyranges to various coveragefile formats\n",
    "\n",
    "exportCov <- function(CovObject, files, fileName, Txdb = TxDb.Hsapiens.UCSC.hg38.refGene) {\n",
    "\n",
    "  ### Let's exporting them to Bigwigs and/or bedgraphs\n",
    "\n",
    "  CovObject <- setSeqInfo(CovObject, Txdb)\n",
    "\n",
    "  print(paste(\"Exporting \", fileName, sep = \"\"))\n",
    "\n",
    "  if (all(tolower(files) == \"all\")) {\n",
    "\n",
    "    # Export BigWigs and BedGraphs\n",
    "\n",
    "    plyranges::write_bed_graph(CovObject, paste(fileName, \".bedGraph\", sep = \"\"))\n",
    "    plyranges::write_bigwig(CovObject, paste(fileName, \".bw\", sep = \"\"))\n",
    "  } else if (any(tolower(files) == \"bigwig\")) {\n",
    "\n",
    "    # Export Bigwigs only\n",
    "    plyranges::write_bigwig(CovObject, paste(fileName, \".bw\", sep = \"\"))\n",
    "  } else if (any(tolower(files) == \"bedgraph\")) {\n",
    "\n",
    "    # Export Bedgraphs only\n",
    "    plyranges::write_bed_graph(CovObject, paste(fileName, \".bedGraph\", sep = \"\"))\n",
    "  } else {\n",
    "    break(\"Error: Incorrect export option give. Check inputs\")\n",
    "  }\n",
    "}\n",
    "\n",
    "validRegionString <- function(regionString) {\n",
    "  if (!is.character(regionString)) {\n",
    "    return(FALSE)\n",
    "  }\n",
    "\n",
    "  pattern <- \"([0-9]{1,2}|chr[0-9]{1,2}):[0-9]*-[0-9]*\"\n",
    "  matchedPattern <- str_extract(regionString, pattern)\n",
    "\n",
    "  if (is.na(matchedPattern)) {\n",
    "    return(FALSE)\n",
    "  } else if (!matchedPattern == regionString) {\n",
    "    return(FALSE)\n",
    "  }\n",
    "\n",
    "  splits <- str_split(regionString, \"[:-]\")[[1]]\n",
    "  start <- splits[2]\n",
    "  end <- splits[3]\n",
    "  if (start > end) {\n",
    "    return(FALSE)\n",
    "  }\n",
    "\n",
    "  # All conditions satisfied\n",
    "  return(TRUE)\n",
    "}\n",
    "\n",
    "StringsToGRanges <- function(regionString) {\n",
    "  if (!validRegionString(regionString)) {\n",
    "    stop(\"Region must be a string matching format 'seqname:start-end', where start<end e.g. chr1:123000-123500\")\n",
    "  }\n",
    "\n",
    "  chrom <- gsub(\":.*\", \"\", regionString)\n",
    "  startSite <- gsub(\".*:\", \"\", regionString) %>%\n",
    "    gsub(\"-.*\", \"\", .) %>%\n",
    "    as.numeric()\n",
    "  endSite <- gsub(\".*-\", \"\", regionString) %>% as.numeric()\n",
    "\n",
    "\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "  return(regionGRanges)\n",
    "}\n",
    "\n",
    "# Helper function to process and write to BAM\n",
    "writeBAMHelper <- function(popFrags, bedFiles, genome, numCores) {\n",
    "\n",
    "  # Coerce 'names' of popFrags to <character> for BamFile conversion\"\n",
    "  if (\"name\" %in% colnames(popFrags)) {\n",
    "    popFrags <- mclapply(seq_along(popFrags), function(x) {\n",
    "      if (\"name\" %in% colnames(popFrags)) {\n",
    "        popFrags[[x]] <- transform(popFrags[[x]], name = as.character(name))\n",
    "      }\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "\n",
    "  ## Fetch chromInfo file from UCSC for given genome, \"hg38\" default.\n",
    "  ## This uses fetchExtendedChromInfoFromUCSC() behind the scenes and\n",
    "  ## requires internet access. See ?fetchExtendedChromInfoFromUCSC\n",
    "  ## for the list of UCSC genomes that are currently supported.\n",
    "  chromInfo <- Seqinfo(genome = genome)\n",
    "\n",
    "  # Change bedFiles extensions to .bam\n",
    "  bamFiles <- gsub(\".bed\", \".bam\", bedFiles)\n",
    "\n",
    "  # Export data from GRanges to bam files, and end the function.\n",
    "  mclapply(seq_along(popFrags), function(x) {\n",
    "    # Export to bed\n",
    "    plyranges::write_bed(popFrags[[x]], bedFiles[x])\n",
    "\n",
    "    # Convert to GAlignments\n",
    "    popGAlignments <- as(popFrags[[x]], \"GAlignments\")\n",
    "\n",
    "    # Add seqinfo\n",
    "    chrNames <- names(seqlengths(popGAlignments))\n",
    "    seqinfo(popGAlignments) <- chromInfo[chrNames]\n",
    "\n",
    "    # Remove metadata cols\n",
    "    mcols(popGAlignments) <- NULL\n",
    "\n",
    "    # Export to BAM\n",
    "    rtracklayer::export(popGAlignments, bamFiles[x])\n",
    "  }, mc.cores = numCores)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8b5b204b-781e-4a4a-a668-c0c98b091a27",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "Need specific help about ggbio? try mailing \n",
      " the maintainer or visit https://lawremi.github.io/ggbio/\n",
      "\n",
      "\n",
      "Attaching package: 'ggbio'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ggplot2':\n",
      "\n",
      "    geom_bar, geom_rect, geom_segment, ggsave, stat_bin, stat_identity,\n",
      "    xlim\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "######################### Section 2: Plotting Coverge and Insertion data for specific regions ----\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(plyranges)\n",
    "library(ggrepel)\n",
    "library(ggbio)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.refGene)\n",
    "library(org.Hs.eg.db)\n",
    "library(OrganismDbi)\n",
    "library(doParallel)\n",
    "\n",
    "\n",
    "\n",
    "## Work Flow:\n",
    "#### 1. Pull out fragments for a given region by group - getPopFrags, from ArchR_Export_ATAC_Data.R\n",
    "######### source('ArchR_Export_ATAC_Data.R\")\n",
    "#### 2. Summarize the bp counts for either coverage or insertions, normalized per thousand cells\n",
    "#### 3. Plot coverage or insertions, with motif positions overlaid.\n",
    "\n",
    "\n",
    "# This calculates bp counts over a region, scaled per 1000 cells\n",
    "# region = region that you want to plot\n",
    "# regionString = region that you want to plot, e.g. chr1:10000-20000\n",
    "# numCores for multithreading\n",
    "\n",
    "\n",
    "##### supporting functions for plotRegion\n",
    "# Modified Track Plots\n",
    "\n",
    "#' Get gene promoter range string\n",
    "#'\n",
    "#' For a given gene, return the gene position from an ArchR project    \n",
    "#' \n",
    "#' @param proj An ArchR Project\n",
    "#' @param gene The name of a gene\n",
    "#' @param upstream The number of basepairs to extend the range upstream of the gene promoter\n",
    "#' @param downstream The number of basepairs to extend the range downstream of the gene promoter\n",
    "#' @param output_string Logical value, default TRUE. If TRUE outputs a range string (ie 'chr1: 1700000-1710000') else\n",
    "#' outputs a granges object\n",
    "#' \n",
    "#' @return A range string or granges object, depending on 'output_string' parameter.\n",
    "get_promoter_range <- function(proj, gene, upstream = 50000, downstream = 50000, output_string = TRUE) {\n",
    "  gene_pos <- ArchR::getGenes(proj, gene)\n",
    "  pro_pos_gr <- GenomicRanges::promoters(gene_pos, upstream = upstream, downstream = downstream)\n",
    "  if (output_string) {\n",
    "    pro_pos_str <- sprintf(\n",
    "      \"%s: %s-%s\",\n",
    "      GenomicRanges::seqnames(pro_pos_gr),\n",
    "      GenomicRanges::start(GenomicRanges::ranges(pro_pos_gr)),\n",
    "      GenomicRanges::end(GenomicRanges::ranges(pro_pos_gr))\n",
    "    )\n",
    "    return(pro_pos_str)\n",
    "  } else {\n",
    "    return(pro_pos_gr)\n",
    "  }\n",
    "}\n",
    "\n",
    "#' CountDF to Region GRanges\n",
    "#'\n",
    "#' Gets the range covered by counts in a count df and converts to granges\n",
    "#' Used in `plotRegion()`\n",
    "#'\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or `getbpInserts()`. Expected columns \"chr\" and \"Locus\"\n",
    "#' @return A granges object for the region defined in countdf\n",
    "countdf_to_region <- function(countdf) {\n",
    "  assertthat::assert_that(\"chr\" %in% names(countdf))\n",
    "  assertthat::assert_that(\"Locus\" %in% names(countdf))\n",
    "  \n",
    "  chrom <- toString(unique(countdf$chr))\n",
    "  startSite <- min(countdf$Locus)\n",
    "  endSite <- max(countdf$Locus)\n",
    "  regionGRanges <- GenomicRanges::GRanges(\n",
    "    seqnames = chrom,\n",
    "    ranges = IRanges::IRanges(start = startSite, end = endSite), strand = \"*\"\n",
    "  )\n",
    "  \n",
    "  return(regionGRanges)\n",
    "}\n",
    "\n",
    "#' Get GeneBody GRanges for a given region\n",
    "#'\n",
    "#' Used in `plotRegion()`\n",
    "#'\n",
    "#' @param regionGRanges regionGRanges A region Granges object to retrieve gene bodies for. For example, the output of countdf_to_region.\n",
    "#' @param TxDb A TxDb database, default TxDb.Hsapiens.UCSC.hg38.refGene\n",
    "#' @param single.strand.genes.only Logical, default FALSE.\n",
    "get_grange_genebody <- function(regionGRanges, TxDb = TxDb.Hsapiens.UCSC.hg38.refGene, single.strand.genes.only = FALSE) {\n",
    "  geneBody <- GenomicFeatures::genes(TxDb, single.strand.genes.only = single.strand.genes.only) %>%\n",
    "    unlist() %>%\n",
    "    plyranges::join_overlap_intersect(regionGRanges)\n",
    "  return(geneBody)\n",
    "}\n",
    "\n",
    "\n",
    "#' Default ggplot theme for counts plot\n",
    ".counts_plot_default_theme <- list(\n",
    "  panel.grid = element_blank(),\n",
    "  plot.margin = unit(c(0, 0, 0, 0), \"cm\"),\n",
    "  legend.title = element_text(size = 12),\n",
    "  legend.text = element_text(size = 10),\n",
    "  axis.text.y = element_blank(),\n",
    "  axis.ticks.y = element_blank(),\n",
    "  panel.border = element_blank(),\n",
    "  strip.background = element_blank(),\n",
    "  strip.text.x = element_text(size = 10, angle = 0),\n",
    "  strip.text.y = element_text(size = 10, angle = 0)\n",
    ")\n",
    "\n",
    "\n",
    "#' Plot normalized counts for each sample\n",
    "#'\n",
    "#' Used in `plotRegion()` for the counts tracks\n",
    "#'\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or `getbpInserts()`\n",
    "#' @param plotType Options include 'overlaid','stacked', or 'RidgePlot'. default is 'stacked', which will plot a seperate track for each group. \n",
    "#' Setting plotType to 'overlaid' will overlay count plot histograms across samples, instead of faceting out separately.\n",
    "#' Setting plotType to 'RidgePlot' will generate a ridgeplot across all groups. \n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param counts_color Optional color palette. A named vector of color values where names are unique values in the `color_var` column\n",
    "#' @param range_label_size Numeric value, default 4. Text size for the y-axis range label\n",
    "#' @param legend.position Any acceptable `legend.position` argument to theme(). Default NULL will place legend for overlaid plots at (0.8,0.8),\n",
    "#' or to the \"right\" for faceted plots.\n",
    "#' @param facet_label_side Direction character value, default \"top\". Can also be \"right\", \"left\", or \"bottom\". Position of facet label.\n",
    "#' @param counts_group_colors Optional named color vector. Values as colors, names are levels of `counts_color_var`. If provided, will color the plots specifically\n",
    "#' using `scale_color_manual()`\n",
    "#' @param counts_color_var Character value, default \"Groups\". Column name from countdf to use to color counts plots.\n",
    "#' Only used if counts_group_colors provided\n",
    "#' @param theme_ls A list of named theme arguments passed to theme(), defaults to `.counts_plot_default_theme`. For example, `list(axis.ticks = element_blank())`\n",
    "#' @return A ggplot object of count histograms by sample.\n",
    "\n",
    "counts_plot_samples <- function(countdf,\n",
    "                                plotType = 'stacked',\n",
    "                                base_size = 12,\n",
    "                                counts_color = NULL,\n",
    "                                range_label_size = 4,\n",
    "                                legend.position = NULL,\n",
    "                                facet_label_side = \"top\",\n",
    "                                counts_group_colors = NULL,\n",
    "                                counts_color_var = \"Groups\",\n",
    "                                theme_ls = .counts_plot_default_theme) {\n",
    "  assertthat::assert_that(\"Counts\" %in% names(countdf))\n",
    "  assertthat::assert_that(\"Locus\" %in% names(countdf))\n",
    "  assertthat::assert_that(\"Groups\" %in% names(countdf))\n",
    "  assertthat::assert_that(counts_color_var %in% names(countdf))\n",
    "  \n",
    "  # Fill in theme any unspecified theme options with defaults\n",
    "  default_theme <- .counts_plot_default_theme\n",
    "  unspec_param <- setdiff(names(default_theme), names(theme_ls))\n",
    "  if (length(unspec_param) > 0) {\n",
    "    theme_ls <- c(theme_ls, default_theme[unspec_param])\n",
    "  }\n",
    "  \n",
    "  # Range df for plotting\n",
    "  df_range <- data.frame(\n",
    "    x = max(countdf$Locus),\n",
    "    y = max(countdf$Counts),\n",
    "    label = paste0(\"Range:\", 0, \"-\", round(max(countdf$Counts), digits = 2))\n",
    "  )\n",
    "  \n",
    "  # Intialize plot\n",
    "  p1 <- ggplot(data = countdf, aes(x = Locus, y = Counts)) +\n",
    "    theme_bw(base_size = base_size)\n",
    "  \n",
    "  \n",
    "  # Plots\n",
    "  if (tolower(plotType) == 'overlaid') {\n",
    "    # Conditional Theme\n",
    "    if (is.null(legend.position)) {\n",
    "      legend.position <- c(0.8, 0.8)\n",
    "    }\n",
    "    theme_ls <- c(\n",
    "      theme_ls,\n",
    "      list(\n",
    "        legend.position = legend.position\n",
    "      )\n",
    "    )\n",
    "    \n",
    "    # Base Plot\n",
    "    p1 <- p1 +\n",
    "      geom_line(aes(color = !!as.name(counts_color_var)), \n",
    "                alpha = 0.75, size = 1.5) +\n",
    "      ylab(NULL) +\n",
    "      labs(Groups = \"Groups\") +\n",
    "      coord_cartesian(clip = \"off\") +\n",
    "      geom_text(\n",
    "        data = df_range,\n",
    "        aes(x = x, y = y, label = label),\n",
    "        size = range_label_size,\n",
    "        hjust = 1,\n",
    "        vjust = 1\n",
    "      ) +\n",
    "      do.call(theme, theme_ls)\n",
    "    \n",
    "    # Conditional Plot Elements\n",
    "    if (!is.null(counts_group_colors)) {\n",
    "      # assertthat::assert_that(all(unique(countdf[[counts_color_var]]) %in% names(counts_group_colors)),\n",
    "      #                        msg = \"Must supply colors for all levels of color variable\")\n",
    "      p1 <- p1 + scale_color_manual(values = counts_group_colors, breaks = names(counts_group_colors))\n",
    "    }\n",
    "  } else if(tolower(plotType) == 'stacked'){\n",
    "    # Conditional Theme\n",
    "    if (is.null(legend.position)) {\n",
    "      legend.position <- \"right\"\n",
    "    }\n",
    "    theme_ls <- c(\n",
    "      theme_ls,\n",
    "      list(\n",
    "        legend.position = legend.position\n",
    "      )\n",
    "    )\n",
    "    \n",
    "    # Base Plot, conditional\n",
    "    if (is.null(counts_color)) {\n",
    "      p1 <- p1 +\n",
    "        geom_area(aes(fill = !!as.name(counts_color_var)), position = \"identity\")\n",
    "    } else {\n",
    "      p1 <- p1 +\n",
    "        geom_area(fill = counts_color, position = \"identity\")\n",
    "    }\n",
    "    \n",
    "    # Base Plot, common elements\n",
    "    p1 <- p1 +\n",
    "      ylab(NULL) +\n",
    "      facet_wrap(vars(Groups), ncol = 1, strip.position = facet_label_side) +\n",
    "      geom_text(\n",
    "        data = df_range,\n",
    "        aes(x = x, y = y, label = label),\n",
    "        size = range_label_size,\n",
    "        hjust = 1,\n",
    "        vjust = 1\n",
    "      ) +\n",
    "      do.call(theme, theme_ls)\n",
    "    \n",
    "    # Conditional Plot Elements\n",
    "    if (!is.null(counts_group_colors)) {\n",
    "      # commenting this section out so we can supply specific colors to highlight just a subset if we want--other samples will just be gray\n",
    "      # assertthat::assert_that(all(unique(countdf[[counts_color_var]]) %in% names(counts_group_colors)),\n",
    "      #                        msg = \"Must supply colors for all levels of color variable\")\n",
    "      p1 <- p1 +\n",
    "        scale_fill_manual(values = counts_group_colors, breaks = names(counts_group_colors))\n",
    "    }\n",
    "  } else if(tolower(plotType) == 'ridgeplot'){\n",
    "    \n",
    "    \n",
    "    ##Conditional elements needed to make RidgePlots work.\n",
    "    countdf_tmp<- as.data.frame(table(countdf$Groups)) \n",
    "    countdf$Groups2 <- match(countdf$Groups,countdf_tmp$Var1)\n",
    "    \n",
    "    \n",
    "  #Base plot, conditional\n",
    "   p1 <- p1 + \n",
    "      geom_ridgeline(data = as.data.frame(countdf),\n",
    "                     aes(x = Locus, y = Groups2, height = Counts, \n",
    "                         fill = Groups),\n",
    "                     alpha = 0.25)+\n",
    "      ylab(NULL) + \n",
    "      scale_y_continuous(breaks=c(1:length(tmp$Var1)), \n",
    "                        label= tmp$Var1) +\n",
    "      geom_text(\n",
    "        data = data.frame(\n",
    "          x = Inf, y = Inf,\n",
    "          label = paste(\"Range:\", 0, \"-\", round(max(countdf$Counts), digits = 2), sep = \"\")\n",
    "        ),\n",
    "        aes(x = x, y = y, label = label), \n",
    "        size = 2, hjust = 0.9, vjust = 1.4\n",
    "      ) + \n",
    "      theme(legend.position = \"none\")\n",
    "    \n",
    "    \n",
    "  } else {stop(\"Error: Plot type not recognized. Please check input for variable 'plotType'\")}\n",
    "  \n",
    "  return(p1)\n",
    "}\n",
    "\n",
    "#' Get scaled values for custom breaks\n",
    "#'\n",
    "#' Get scaled values for custom breaks in a given value vector for use\n",
    "#' defining breaks in scale_gradientn(), for example\n",
    "#' @param breaks\n",
    "#' @param x\n",
    "breaks_to_scaledbreaks <- function(breaks, x) {\n",
    "  rescaled_weights <- scales::rescale(x)\n",
    "  rescaled_breaks <- quantile(rescaled_weights, probs = ecdf(x)(breaks))\n",
    "  return(rescaled_breaks)\n",
    "}\n",
    "\n",
    "cleanup_breaks <- function(breaks, x) {\n",
    "  breaks <- sort(breaks)\n",
    "  n_lower <- sum(breaks < min(x))\n",
    "  while (n_lower > 1) {\n",
    "    breaks <- breaks[-1]\n",
    "    n_lower <- sum(breaks < min(x))\n",
    "  }\n",
    "  n_higher <- sum(breaks > max(x))\n",
    "  while (n_higher > 1) {\n",
    "    breaks <- breaks[-length(breaks)]\n",
    "    n_higher <- sum(breaks > max(x))\n",
    "  }\n",
    "  return(breaks)\n",
    "}\n",
    "\n",
    "\n",
    "#' Get GRanges of motif annotations within a region\n",
    "#'\n",
    "#' Helper to return the motif annotations within a specfic region outside of\n",
    "#' the plotting function. Can supply either a counts df or a region string.\n",
    "#'\n",
    "#' @param ArchRProj The archr project containing motif annotations\n",
    "#' @param motifSetName The name of the motif annotations in the ArchR project\n",
    "#' @param countdf A counts data frame object from getPop\n",
    "get_motifs_in_region <- function(\n",
    "  ArchRProj, \n",
    "  motifSetName, \n",
    "  countdf = NULL, \n",
    "  regionString = NULL, \n",
    "  numCores = 1, \n",
    "  metaColumn = NULL, \n",
    "  cellSubsets = \"ALL\"\n",
    ") {\n",
    "  \n",
    "  if (is.null(countdf) & !is.null(regionString)) {\n",
    "    popFrags <- getPopFrags(ArchRProj = ArchRProj, metaColumn = metaColumn, region = regionString, numCores = numCores, cellSubsets = cellSubsets)\n",
    "    countdf <- getbpCounts(regionString = regionString, popFrags = popFrags, numCores = numCores, returnGRanges = FALSE)\n",
    "  }\n",
    "  chrom <- toString(unique(countdf$chr))\n",
    "  startSite <- min(countdf$Locus)\n",
    "  endSite <- max(countdf$Locus)\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "  \n",
    "  specMotifs <- unlist(getPositions(ArchRProj, name = motifSetName)) %>%\n",
    "    plyranges::mutate(name = gsub(\"_.*\", \"\", names(.))) %>%\n",
    "    plyranges::join_overlap_intersect(regionGRanges) %>%\n",
    "    plyranges::mutate(index = seq(1, length(.), by = 1)) %>%\n",
    "    plyranges::mutate(labels = paste(name, index, sep = \"_\"))\n",
    "  \n",
    "  return(specMotifs)\n",
    "}\n",
    "\n",
    "#' Plot Motifs on Counts Plot\n",
    "#'\n",
    "#' Add motif annotation labels on an existing counts plot. If providing weights\n",
    "#' for motif text colors, ensure the input counts plot is formatted accordingly\n",
    "#' (ie counts_color = \"gray90\") for best visibility.\n",
    "#'\n",
    "#' @param p1 The output of `counts_plot_samples()`\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or\n",
    "#'   `getbpInserts()`\n",
    "#' @param ArchRProj An archR project containing motif annotations\n",
    "#' @param motifSetName The name of the motif set in ArchRProj to use for\n",
    "#'   annotation. Example: 'JasparMotifs'\n",
    "#' @param customFeatures Optional - A GRanges containing a user-specific set of\n",
    "#'   motifs or other features. If specified these will be used instead of the\n",
    "#'   ArchR motif set.\n",
    "#' @param motif_y_space_factor A factor for vertical spacing between motif\n",
    "#'   labels. Default 4. Increase to make labels farther apart, decrease to make\n",
    "#'   labels closer.\n",
    "#' @param motif_stagger_labels_y = FALSE Logical value, default FALSE. If TRUE,\n",
    "#'   will  stagger motif labels in adjacent columns in the vertical direction\n",
    "#' @param motif_weights Optional numeric vector, default NULL. If provided will\n",
    "#'   be used to color motif labels by the weighted values. Values must be\n",
    "#'   uniquely named with motif names, for example c(`KLF5`= 3.2, `STAT1 = 0.2`,\n",
    "#'   `EOMES` = -1.4`). Weights can be anything relevant, for example if the\n",
    "#'   peak/region is associated with a specific group/sample then global motif\n",
    "#'   enrichment results for that group: `-log10(FDR)*sign(change)`\n",
    "#' @param motif_weight_name Character value, default \"Motif Weight\". Used to\n",
    "#'   label the color legend.\n",
    "#' @param weight_colors Named numeric vector. Names should be color values and\n",
    "#'   breaks should be the corresponding values of motif_weights. Values outside\n",
    "#'   the highest and lowest value will appear as max or min defined color value.\n",
    "#' @param motif_lab_size Numeric value, default 1. Size of motif labels.\n",
    "#' @param motif_lab_alpha Numeric value, default 0.25. Alpha for motif labels.\n",
    "#' @param motif_line_size Numeric value, default 1. Size of motif lines.\n",
    "#' @param motif_line_alpha Numeric value, default 0.25. Alpha for motif lines.\n",
    "#' @return The input ggplot object with motif labels overlaid\n",
    "counts_plot_motif_overlay <- function(p1,\n",
    "                                      countdf,\n",
    "                                      ArchRProj = NULL,\n",
    "                                      motifSetName = NULL,\n",
    "                                      customFeatures = NULL,\n",
    "                                      motif_y_space_factor = 4,\n",
    "                                      motif_stagger_labels_y = FALSE,\n",
    "                                      motif_weights = NULL,\n",
    "                                      motif_weight_name = \"Motif Weight\",\n",
    "                                      motif_weight_colors = c(darkblue = -10, gray = 0, darkred = 10),\n",
    "                                      motif_lab_size = 1,\n",
    "                                      motif_lab_alpha = 0.25,\n",
    "                                      motif_line_size = 0.75,\n",
    "                                      motif_line_alpha = 0.25) {\n",
    "  \n",
    "  # Retrieve annotations in region and format\n",
    "  if (!is.null(customFeatures)){\n",
    "    \n",
    "    # Use custom features/motifs\n",
    "    specMotifs <- customFeatures\n",
    "    # Filter these to our region\n",
    "    chrom <- toString(unique(countdf$chr))\n",
    "    startSite <- min(countdf$Locus)\n",
    "    endSite <- max(countdf$Locus)\n",
    "    regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "    specMotifs <- specMotifs %>%\n",
    "      plyranges::mutate(name = gsub(\"_.*\", \"\", names(.))) %>%\n",
    "      plyranges::join_overlap_intersect(regionGRanges) %>%\n",
    "      plyranges::mutate(index = seq(1, length(.), by = 1)) %>%\n",
    "      plyranges::mutate(labels = paste(name, index, sep = \"_\"))\n",
    "    \n",
    "  } else if (!is.null(motifSetName) & !is.null(ArchRProj)){\n",
    "    \n",
    "    # Get our motifs from the ArchR Project\n",
    "    assertthat::assert_that(\n",
    "      motifSetName %in% names(ArchRProj@peakAnnotation),\n",
    "      msg = sprintf(\"%s not found in ArchRProj\", motifSetName)\n",
    "    )\n",
    "    specMotifs <- get_motifs_in_region(\n",
    "      countdf = countdf,\n",
    "      ArchRProj = ArchRProj,\n",
    "      motifSetName = motifSetName\n",
    "    )\n",
    "    \n",
    "  } else {\n",
    "    stop(\"To plot motifs either motifSetName and ArchRProj must be provided, or provide a GRanges to customFeatures\")\n",
    "  }\n",
    "  \n",
    "  reduceMotifs <- plyranges::reduce_ranges(specMotifs, count = plyranges::n(), names = paste(labels, collapse = \",\"))\n",
    "  splitMotifs <- strsplit(reduceMotifs$names, split = \",\")\n",
    "  \n",
    "  # Calculate y spacing for motif labels\n",
    "  minSize <- max(countdf$Counts) / (length(specMotifs) / motif_y_space_factor)\n",
    "  y1 <- rep(0, length(specMotifs))\n",
    "  names(y1) <- specMotifs$labels\n",
    "  \n",
    "  for (i in 1:length(splitMotifs)) {\n",
    "    if (i %% 2 == 0 & motif_stagger_labels_y) { # stagger y spacing in neighboring columns\n",
    "      start_val <- minSize / 2\n",
    "    } else {\n",
    "      start_val <- 0\n",
    "    }\n",
    "    \n",
    "    y1[splitMotifs[[i]]] <- seq(start_val, start_val + minSize * (length(splitMotifs[[i]]) - 1), by = minSize)\n",
    "  }\n",
    "  \n",
    "  # TF label coordinates and labels\n",
    "  tmp_motifdf <- data.frame(\n",
    "    x1 = start(specMotifs),\n",
    "    x2 = start(specMotifs) + width(specMotifs),\n",
    "    y = y1,\n",
    "    name = specMotifs$labels\n",
    "  ) %>%\n",
    "    pivot_longer(cols = c(\"x1\", \"x2\"), names_to = NULL, values_to = \"x\") %>%\n",
    "    group_by(name) %>%\n",
    "    mutate(labels = ifelse(max(x) == x, gsub(\"_.*\", \"\", name), NA))\n",
    "  \n",
    "  # Incorporate Weights\n",
    "  if (!is.null(motif_weights)) {\n",
    "    assertthat::assert_that(is.numeric(motif_weight_colors))\n",
    "    \n",
    "    if (length(intersect(names(motif_weights), tmp_motifdf$labels)) == 0) {\n",
    "      warning(sprintf(\n",
    "        \"None of the supplied motif weight names match expected motif labels. Example motif label format: %s\",\n",
    "        paste(head(unique(na.omit(tmp_motifdf$label)), 3), collapse = \", \")\n",
    "      ))\n",
    "    }\n",
    "    tmp_motifdf <- tmp_motifdf %>%\n",
    "      mutate(mweight = ifelse(labels %in% names(motif_weights), motif_weights[labels], NA)) %>%\n",
    "      mutate(mweight = ifelse(!all(is.na(mweight)), max(mweight, na.rm = T), NA)) %>% # still grouped by name. this ensures weights are applied to each row of same motif\n",
    "      mutate(mweight = case_when( # clip ends within color range\n",
    "        mweight > max(motif_weight_colors) ~ max(motif_weight_colors),\n",
    "        mweight < min(motif_weight_colors) ~ min(motif_weight_colors),\n",
    "        is.na(mweight) ~ as.numeric(NA),\n",
    "        TRUE ~ mweight\n",
    "      ))\n",
    "  } else {\n",
    "    tmp_motifdf$mweight <- 1\n",
    "  }\n",
    "  \n",
    "  # Plot\n",
    "  p1 <- p1 +\n",
    "    geom_line(\n",
    "      data = tmp_motifdf,\n",
    "      aes(x = x, y = y, group = name, color = mweight),\n",
    "      alpha = motif_line_alpha,\n",
    "      size = motif_line_size\n",
    "    ) +\n",
    "    # vertical labels\n",
    "    ggrepel::geom_text_repel(\n",
    "      data = tmp_motifdf[tmp_motifdf$y > 0 & !is.na(tmp_motifdf$labels), ], # removed the NA rows to prevent warnings for intentionally missing labels\n",
    "      aes(x = x, y = y, label = labels, color = mweight),\n",
    "      direction = \"x\",\n",
    "      arrow = arrow(length = unit(0.5, \"mm\")),\n",
    "      alpha = motif_lab_alpha,\n",
    "      size = motif_lab_size,\n",
    "      segment.size = 0.25,\n",
    "      max.overlaps = 50,\n",
    "      min.segment.length = 0\n",
    "    ) +\n",
    "    # motif labels along x-axis\n",
    "    ggrepel::geom_text_repel(\n",
    "      data = tmp_motifdf[tmp_motifdf$y == 0 & !is.na(tmp_motifdf$labels), ], # removed the NA rows to prevent warnings for intentionally missing labels\n",
    "      aes(x = x, y = y, label = labels, color = mweight),\n",
    "      arrow = arrow(length = unit(0.5, \"mm\")),\n",
    "      alpha = motif_lab_alpha,\n",
    "      size = motif_lab_size,\n",
    "      direction = \"y\",\n",
    "      segment.size = 0.25,\n",
    "      min.segment.length = 0,\n",
    "      nudge_y = -max(countdf$Counts) / 20\n",
    "    ) +\n",
    "    ylim(-max(countdf$Counts) / 10, max(countdf$Counts)) +\n",
    "    xlim(min(countdf$Locus), max(countdf$Locus))\n",
    "  \n",
    "  # Color text by motif weights\n",
    "  if (!is.null(motif_weights)) {\n",
    "    scaled_breaks <- breaks_to_scaledbreaks(motif_weight_colors, na.omit(tmp_motifdf$mweight))\n",
    "    \n",
    "    p1 <- p1 + scale_color_gradientn(\n",
    "      colors = names(motif_weight_colors), values = scaled_breaks,\n",
    "      na.value = \"gray\", # may want to parameterize this\n",
    "      name = motif_weight_name\n",
    "    )\n",
    "  } else {\n",
    "    p1 <- p1 + scale_color_gradient(low = \"black\", high = \"black\", na.value = \"gray\", name = motif_weight_name, guide = \"none\")\n",
    "  }\n",
    "  \n",
    "  return(p1)\n",
    "}\n",
    "\n",
    "\n",
    "#' Plot colors in a color palette\n",
    "#'\n",
    "#' Utility function to quickly visualize a color palette formatted for motif weight visualization\n",
    "#'\n",
    "#' @param pal A palette, named list of values where names are colors and values are the color breaks\n",
    "#' @return Plots a base plot to visualize the input palette\n",
    "#' @examples\n",
    "#' motif_color_pal <- c(-20, -10, 0, 10, 20)\n",
    "#' names(motif_color_pal) <- c(\"darkblue\", \"cyan\", \"gray\", \"orange\", \"darkred\")\n",
    "#' plot_pal(motif_color_pal)\n",
    "plot_pal <- function(pal, cex = 4, cex.lab = 1) {\n",
    "  npal <- length(pal)\n",
    "  plotdim(npal * 1, 3)\n",
    "  plot(1:npal, rep(1, npal), pch = 19, cex = cex, col = names(pal))\n",
    "  text(1:npal, rep(1, npal), labels = pal, cex = cex.lab)\n",
    "}\n",
    "\n",
    "#' Get Gene Body Model\n",
    "#'\n",
    "#' Get Gene Body model for specific gene in region ranges\n",
    "#'\n",
    "#' Used in `plotRegion()` to preformat GRanges for selected genes\n",
    "#'\n",
    "#' @param whichGene Character value. Gene symbol for a single gene to plot.\n",
    "#' @param regionGRanges Granges object for region of interest. Ie, an output of `countdf_to_region()`\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or `getbpInserts()`\n",
    "#' @param orgdb An organism database containing the gene, default `org.Hs.eg.db`\n",
    "#' @param db_id_col Character value. Column in `orgdb` containing the output id. Default \"REFSEQ\".\n",
    "#' @param verbose Logical value, default TRUE.\n",
    "#' @return GRanges object for gene to plot\n",
    "get_gene_body_model <- function(whichGene,\n",
    "                                regionGRanges,\n",
    "                                countdf,\n",
    "                                orgdb = org.Hs.eg.db,\n",
    "                                db_id_col = \"REFSEQ\",\n",
    "                                verbose = TRUE) {\n",
    "  # Get REFSEQ values for gene symbol\n",
    "  # this requires package \"RMariaDB\" available on CRAN and corresponding linux library \"libmariadb-dev\"\n",
    "  txList <- tryCatch(\n",
    "    unlist(AnnotationDbi::mapIds(\n",
    "      x = orgdb,\n",
    "      keys = whichGene,\n",
    "      column = db_id_col,\n",
    "      keytype = \"SYMBOL\",\n",
    "      multiVals = \"list\"\n",
    "    )),\n",
    "    error = function(e) {\n",
    "      stop(sprintf(\n",
    "        \"Error attempting to query symbol for 'whichGene' argument [%s] from specified database. Full error message: %s\",\n",
    "        paste(whichGene, collapse = \",\"), e\n",
    "      ))\n",
    "      return(NULL)\n",
    "    }\n",
    "  )\n",
    "  \n",
    "  # Pull specified database and table for the transcripts by\n",
    "  newTxDB <- GenomicFeatures::makeTxDbFromUCSC(genome = \"hg38\", tablename = \"refGene\", transcript_ids = txList)\n",
    "  if (verbose) {\n",
    "    print(txList)\n",
    "  }\n",
    "  \n",
    "  # Get granges object from database and intersect with region granges\n",
    "  newTxDBgenes <- unlist(genes(newTxDB, single.strand.genes.only = FALSE)) %>%\n",
    "    plyranges::filter(!grepl(\"_\", seqnames))\n",
    "  newGRanges <- biovizBase::crunch(newTxDB, which = newTxDBgenes) %>%\n",
    "    plyranges::join_overlap_intersect(regionGRanges)\n",
    "  colnames(values(newGRanges))[4] <- \"model\" # rename 'Type' column to models\n",
    "  \n",
    "  if (length(newGRanges) > 0) {\n",
    "    newGRangesf <- as.data.frame(newGRanges)\n",
    "    startSide <- newGRangesf[newGRangesf$start %in% start(regionGRanges) & newGRangesf$model == \"gap\", ]\n",
    "    endSide <- newGRangesf[newGRangesf$end %in% end(regionGRanges) & newGRangesf$model == \"gap\", ]\n",
    "    beginexon <- NULL\n",
    "    endexon <- NULL\n",
    "    \n",
    "    if (dim(startSide)[1] > 0) {\n",
    "      beginexon <- data.frame(\n",
    "        seqnames = unique(countdf$chr),\n",
    "        start = min(countdf$Locus) - 1,\n",
    "        end = min(countdf$Locus) - 1,\n",
    "        width = 1,\n",
    "        strand = \"*\",\n",
    "        tx_id = unique(startSide$tx_id),\n",
    "        tx_name = unique(startSide$tx_name),\n",
    "        gene_id = unique(startSide$gene_id),\n",
    "        model = \"exon\"\n",
    "      )\n",
    "    }\n",
    "    \n",
    "    if (dim(endSide)[1] > 0) {\n",
    "      endexon <- data.frame(\n",
    "        seqnames = unique(countdf$chr),\n",
    "        start = max(countdf$Locus) + 1,\n",
    "        end = max(countdf$Locus) + 1,\n",
    "        width = 1,\n",
    "        strand = \"*\",\n",
    "        tx_id = unique(endSide$tx_id),\n",
    "        tx_name = unique(endSide$tx_name),\n",
    "        gene_id = unique(endSide$gene_id),\n",
    "        model = \"exon\"\n",
    "      )\n",
    "    }\n",
    "    \n",
    "    newModel <- rbind(newGRangesf, beginexon, endexon) %>%\n",
    "      mutate(nameList = paste(whichGene, tx_name, sep = \": \")) %>%\n",
    "      makeGRangesListFromDataFrame(., split.field = \"nameList\", keep.extra.columns = TRUE)\n",
    "    \n",
    "    # check for transcripts that are identical within the window\n",
    "    uniqueTranscripts <- stack(newModel)[!duplicated(ranges(stack(newModel)))] %>%\n",
    "      plyranges::filter(model != \"utr\") %>%\n",
    "      as.data.frame() %>%\n",
    "      dplyr::select(tx_name) %>%\n",
    "      unique(.)\n",
    "    newModel <- newModel[grepl(paste(unlist(uniqueTranscripts), collapse = \"|\"), names(newModel))]\n",
    "    \n",
    "    names(newModel) <- rep(whichGene, length(newModel))\n",
    "    \n",
    "    return(newModel)\n",
    "  } else {\n",
    "    print(\"Warning: whichGene did not contain genes names for any genes in the window. Plotting all transcripts\")\n",
    "    return(NULL)\n",
    "  }\n",
    "}\n",
    "\n",
    "#' Common theme for gene plots\n",
    ".gene_plot_theme <- list(\n",
    "  panel.grid = element_blank(),\n",
    "  panel.border = element_blank(),\n",
    "  axis.ticks = element_blank(),\n",
    "  axis.text.y = element_blank(),\n",
    "  plot.margin = unit(c(.5, .5, .5, .5), \"cm\")\n",
    ")\n",
    "\n",
    "#' Generate a ggplot based on a Gene Model\n",
    "#'\n",
    "#' Creates a gene track ggplot based on gene model GRanges\n",
    "#' Used conditinally in `plotRegion()` for gene plot.\n",
    "#'\n",
    "#' @param A gene granges model, ie generated by `get_gene_body_model()`\n",
    "#' @param x_min Numeric value. Min x-axis limit. If NULL (default) will use min value from newModel granges.\n",
    "#' @param x_max Numeric value. Max x-axis limit. If NULL (default) will use max value from newModel granges.\n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param theme_ls Named list of parameters passed to `theme()`. For defaults see `.gene_plot_theme`\n",
    "plot_whichGene <- function(newModel,\n",
    "                           x_lim = NULL,\n",
    "                           base_size = 6,\n",
    "                           theme_ls = .gene_plot_theme) {\n",
    "  # if(is.null(x_min)){\n",
    "  #     xmin = min(newModel$\n",
    "  # }\n",
    "  \n",
    "  # Fill in theme any unspecified theme options with defaults\n",
    "  default_theme <- .gene_plot_theme\n",
    "  unspec_param <- setdiff(names(default_theme), names(theme_ls))\n",
    "  if (length(unspec_param) > 0) {\n",
    "    theme_ls <- c(theme_ls, default_theme[unspec_param])\n",
    "  }\n",
    "  \n",
    "  p_gene <- ggbio() +\n",
    "    geom_alignment(newModel,\n",
    "                   aes(type = model),\n",
    "                   cds.rect.h = 0.25,\n",
    "                   rect.height = 0.25 / 4\n",
    "    )\n",
    "  \n",
    "  if (!is.null(x_lim)) {\n",
    "    p_gene <- p_gene + xlim(x_lim[1], x_lim[2])\n",
    "  }\n",
    "  \n",
    "  p_gene <- p_gene +\n",
    "    scale_y_continuous(expand = c(0.3, 0.3)) +\n",
    "    theme_bw(base_size = base_size) +\n",
    "    coord_cartesian(clip = \"off\") +\n",
    "    do.call(theme, theme_ls)\n",
    "  \n",
    "  return(p_gene)\n",
    "}\n",
    "\n",
    "#' Plot a genebody from a specific database\n",
    "#'\n",
    "#' Used in `plotRegion()` to plot gene track for granges input containing all genes in a region\n",
    "#'\n",
    "#' @param organismdb Database object, for example the output of `OrganismDbi::makeOrganismDbFromTxDb()`\n",
    "#' @param geneBody_gr A GRanges object containing the gene, for example an output of `get_grange_genebody()`\n",
    "#' @param xmin\n",
    "#' @param xmax\n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param plot_theme Named list of `ggplot2::theme()` parameters.\n",
    "#' @param collapseGenes Logical value, default FALSE. If TRUE will collapse all genes into one line.\n",
    "#' sites into a single row\n",
    "#' @return A ggplot object\n",
    "plot_geneBody <- function(organismdb,\n",
    "                          geneBody_gr,\n",
    "                          x_lim = NULL,\n",
    "                          base_size = 12,\n",
    "                          theme_ls = .gene_plot_theme,\n",
    "                          collapseGenes = FALSE) {\n",
    "  # Fill in theme any unspecified theme options with defaults\n",
    "  default_theme <- .gene_plot_theme\n",
    "  unspec_param <- setdiff(names(default_theme), names(theme_ls))\n",
    "  if (length(unspec_param) > 0) {\n",
    "    theme_ls <- c(theme_ls, default_theme[unspec_param])\n",
    "  }\n",
    "  \n",
    "  if (collapseGenes) {\n",
    "    stat_val <- \"reduce\"\n",
    "  } else {\n",
    "    stat_val <- NULL\n",
    "  }\n",
    "  \n",
    "  g_geneBody <- ggbio::autoplot(organismdb, wh = geneBody_gr)#, stat = stat_val)\n",
    "  \n",
    "  if (!is.null(x_lim)) { # had to move this here, otherwise theme would be overwritten...\n",
    "    g_geneBody <- g_geneBody + xlim(x_lim[1], x_lim[2])\n",
    "  }\n",
    "  \n",
    "  g_geneBody <- g_geneBody +\n",
    "    scale_y_continuous(expand = c(0.3, 0.3)) +\n",
    "    theme_bw(base_size = base_size) +\n",
    "    coord_cartesian(clip = \"off\") +\n",
    "    do.call(theme, theme_ls)\n",
    "  \n",
    "  \n",
    "  return(g_geneBody)\n",
    "}\n",
    "\n",
    "##### simplifiedOrgDb: Generates a simplified OrganismDb object for plotting only the longest transcript for each gene. \n",
    "\n",
    "#' Generate a simplified Organism Db object.\n",
    "#'\n",
    "#' Pulls out only the longest transcripts for each gene and repackages them into an organism database object for plotting. \n",
    "#'\n",
    "#' @param TxDb A TxDb object with transcript info. \n",
    "#' @param orgdb A OrgDb object with gene name info. \n",
    "\n",
    "simplifiedOrgDb <- function(TxDb = TxDb, orgdb = orgdb){\n",
    "  \n",
    "  # retrieve transcript lengths\n",
    "  txlen <- transcriptLengths(TxDb, with.utr5_len=TRUE, with.utr3_len=TRUE)\n",
    "  setDT(txlen)\n",
    "  txlen$len <- rowSums(as.matrix(txlen[, .(tx_len, utr5_len, utr3_len)]))\n",
    "  \n",
    "  setkey(txlen, gene_id, len, tx_id)\n",
    "  \n",
    "  # filter longesttranscript by gene_id\n",
    "  ltx <- txlen[!is.na(gene_id)][, tail(.SD,1), by=gene_id]$tx_id\n",
    "  \n",
    "  # filter txdb object\n",
    "  txb <- as.list(TxDb)\n",
    "  txb$transcripts <- txb$transcripts[txb$transcripts$tx_id %in% ltx, ]\n",
    "  txb$splicings <- txb$splicings[txb$splicings$tx_id %in% ltx,]\n",
    "  txb$genes <- txb$genes[txb$genes$tx_id %in% ltx,]\n",
    "  chrominfo <- as.data.frame(seqinfo(TxDb)) %>%\n",
    "    dplyr::mutate(chrom = rownames(.)) %>% \n",
    "    dplyr::rename(length = seqlengths, is_circular = isCircular)\n",
    "  txb2 <- GenomicFeatures::makeTxDb(txb$transcripts, txb$splicings, txb$genes, chrominfo, metadata = \n",
    "                                      filter(metadata(TxDb), \n",
    "                                             grepl('Genome|Organism|Data source|UCSC Table',name)))\n",
    "  Homo.sapiens.hg38 <- OrganismDbi::makeOrganismDbFromTxDb(txb2, \n",
    "                                                           orgdb = orgdb)\n",
    "  \n",
    "  return(Homo.sapiens.hg38)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "##### plotRegion: Plots the region that you've summarized across all cell groupings.\n",
    "\n",
    "# Preformatting for Motif Overlay:\n",
    "# Use any of these lines of code to add a motif position set to an ArchR Project with peak calls\n",
    "### proj<- addMotifAnnotations(ArchRProj = proj, motifSet = \"JASPAR2020\", name = \"JasparMotifs\")\n",
    "########### Other options: motifSet = \"cisbp\", \"encode\", \"homor\".\n",
    "########### name = \"whatever name you want here\". This is \"name\" field is what you need to pass to plotRegion to access the motif positions.\n",
    "\n",
    "#' Plot Region\n",
    "#'\n",
    "#' Plots the region that you've summarized across all cell groupings\n",
    "#' (groups=initial getPopFrags() split) with optional motif overlay, chromosome\n",
    "#' position ideogram, and additional GRanges tracks. If plotting motif overlay,\n",
    "#' ensure that motif annotations have been added to your project peak calls.\n",
    "#' Example: `proj<- addMotifAnnotations(ArchRProj = proj, motifSet =\n",
    "#' \"JASPAR2020\", name = \"JasparMotifs\")`. A basic plot can be rendered with just\n",
    "#' a counts data frame, but additional formatting arguments allow for further\n",
    "#' customization.\n",
    "#'\n",
    "#' @param countdf A dataframe that comes from `getbpCounts()` or\n",
    "#'   `getbpInserts()`\n",
    "#' @param plotType Options include 'overlaid','stacked', or 'RidgePlot'. default\n",
    "#'   is 'stacked', which will plot a seperate track for each group. Setting\n",
    "#'   plotType to 'overlaid' will overlay count plot histograms across samples,\n",
    "#'   instead of faceting out separately. Setting plotType to 'RidgePlot' will\n",
    "#'   generate a ridgeplot across all groups.\n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param counts_color Optional color palette. A named vector of color values\n",
    "#'   where names are unique values in the `color_var` column\n",
    "#' @param range_label_size Numeric value, default 4. Text size for the y-axis\n",
    "#'   range label\n",
    "#' @param legend.position Any acceptable `legend.position` argument to theme().\n",
    "#'   Default NULL will place legend for overlaid plots at (0.8,0.8), or to the\n",
    "#'   \"right\" for faceted plots.\n",
    "#' @param facet_label_side Direction character value, default \"top\". Can also be\n",
    "#'   \"right\", \"left\", or \"bottom\". Position of facet label.\n",
    "#' @param counts_group_colors Optional named color vector. Values as colors,\n",
    "#'   names are levels of `counts_color_var`. If provided, will color the plots\n",
    "#'   specifically using `scale_color_manual()`\n",
    "#' @param counts_color_var Character value, default \"Groups\". Column name from\n",
    "#'   countdf to use to color counts plots. Only used if counts_group_colors\n",
    "#'   provided\n",
    "#' @param counts_theme_ls A list of named theme arguments passed to theme(). For\n",
    "#'   example, `list(axis.ticks = element_blank())`. Default NULL will use\n",
    "#'   `.counts_plot_default_theme`.\n",
    "#' @param ArchRProj An archR project containing motif annotations. Only needed\n",
    "#'   if supplying `motifSetName` arg for motif label overlay.\n",
    "#' @param motifSetName The name of the motif set in ArchRProj to use for\n",
    "#'   annotation. Example: 'JasparMotifs'\n",
    "#' @param customFeatures Optional - A GRanges containing a user-specific set of\n",
    "#'   motifs or other features. If specified these will be used instead of the\n",
    "#'   ArchR motif set.\n",
    "#' @param motif_y_space_factor A factor for vertical spacing between motif\n",
    "#'   labels. Default 4. Increase to make labels farther apart, decrease to make\n",
    "#'   labels closer.\n",
    "#' @param motif_stagger_labels_y = FALSE Logical value, default FALSE. If TRUE,\n",
    "#'   will  stagger motif labels in adjacent columns in the vertical direction\n",
    "#' @param motif_weights Optional numeric vector, default NULL. If provided will\n",
    "#'   be used to color motif labels by the weighted values\n",
    "#' @param motif_weight_name Character value, default \"Motif Weight\". Used to\n",
    "#'   label the legend for motif colors\n",
    "#' @param weight_colors Named numeric vector. Names should be color values and\n",
    "#'   breaks should be the corresponding values of motif_weights. Values outside\n",
    "#'   the highest and lowest value will appear as max or min defined color value.\n",
    "#' @param motif_lab_size Numeric value, default 1. Size of motif labels.\n",
    "#' @param motif_lab_alpha Numeric value, default 0.25. Alpha for motif labels.\n",
    "#' @param motif_line_size Numeric value, default 1. Size of motif lines.\n",
    "#' @param motif_line_alpha Numeric value, default 0.25. Alpha for motif lines.\n",
    "#' @param showGene Logical value, default TRUE. Whether or not the gene track\n",
    "#'   should be plotted.\n",
    "#' @param TxDb A TxDb (transcript database) object used for gene body references\n",
    "#'   when autoplotting all genes in region. Default\n",
    "#'   `TxDb.Hsapiens.UCSC.hg38.refGene`, UCSC Ref Gene from Hg38.\n",
    "#' @param whichGene Optional character value. Gene symbol for a single gene to\n",
    "#'   plot instead of all genes in region. Default NULL.\n",
    "#' @param orgdb An organism database containing the gene regions for `whichGene`\n",
    "#'   plotting, default `org.Hs.eg.db`\n",
    "#' @param db_id_col Character value. Column in `orgdb` containing the output id\n",
    "#'   for `whichGene` plotting. Default \"REFSEQ\".\n",
    "#' @param collapseGenes Options include 'collapseAll', 'longestTx', or 'None'\n",
    "#'   Default 'None' will plot the expanded view of the reference genes,\n",
    "#'   'collapseAll' if you want collapse the gene tracks into one, and\n",
    "#'   'longestTx' will only plot the longest transcript of each gene.\n",
    "#' @param gene_theme_ls Named list of parameters passed to `theme()` for the\n",
    "#'   gene plot. Default NULL will use `.gene_plot_theme`\n",
    "#' @param additionalGRangesTrack A GRanges object containing additional track\n",
    "#'   plot data\n",
    "#' @param showIdeogram Logical value, default TRUE. If TRUE plots the chromosome\n",
    "#'   ideogram at the top of the multi-track plot\n",
    "#' @param ideogram_genome Character value, a genome name for the ideogram plot.\n",
    "#'   Default 'hg19'.\n",
    "#' @param relativeHeights Named numeric vector of relative heights for each of\n",
    "#'   the 4 track plots to enable clean visualization when there are many tracks.\n",
    "#'   Unused tracks will be ignored. Default value = c(`Chr` = 0.9, `Normalized\n",
    "#'   Counts` = 7, `Genes`= 2, `AdditionalGRanges` = 4.5)\n",
    "#' @param verbose Logical value, default FALSE. If TRUE will show all messages\n",
    "#'   generated by internal function calls. If FALSE, will attempt to use\n",
    "#'   suppressMessages() to minimize message output.\n",
    "#' @return The input ggplot object with motif labels overlaid\n",
    "#' @examples\n",
    "#' \\dontrun{\n",
    "#' # my_count_df is a counts data frame generated by getbpCounts()\n",
    "#'\n",
    "#' # Simple counts + ideogram + all genes:\n",
    "#' plotRegion(countdf = my_count_df)\n",
    "#'\n",
    "#' # Motif overlay for a project (my_proj) containing \"JasparMotifs\" annotations:\n",
    "#' plotRegion(countdf = my_count_df, ArchRProj = my_proj, motifSetName = \"JasparMotifs\", motif_lab_alpha = 1, motif_line_alpha = 1)\n",
    "#' }\n",
    "#'\n",
    "#' # Motif overlay w/ weights.:\n",
    "#' plotRegion(\n",
    "#'   countdf = my_count_df, ArchRProj = my_proj, motifSetName = \"JasparMotifs\", motif_lab_alpha = 1,\n",
    "#'   motif_line_alpha = 1, motif_weights = my_enrichment_weights\n",
    "#' )\n",
    "#'\n",
    "plotRegion <- function(countdf,\n",
    "                       # base count plot args\n",
    "                       plotType = 'stacked',\n",
    "                       base_size = 12,\n",
    "                       counts_color = NULL,\n",
    "                       range_label_size = 2,\n",
    "                       legend.position = NULL,\n",
    "                       facet_label_side = \"top\",\n",
    "                       counts_color_var = \"Groups\",\n",
    "                       counts_group_colors = NULL,\n",
    "                       counts_theme_ls = NULL,\n",
    "                       # Motif args\n",
    "                       ArchRProj = NULL, # For overlaying motifs only\n",
    "                       motifSetName = NULL,\n",
    "                       customFeatures = NULL,\n",
    "                       motif_y_space_factor = 4,\n",
    "                       motif_stagger_labels_y = FALSE,\n",
    "                       motif_weights = NULL,\n",
    "                       motif_weight_name = \"Motif Weight\",\n",
    "                       motif_weight_colors = c(darkblue = -10, gray = 0, darkred = 10),\n",
    "                       motif_lab_size = 1,\n",
    "                       motif_lab_alpha = 0.25,\n",
    "                       motif_line_alpha = 0.25,\n",
    "                       motif_line_size = 0.75,\n",
    "                       # Genes plot args\n",
    "                       showGene = TRUE,\n",
    "                       TxDb = TxDb.Hsapiens.UCSC.hg38.refGene, #\n",
    "                       whichGene = NULL, # for plotting specific gene in region\n",
    "                       orgdb = org.Hs.eg.db, \n",
    "                       db_id_col = \"REFSEQ\",\n",
    "                       collapseGenes = 'None',\n",
    "                       gene_theme_ls = NULL,\n",
    "                       single.strand.genes.only = FALSE,\n",
    "                       # Additional Tracks\n",
    "                       additionalGRangesTrack = NULL,\n",
    "                       loopdf = NULL,\n",
    "                       # Ideogram\n",
    "                       showIdeogram = TRUE,\n",
    "                       ideogram_genome = \"hg19\",\n",
    "                       # Combined Tracks\n",
    "                       relativeHeights = c(`Chr` = 0.9, `Normalized Counts` = 7, `Genes` = 2, `AdditionalGRanges` = 4.5),\n",
    "                       verbose = FALSE) {\n",
    "  # Validate input\n",
    "  supported_tracks <- c(\"Chr\", \"Normalized Counts\", \"Genes\", \"AdditionalGRanges\")\n",
    "  if (length(setdiff(names(relativeHeights), supported_tracks)) > 0) {\n",
    "    warning(sprintf(\n",
    "      \"1 or more values of relative heights not in supported tracks: %s.\\n Supported track names: %s\",\n",
    "      paste(setdiff(names(relativeHeights), supported_tracks), collapse = \", \"),\n",
    "      paste(supported_tracks, collapse = \", \")\n",
    "    ))\n",
    "  }\n",
    "  \n",
    "  # function wrapper based on verbosity to hide messages\n",
    "  verbf <- function(x) {\n",
    "    if (verbose) {\n",
    "      x\n",
    "    } else {\n",
    "      suppressMessages(x)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # Variables\n",
    "  chrom <- toString(unique(countdf$chr))\n",
    "  .relativeHeights_default <- c(`Chr` = 0.9, `Normalized Counts` = 7, `Genes` = 2, `AdditionalGRanges` = 4.5,  `Loops` = 1.5) # retain in case any missing\n",
    "  \n",
    "  # Extract region from countdf as granges\n",
    "  regionGRanges <- countdf_to_region(countdf = countdf)\n",
    "  \n",
    "  # Base Plot of Sample Counts\n",
    "  p1 <- verbf(\n",
    "    counts_plot_samples(countdf,\n",
    "                        plotType = plotType,\n",
    "                        base_size = base_size,\n",
    "                        counts_color_var = counts_color_var,\n",
    "                        counts_color = counts_color,\n",
    "                        range_label_size = range_label_size,\n",
    "                        legend.position = legend.position,\n",
    "                        facet_label_side = facet_label_side,\n",
    "                        counts_group_colors = counts_group_colors,\n",
    "                        theme_ls = counts_theme_ls\n",
    "    )\n",
    "  )\n",
    "  \n",
    "  # Add Motifs to Base Plot if Requested\n",
    "  if (\n",
    "    (!is.null(motifSetName) & !is.null(ArchRProj)) || !is.null(customFeatures)\n",
    "  ) {\n",
    "    p1 <- verbf(\n",
    "      counts_plot_motif_overlay(\n",
    "        p1 = p1,\n",
    "        countdf = countdf,\n",
    "        ArchRProj = ArchRProj,\n",
    "        motifSetName = motifSetName,\n",
    "        customFeatures = customFeatures,\n",
    "        motif_y_space_factor = motif_y_space_factor,\n",
    "        motif_stagger_labels_y = motif_stagger_labels_y,\n",
    "        motif_weights = motif_weights,\n",
    "        motif_weight_name = motif_weight_name,\n",
    "        motif_weight_colors = motif_weight_colors,\n",
    "        motif_lab_size = motif_lab_size,\n",
    "        motif_lab_alpha = motif_lab_alpha,\n",
    "        motif_line_alpha = motif_line_alpha,\n",
    "        motif_line_size = motif_line_size\n",
    "      )\n",
    "    )\n",
    "  } else if (!is.null(motifSetName) & is.null(ArchRProj)) {\n",
    "    stop(\"No ArchR Project found. If you wish to plot motif positions overlaid, please provide an ArchR project along with your motifSetName.\")\n",
    "  } else {\n",
    "    p1 <- verbf(\n",
    "      p1 + xlim(min(countdf$Locus), max(countdf$Locus))\n",
    "    )\n",
    "  }\n",
    "  \n",
    "  # Build P2, Ref Genes track\n",
    "  if (showGene) {\n",
    "    # If user provided a specific gene symbol(s), pull new granges from database and format models\n",
    "    if (!is.null(whichGene)) {\n",
    "      newModel <- verbf(\n",
    "        get_gene_body_model(\n",
    "          whichGene = whichGene,\n",
    "          countdf = countdf,\n",
    "          regionGRanges = regionGRanges,\n",
    "          orgdb = orgdb,\n",
    "          verbose = verbose\n",
    "        )\n",
    "      )\n",
    "    } else {\n",
    "      newModel <- NULL\n",
    "    }\n",
    "    \n",
    "    \n",
    "    if (!is.null(whichGene) & !is.null(newModel)) {\n",
    "      # Successful newmodel generated\n",
    "      p2 <- verbf(\n",
    "        plot_whichGene(newModel,\n",
    "                       base_size = base_size,\n",
    "                       x_lim = range(countdf$Locus),\n",
    "                       theme_ls = gene_theme_ls\n",
    "        )\n",
    "      )\n",
    "    } else {\n",
    "      # Create your reference for plotting gene body\n",
    "      \n",
    "      if(tolower(collapseGenes) == 'longesttx'){\n",
    "        \n",
    "        Homo.sapiens.hg38 <- simplifiedOrgDb(TxDb = TxDb, orgdb = orgdb)\n",
    "        \n",
    "      }else{\n",
    "        \n",
    "        Homo.sapiens.hg38 <- verbf(OrganismDbi::makeOrganismDbFromTxDb(TxDb, orgdb = orgdb))\n",
    "      }\n",
    "      \n",
    "      geneBody <- verbf(get_grange_genebody(regionGRanges, TxDb = TxDb, single.strand.genes.only = single.strand.genes.only))\n",
    "   \n",
    "      if (length(geneBody) > 0) {\n",
    "        p2 <- verbf(\n",
    "          plot_geneBody(\n",
    "            organismdb = Homo.sapiens.hg38,\n",
    "            geneBody_gr = geneBody,\n",
    "            collapseGenes = tolower(collapseGenes) == 'collapseall',\n",
    "            base_size = base_size,\n",
    "            x_lim = range(countdf$Locus),\n",
    "            theme_ls = gene_theme_ls\n",
    "          )\n",
    "        )\n",
    "      } else {\n",
    "        print(\"No gene body in range\")\n",
    "        showGene <- FALSE # to prevent errors resulting from p2 nonexistence\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## TODO process additional granges\n",
    "  if (!is.null(additionalGRangesTrack)) {\n",
    "    \n",
    "    # Check for name metadata column\n",
    "    if (\"name\" %in% colnames(mcols(additionalGRangesTrack))) {\n",
    "      # Only plot the overlap of this region and the additional GRanges Track\n",
    "      overlapGRanges <- verbf(plyranges::join_overlap_intersect(additionalGRangesTrack, regionGRanges))\n",
    "      if (length(overlapGRanges) > 0) {\n",
    "        # Use the subset within our region as the track we want to plot\n",
    "        print(\"GRanges has overlap\")\n",
    "        # Assign exon status to each row\n",
    "        mcols(overlapGRanges)$type <- rep(\"exon\", each = length(overlapGRanges))\n",
    "        # split into list of GRanges -> GRangesList named for the names metadata col\n",
    "        additionalGRangesTrack <- split(overlapGRanges, overlapGRanges$name)\n",
    "      } else {\n",
    "        print(length(overlapGRanges))\n",
    "        print(\"No overlap of additional GRanges and this region\")\n",
    "        additionalGRangesTrack <- NULL\n",
    "        # Avoids the following error:\n",
    "        # Error: Faceting variables must have at least one value\n",
    "        # Error in deparse(x[[i]], nlines = nlines) :   no slot of name \"elementType\" for this object of class \"GRanges\"\n",
    "      }\n",
    "    } else {\n",
    "      print(\"Additional GRanges does not have `name` in metadata - ignoring additional track\")\n",
    "      additionalGRangesTrack <- NULL\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Generate Loop track\n",
    "  if(!is.null(loopdf)){\n",
    "    \n",
    "    p5 <- ggplot() + \n",
    "            geom_curve(aes(x = midPeak1, xend = midPeak2,\n",
    "                                    y = y, yend = y, color = Correlation), \n",
    "                                curvature = -.5,\n",
    "                                data = cbind(loopdf,y= rep(0, dim(loopdf)[1]))) +\n",
    "           theme_minimal() + ylab(NULL) + xlab(NULL) +  ylim(-1, 0) +\n",
    "           scale_colour_viridis_c(breaks = c(ceiling(10*min(loopdf$Correlation))/10,\n",
    "                                        0,\n",
    "                                        floor(10*max(loopdf$Correlation))/10))  +\n",
    "            coord_cartesian(ylim = c(-0.25, 0))  +\n",
    "      theme(panel.grid = element_blank(), panel.border = element_blank(),\n",
    "             axis.text.x = element_blank(),  axis.ticks.x = element_blank(),\n",
    "             axis.text.y = element_blank(),  axis.ticks.y = element_blank())\n",
    "    \n",
    "  }\n",
    "  \n",
    "  # Combine plots P1...P4\n",
    "  # Base tracks are:\n",
    "  # 1: Chromosome Ideogram\n",
    "  # 2: a) Normalized Counts\n",
    "  # 2: b) Optional: Loop track\n",
    "  # 3: Ref Genes\n",
    "  # Additional GRanges track is user-defined and labeled according to its\n",
    "  # 'names' metadata column\n",
    "  # 4: AdditionalGRanges\n",
    "  ## TODO\n",
    "  \n",
    "  # Construct Named plot list top to Bottom by appending to track list if present\n",
    "  track_list <- list()\n",
    "  \n",
    "  if (showIdeogram) {\n",
    "    # Show Ideogram\n",
    "    p3 <- verbf(Ideogram(genome = ideogram_genome, subchr = chrom))\n",
    "    \n",
    "    track_list <- c(track_list, list(\"Chr\" = p3))\n",
    "  }\n",
    "  \n",
    "  # Counts\n",
    "  track_list <- c(track_list, list(\"Normalized Counts\" = p1))\n",
    "  \n",
    "  # Loops\n",
    "  if (!is.null(loopdf)) {\n",
    "    track_list <- c(track_list, list(\"Loops\" = p5))\n",
    "  }\n",
    "  \n",
    "  # Genes\n",
    "  if (showGene) {\n",
    "    track_list <- c(track_list, list(\"Genes\" = p2))\n",
    "  }\n",
    "  \n",
    "  # Additional Ranges\n",
    "  if (!is.null(additionalGRangesTrack)) {\n",
    "    print(\"Combining base tracks with an additional GRanges track\")\n",
    "    p4 <- verbf(autoplot(additionalGRangesTrack))\n",
    "    track_list <- c(track_list, list(\"AdditionalGRanges\" = p4))\n",
    "  }\n",
    "  \n",
    "  # height params\n",
    "  if (!all(names(track_list) %in% names(relativeHeights))) {\n",
    "    missing_heights <- setdiff(names(track_list), names(relativeHeights))\n",
    "    append_heights <- .relativeHeights_default[missing_heights]\n",
    "    relativeHeights <- c(relativeHeights, append_heights)\n",
    "    warning(sprintf(\n",
    "      \"Relative heights were not defined for included plots [%s]. Using defaults for these tracks [%s]\",\n",
    "      paste(missing_heights, collapse = \", \"),\n",
    "      paste(append_heights, collapse = \", \")\n",
    "    ))\n",
    "  }\n",
    "  trackHeights <- relativeHeights[names(track_list)] # ensure intended order\n",
    "  # Plot All Supplied Plots\n",
    "  g_tracks <- verbf(\n",
    "    ggbio::tracks(\n",
    "      track_list,\n",
    "      heights = trackHeights,\n",
    "      xlim = range(countdf$Locus),\n",
    "      track.bg.color = \"transparent\",\n",
    "      padding = unit(-1, \"lines\"),\n",
    "      label.bg.fill = \"transparent\",\n",
    "      label.bg.color = \"transparent\",\n",
    "      label.width = unit(2, \"lines\")\n",
    "    )\n",
    "    # coord_cartesian(clip = \"off\")\n",
    "  )\n",
    "  \n",
    "  return(g_tracks)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "73a750a4-efc4-43d4-ab07-60b90b0535bf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "projHeme1 <- loadArchRProject(path = '02_TEAseq/PenSen_ATAC/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f60297bd-02e8-43a8-afb0-b06c4e810f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data<-read.csv('02_TEAseq/meta_data_GEO.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "acb2b31f-b71d-4772-a99b-d6fd7af99c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data<-meta_data %>%\n",
    "  mutate(Age_Group = case_when(\n",
    "    str_starts(subject_id.ch1, \"BR\") ~ \"Older\",\n",
    "    TRUE ~ \"Children\"\n",
    "  ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a0ae7b46-f7a5-4875-9bc5-d89c1a2aeb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_meta<-meta_data[c('combined_sample_id','Age_Group')]\n",
    "colnames(df_meta)[1]<-'Sample'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cf63d9f1-0cf3-48cf-8528-184569b5b48e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(Sample)`\n"
     ]
    }
   ],
   "source": [
    "df<-left_join(as.data.frame(projHeme1@cellColData),df_meta)\n",
    "projHeme1@cellColData$Age_Group<-df$Age_Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e61dac-955a-48b7-8e3d-5cbb7a4a51be",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_meta<-meta_data[c('combined_sample_id','subject_id.ch1')]\n",
    "colnames(df_meta)[1]<-'Sample'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93374358-478f-482a-a282-a09cd756ce76",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(Sample)`\n"
     ]
    }
   ],
   "source": [
    "df<-left_join(as.data.frame(projHeme1@cellColData),df_meta)\n",
    "projHeme1@cellColData$subject_id.ch1<-df$subject_id.ch1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "74dd9350-5a44-4d57-be16-f58b1f2095d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Dropping ImputeWeights Since You Are Subsetting Cells! ImputeWeights is a cell-x-cell Matrix!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "idxSample <- BiocGenerics::which(projHeme1$AIFI_L3 %in% \"CM CD4 T cell\")\n",
    "cellsSample <- projHeme1$cellNames[idxSample]\n",
    "projHeme1_subset<-projHeme1[cellsSample, ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "915306c5-d25a-4670-b9f2-725202f6f835",
   "metadata": {},
   "source": [
    "## plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a38aa7dd-ad59-48eb-9e5e-8ccdbf619af7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "492455\"\n"
     ]
    }
   ],
   "source": [
    "pop_frags <- getPopFrags(ArchRProj = projHeme1_subset,\n",
    "                         metaColumn = 'Age_Group', \n",
    "                         numCores=60,sampleSpecific = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "69ea7d3f-8aed-4acc-bd84-f04c04e29107",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in all(lapply(seq_along(tmpCounts), function(x) {:\n",
      "\"coercing argument of type 'list' to logical\"\n"
     ]
    }
   ],
   "source": [
    "countdf <- getbpCounts(regionString = \"chr5:132,672,086-132,684,576\", \n",
    "                    popFrags = pop_frags, numCores = 16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91bab07d-cf67-4880-b5cc-e53d003d18b8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "eeeafd90-6924-4217-8033-b45ed27d7c65",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mCoordinate system already present. Adding new coordinate system, which will\n",
      "replace the existing one.\n",
      "\u001b[1m\u001b[22mCoordinate system already present. Adding new coordinate system, which will\n",
      "replace the existing one.\n",
      "Warning message in !vapply(ggl, fixed, logical(1L)) & !vapply(PlotList, is, \"Ideogram\", :\n",
      "\"longer object length is not a multiple of shorter object length\"\n",
      "\u001b[1m\u001b[22mCoordinate system already present. Adding new coordinate system, which will\n",
      "replace the existing one.\n",
      "\u001b[1m\u001b[22mCoordinate system already present. Adding new coordinate system, which will\n",
      "replace the existing one.\n",
      "Warning message in !vapply(ggl, fixed, logical(1L)) & !vapply(PlotList, is, \"Ideogram\", :\n",
      "\"longer object length is not a multiple of shorter object length\"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsUAAAI3CAIAAAD887FdAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdd1hT1xsH8DfsjaAoS3Ggorh33XuvWnetWtuqVWur1V/dxWrVWltXXdXa1tZZ\n615VVJy4cAIiIHvvlZ3c+/sjigGSkHADIfD9PE+fknvPvXkDSL4599xzeCzLEgAAAAAHJoYu\nAAAAAIwe8gQAAABwhTwBAAAAXCFPAAAAAFfIEwAAAMCVGZeDpTlJCdki8xqenk4W+ioIQIOk\npKTk5GRDVwGgs1WrVuXk5KxZs6Zv375EJBAIvvrqq8ePHy9evNjb29vQ1VU0Nzc3d3d3Q1cB\nesbjcr9oyOqWLfyCfb99EezXQo81AajEsqxYLJbL5cob7ezsDFUPgPa+rlfvRUHBjYKCy1eu\ndOvWbejQoS8CAsbVrp3o5PTvixeGrq6i9e/f/8CBA25uboYuBPSJU/+Eo6MjEYnFYj0VA6CJ\nXC6/du1ao0aNnJ2dCzc+evRIIBCkpKSUbN+wYcOSG0NCQn777Tftn/Tnn38uuTEtLe3JkyfK\nWwYNGlT4dUxMzPnz5zWcc+7cuSq3q3whKl+FgsrXorJgDe3VHa78GhWvTvPhigO1aVPm2tSd\nsOSPo5Dyz0Vzy8LGKn98w4YNU/kzVfejLOnmrFmu5uYsy06fPv3LL78MvHp1Yd26dqamHp07\nt5s4UXPZCup+z0n9L4m672qpP6xiP6mS3zdxUvCTOJMGHZrXMVNRsHKpJWsLCQmZNm1afHw8\n8kQVwylPuLdv70Z3Yl69ElN7S31VBKDGs2fPBAKBTCZzcXEp3Oji4pKfnx8XF1eyva+vb8mN\ncrlckYO11L59+5IbExMTc3Jy1DWztbUNCgrS9ZxEpPKFqHwVCipfi7qTq2uv7nDl16jYqPlw\n7duUuTZ1Jyz541D3jBpaFjZW+ePz9fVV+TPV8IqKeWxnl5uT09fJ6WFc3OrVqwc5O9uZmhJR\nzZo1bRo00Oa06n7PSf0viarvavMRsxqW+sMq9fvGF0U8iTOtWa9BA0sV7ZVLVartxc4Pl1/K\nJyIaMWLE2rVriYi6Lj6zpIfKGsDocMoTJt3nL+m5/8uzGzc8/eDbNkgUUL7kcvmKFSsePXpk\n6EIAysiMx+vk4OCfldVOfbdTOXPyHTZT8ZWLg6UFkaRin96q78pFXdnOnTvfv3+/U6dOFfvk\nUL445Yn8iMSGny3oF/69X7eWT+fMHNy2kYeDuXIDhya9ejax51YhwDtpaWk2NjaGrgKg7Jrb\n2FzJymJYlng8Qzx/8Oldp7p+PpqIGvi02r/lm4tnTl+4GZYtM0QtULVwyhOxh+aMWB1CREQR\npzYtPlWiAYZqgh7xeLylS5dKpVJLS3SGgbFys7Q05fFSJBIvKytDPH/uy4v76fPRRJSUJbD1\n6jbhq24fTAu/eeH0mYt3o/LkRRuzuWGXjh6/EhgSlyPm2TrWqefbqpmH/bu3DYYf//TR8+i0\n4ycuurXsP33ORx1qKULS83++/SG8/WcLWif9/Ofmh1Hpdcbt2DKhAl8lGAKnPOHcYcKsWYka\nGnh0cNawF0Anbdq06dixY0xMTP369Q1dC0AZmRA5m5llyWRehq4kITJkw45j3QcOHTq4e98P\nF/cdl/Hi+rnTZ/57GM9XNMi5v23R+lvm3aZN+bph4suHotzUmJCI5Drt6r5535ClPn9iXbtF\nn+EOY96rs3v9jg1bXfauGeL09vxs5t2TN9zHfvHTTOuo+1FvjhEHbt18u2D4sMFb9xzsHG82\naXQ7F9OKfuFQTriNxxy+cvdwfVUCUIq8vLyxY8f+8MMPyBNg1GxMTUUMY+gqiIhk2eEBR8MD\n/vmtQedB46dO7DZoestBE5OfBhARyYMP77ya6T191+IR8sTEtAhL85r1Wvasp3R0Ft/+ve4N\nHIjIqfGgme8HfLz3v+spQ8a4vtnNj6FWy/q2q+tI1LZfLSLKqdXm/S8G9qG8hKGDhv2+rkXA\nQb+HwfM2rRpYxyBXfkDfMD8mGI3Xr18HBAQYugoArnhEXCb+0TNTR++ug0aMHNLRw4KIyUvK\ntGk5mIhIZiY34zXt1ctV7ZH2bu4OhQ9qetWzo5SUVKX9DZo1K3Jl0mf84o/7t67vYGkhFeY1\n7jp2xfiGeUH7Dz+S6vkVgYFw6p8ogpVJJDKm6L8REzNLCzMkTwCAd0QMY2li+M9yPAu7PpMW\nDhvcrYmTOTEFMYEnzp49dyM4g63d7t99fmTp8+nsvvskGq5ZW9kojwCxsbEhAb9AaYOjo8Y3\nGJ5Xl45uh46GhsZSx2o3Q2iVxDlP8EMPf++3/dj1J9EZohIdeBiPCXrk5ua2aNEizIEDRo0l\nypbJapjp77Oczt4sj9C2dfOObZqzgvgHp86dOXftedrbmQnTHiv+b9XC19o/i6immvPwWI2f\nFk1MDZ+ZoCJx+51mw7cMeW/BrTwiMrd3rV/bttjAmnrOGIcPelO3bt0ff/zR0FUAcJIplYoY\nxs3CUGse+Xy4Y5XiK5449+yfm89efZwiUtPW2ibrxo2UEWPKpxI27v6jZLLp06xe6W3BGHDK\nE/Irm767lUdmLT4/furnUY0McvMTVB9JSUkBAQHNmjVr27atoWsBKKNXAoEpj2dtaqi7GmrW\n9XzzZ//p81d7zz/W0FQQmZgWefS7TeYf9G4kk0lEuSkxoRmO7xXe36GbzCtbdyQ0G9mrJV8q\nM7eyj7z3774br+3bzZ3YEctJVhHc5p94+jSbyGas35ZRjfAbAeUtISHB39+/T58+hi4EoIxY\nogf5+WYWFi/5/Oa2toYoIfSvedO77jhCRPLSmtp4j12x3ubQsSsHfvwjV2pqV6OOV/OWbmV9\n06jZY+KIq2fO7TwZGpMysG+3i8HSLlP8JoxsVwtD7KoKTnnCzMyMiFzr1UOYgIpx+vTpPXv2\nGLoKAJ3JZDIiepyfL7C2/nz69CM7dzaxsTHjcVrhuUyyE+O1b8xzbDbs82+HjUxMvHnzpvIO\n22YDJzUr2tZ72r4z094+aDVu9friJ7Oq03bYZ22H0YMHDzp37nz//upOndrpWj1UZpzyhGeX\nLh4UmBAVJaWO5qU3B+BqzJgxZmZmxVZmUrfuokik4rJwWFhYbm6u9s+ochWotLS06Ohodc1i\nYmLUrQOp4Zyk5oWofBUKKl+LhqXItHnthYcrv0bFRs2Ha9+mzLWpO2HJH4e6Z9TQsrCxyh9f\nSEiIyp+p5oXflCUlJSWJRCcyMvb89tu4ceMuXrx4MCFhQu3aT58+dS+xKJfK02pYX1TdL4m6\n72qpPyydvm8l2yuXWrK2sLAwDacCI8ZyEv/bECey6LLplZTbeQBKFxcX161bN0P/iwEoiwYu\nLpaWllu3blX8MsfGxrZv397SxKRHx46GLs0AunXrFhcXZ9i/J6B3uvW25YffvBGeV2RTwdOd\nX/tdErWbseDjvr51iy4HhvXAQM9yc3Pz8t79BqalpRmwGADtRURE9O7d29X13exQDMPcunXL\nzs7OgFUZipubm7u7u6GrAD3TLU8E+7Vo+WYBMK1g/gkAAIDqQLfxE6UuAFYM1gMDAACoDip+\ndDEAAABUNZgPFQAAALgqQ55gbq/p36VL98UX+epa8C8u7t6lS/81dyrFirwAAABQznTPE8zN\nv3dcvf/IskkHtZO72XZobPHo/tWdhxAoAAAAqgPd88SDc+dSibqMGOGivo3LiBGdiVLOnn3E\noTQAAAAwEjrnCUl0dBKRpZeXq6ZWbl5elkQJ0dHSspcGAAAARkLnPJGSnMwSuZU2FYm7uysR\nm5ycWtbCAAAAwGjo3j8hkRCRiUkpByoaiMXispUFAAAARkTnPOHi4kJESTExGq9kSGJikgsb\nAwAAQNWmc55w7NDBm0h09Zy/UH0jwZWzV0VETTp2dOBSHAAAABgF3e/vaDV2XFOizMPLvn+o\nZh1lwb01y45kEc9n3Fis3QEAAFAN6J4neK0XrJ3gQtKn3w/qt/BwcHaRKSaYrOcHv+o7ZMNz\nGdWevG5BS57eCgUAAIBKq2zrd+TfXNZn2PqgAiKyqu3bsVPzejVtSJARF/rwYUiaiIgcOq08\nf/W77tVxIV4AAIDqp8zrgQlDD62Yv2zX1djiwyisG/Sfs3772gk+VtyrAwAAAGPAbX1RWXZE\n4PXr90PiUrMFPBun2vV8u/Tp26VxDd1WQQcAAADjhvXKAQAAgCusVw4AAABcIU8AAAAAV8gT\nAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gTAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gT\nAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gTAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gT\nAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gTAAAAwBXyBAAAAHCFPAEAupJeXz927IJjiepb\nPNg2aeznB16/eRT++8yxU34JUn++m+vHjl1wLE6/VQJARTIzdAEAUBmx+ZHXT529+jA0PkPA\nc6jt6dmo7YAxIzq7WxIRMekvAwMT3hOqPzwr4l7g42YFbx4VxAQF3rfJVtv6zfkE+nwBAFCx\n0D8BAMUwKdfWjOjSd+4fL21aD/7ws5nTRnSqkxewYXyXtl/9J9XuFJ3mHz6+a1qj8q0TACoR\n9E8AQBGiJz9+OGN3zsBtl3d+UPftX4i+wyZ9PP2PeX/kM9qdxLlxl/fKrUIAqISQJwBAWfLB\nNTvDLIfs3vQuTCjYN5++b424yDZ5WuCfO/++EZFr6dFx1MyZw7yt3+55sG3SxpdDftg1VXUX\nhSzlzl97jgSE55i7th726eyhRXamnVk654DFzN+XNb63b9+JB6+z2iw8uqATEckyHp86ePz6\n09gsqbVr0x5jPp7Uw9PizUHS6+sn7Ugbv3VT7zg1JQFAucL1DgBQknH5/H2ZRa+Rgx1U7DS3\ntOS9e5R1cdn8g3k+w6ZM7ucUsnPmiC9PZ73bF3Ev8HFMgYpzEDGxBz8eNH7dLdPO7384rrvj\nvW9n/hIqV9ovTnoRGBh8d9/n/ztPbUaMG+hGBUQkerFjYp9RfpflLYd99OmH/dyjd300YOLO\n4LfXX5j0l4GBz56dUF8SAJQr9E8AgJLIyEiiet7e5qW2fH0qeN7NPWNqEhENaCcL6+y3/UDU\nqK8alnZczgm/Vdek7+87uWWIPRHRoCFNl/f5iKhoT8ajM/G/3toyyJaIBspkJH/605z1D+t8\nffH0guZmRET9hnR1er/Xd9/8Puz8TC+uJQEAd+ifAAAlYrGYyMLSovSW5n3Hja759oFbp86e\nFPEqvPThFZLrl66JnEdMHmz/dotVz4mj3Is3cxw6bqCt4kszMzN68u/xKLbz1I+bv/sI5DVm\nTAf508vX0jiXBAB6gP4JAFBSq2ZNooyMDCLvUlq6eXoqfR5xcnIiWXpGLpGT5sPSkpJk5OHp\nqXThhOp6ehJlFmnmWbeuUgtRTEwamfEOzppwnliWffNfVjiRXXIyUW1uJQGAHiBPAIASn86d\nHejYgwex1MVLc0tzM+U/Hzwej4hYttQnsLa2Jkrl85W3FRR9SERkY2Oj/FxWViZk4Tvqqzlt\ni7WzdFe6mlHGkgBAD5AnAECJac8Z05sc3/bnL9c//rGPfdF9uXcCojv3bsPtr0bN1q08aP+j\noDRq9rZbIffhwwiiupqq6tCpvemFuLwand5rZsrp6QGgnGD8BAAoM2m5cNfK98SH50z67nx4\n/tsP96KkwL2fj5i6P1yu8WBttJ82q71Z4NYV/8RIiIhkiZfWno9xLeUg18nfzGgQs+erVRdi\nCmflFMTe2L/tXAznggBAH9A/AQBFmfvMPHKp0Y71P60Y3GqJW5P6tcwEiZEReTXfG71wTtfS\n7/soVf0Z+/alzlywqHfbzU08zXJlrRbv+Cz14sIEjQfZvrfqxEHn5d8u6t1meb0G7rbClJhM\nmw4fzPlfTY2HAUBF4bG4uggAKrHizNio+EyBiUMdD0/XmtaFHz+YtLD7r0WebdrULZwtSpz0\n4nGsScOOvnXMiIiyIu69yndr2c7LjogoPyYoONXRp7O30sBIWXZ0aFSOWR3vpp72vPSw+5Ei\nz9Zt6tqUPFVRsvyEyNcpArOanvXr1rZ716D0kgCgXCFPAAAAAFcYPwEAAABcIU8AAAAAV8gT\nAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gTAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gT\nAAAAwBXyBAAAAHCFPAEAAABcIU8AAAAAV8gTAAAAwBXyBAAAlD/ZkdEmdVr27t2zc5NaNXy/\n9s8tv6fK9F/YqpZri7ZNankO3B0q02FvsJ+PU5PuvXv37t2799JLksLtiX9Pae1p331LSvkV\nbfyQJwAAoEJYDFgfEHDzfviTb2vsOnCDiIjSQwICAm7cCQpPEynaSBKfPIoVSrMinjyOyi58\ns5dkRj59+jpbmh91LziVVWyT5cUHP3oUmiJQPJYmPb0VlkVEFLFr0bFOf4cHPwl7vFS+0u9E\nQZEiNO8lajF5/Q+bdh85f339YAvFFiZyy9fXRq+e6Kjv70cVgzwBAAAVgkl9ERBw7eKh3Wdj\nWrVqSkREocf8/Py+XTJ7WOsWH/2TRkRZJ78Y+/G43j0/nDWpU+PBv0YRUf7VRe0bd5s8c0L3\n/uOmjVxxRU5EKadntfXp/tEX88a2azZkZxhDJHj0x7pTkUQkDXoU3mvUIAcik3rvj2r86FG4\ncg2a99rW72QasGLhZwN8GvRYc0dARCR5vG5F4qwtY10r5FtkzMwMXQAAAFQPsueH/Fb9G//8\npfWcgPmKPNFr9aVjr0Jfp+bFH5qz6WD8uAXmRLlOUx9fG+/MPlvZ4otzSTMnnVlzuP2hyD8G\n2zNxv/T39iciCvzxy1tdf/xjkgeP+P4rP9tybc7u/iO3XBxJRCTg861sbXlERGRra5ufn8+m\nBd8IzSAydW3Zw63EXuUKG0w/EDCdiORJf49pteT3z25Nj1z61dMuK7o/DngcL84V3Q2KG9O+\nXoV9w4wL8gQAAFQIiwHrA/4eTrn/fdZh1sbJQSta8CJ29+vtl+jWqLadeUZ6kn0SkReRV9t2\nzkTE8/Bwy8vIo8yMDE9fX3siMqnn28zen4jY5OT07OC9q18REZFpk65yEZHVm2dx9PSUH4zM\npL41iSIiIuvWrcsE7/f77jaR9eD1F5eU2KuqUlP3zh081qSlU/KrdJOMoHV+pyjvdV6c5aZf\nu47ZM7lCvlnGB3kCAAAqkuOgdau2tV31z9wTox5fferz6a+Lepqn3v5lWSTLqmreqE+ftA+/\n3Oj1WXPB1c0n8h268ojXfeQQx6T6H84f0MjZ0oRsG3SyImnS03t59Xr4OFPvSRM+mTVzY81p\n7oE/7nedEtjQtNHagL5vTycuvpdInPDkkaB+tyZOgthHD6ILWGluxKWfthcMPd2UvFseCPiE\niIjuLfJc5HkCYUI9jJ8AAIDyx6vdomezmoqvXSZ/t5B3/b8ky/e/39E+6vcd+y6mvDflk/ca\nORBZeLTt4GVNRETm7q07eNmQebcN59c1eX5wz9GX7UYO8KhVy5So9sQ/Ts6r8ejo9nWr/fz8\ndt7MUBo/QZYDt/mv8wn9e+8Z/qjDpxY0KlqHir159/ZvPBdFRBk3d/v5+X334283pKMOXt/w\nnoXScQ6N3mvraUGgFo9VHQgBAAAqhfyo+0FxAnHWy+Nrvs9bFnJ0XA1DVwQq4HoHAABUakn+\n2/wOJZk7erT+/OSPCBOVFfonAAAAgCuMnwAAAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngAA\nAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngAAAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngAA\nAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngAAAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngAA\nAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngAAAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngAA\nAACukCcAAACAK+QJAAAA4Ap5AgAAALhCngCg/x49UHwhZxiRRGLYYgAAjBHyBAAFhb9SfPEq\nPu6rndsNWwwAgDFCngCgW8HPt508TkRyhnn6OsLQ5QAAGB/kCQCKS0u7ExJMRHKGYVlDVwMA\nYISQJwCIWDYsLq5AKEzMSGcRKAAAdFcOeUKWnxgdlyPV/4kBygnDsnkCvkgiyRcIWEKeAADQ\nGcc8wb48tnLRov/tfyRSPM4MWNnN1dGzoZdLvb4b7hdwrw+gwnz0w9rtp07gggcAQBlwyxPC\ni+vnrP1pd5hVIysiItkdv6lr78nrN6trJ0+5vnTa+ud6qRGgnCkyRFhcXFJmBq53AACUAbc8\ncf34v5lk0m/cGCciIvbWoSPxrb69Ex766sLn9Yle7f/9jj5qBKhIiBMAALrjlCdSoqIERJ7e\n3lZERPTqzp2M5qPHNTcj6x7vD6pFlBIWlquXKgEqjEgiCYoI54uEhi4EAMCYcMoTQqGQiExM\nFCcRP3ny0qZdu6ZEROTs7ERE6enpXAsEKH8MwxR+LZSIb794HpWcjEgBAKA9TnnC3cODiBJD\nQ/OISH434LakXYcOijMmJ6cQkatrHe4lApS3Ypc4GJYtEAoFYrFhqgEAMEJmXA627N2/h9nJ\nW/+t+WSj2cD4Xw6mt1o+xIOIiJLDw/OJPHx87PVSJUC5Uu6fICKZXJbL54slUplcbmZqaqiq\nAACMCLfxmO7T/OY0MZNHH/9m6sxfgmp+sm5uEyIiSjx+/B6R59ixnfVRI0B5Ylm22D0dz6Oi\nWJa5E/JCKpcZqioAAOPCqX+CyLbv1gdP+xw89SzHsdmgSePaOxARkSBa3vzjTzr0/KwLTw8l\nApSrkhNO3A5+PrZnr8tBDwd37GxlbsHj4fcYAKAUHPMEETn6jp7jO7rIJpvuX+3tzvnEABVD\n1YQTBUJhQnpadn6ejaWluRn3fyYAAFUcp+sdId+1MTMza/NdiE67ACoVedHBEwqp2dliqTSH\nz5fK5RVfEgCA0eGUJ1hGJpfLZYyKj3eKXXJMDQSVnsoJthMz0nP5fJlMJsJdHgAAWiiv9UX5\nfD4RWVpaltP5AfRFZZ4QSSX5QmFiZoZiHXMAANCsLBeGZQWZGQVSIsoskCkepqSkvNvNSguS\n7u89F0tk1qCBp57qBCgvobExJTcKRCK5XC6R4f4OAACtlCVPhG3q1XL1u4ERr37q7faTqnYu\n46cOtiprYQAVJC4tteTGp68jiUgskWAwJgCANsryt9LavUX79lZEJEoKCUkWWbn5+ror5Qae\nqYVdrfqtek+Z/8VQR33VCVBeJFJpyY1iiZSIhBIJ5rMCANAGj8vqzMF+LVquDvH99kWwXws9\n1gRQkb7cse1MoOqVcBt7eI7v3ScyMXHDp7MquCoAAOPCqS/Xe/a/T0YLrV299VUNQMVj1a9Q\nHpGYIBJL8oWCiqwHAMAYccoTVq5N27jqqxIAw2BU3fBcSCqXv4iKqrBiAACMFPexZsKoy7/v\nO3b9SVRKrlBa7C9zo08PHfy0IeenAChHGvoniKhAKBBJJBVWDACAkeKYJ3L8F/YcufmFUM3u\ngsHoKIZKTSaXp+XkaGhw4f49oQRTWgEAlIJbnnj205zNL4Tk/N5XP66Z0s27jm2xofDmDrgc\nApWaQCx6HvVaQ4McfoGJSXlN+wYAUGVwyhOJAQERRGZ9157aPKO2vioCqEByhpFqnLRKcbFD\nKBFbW2CyVwAAtTh98BKJRERUr0MHhAkwUrkFBdrcMo0hFAAAmnHKE57NmtkRpaen66sagAqW\nkZenTbMnERHlXQkAgFHjlCcsB8+d1cQ0//zh81r9TQaodF4nJWrTjGFVrGkOAACFOOUJhlot\nPbh5iNmR6UMW/n37ZVJ2QTFCCf4KQ6WWnZ+vTTOVa5ACAEAhTnkidF3HWh3nX0ySZdzd/FGP\n5h7O9sV0XBeqr0IByoPmm0ULvYqPK+9KAACMGqf7OyxrNWzaVNPY+Ia1MCQeKq+M3NzY1BRt\nWt4NCf5i9AflXQ8AgPHilCcazzsTNk9flQBUNKlMJpapWFy0JFzvAADQDBP1QPUlkcnkcrk2\nLZEnAAA0Q56A6isrPy+Xz9empUSqVTcGAEC1xel6R8LxxYuOx2toUHfsph/HenJ5CoDy8yDs\n5cu4WG1aap5DEwAAOOWJnOCLR4+GaGjg67MCeQIqp3yBIDxBUxpWpnkNUgAA4JQnGs7461b/\not3F0ry4kKv7vt9602TQul1LB7TBYuVQSb2Kjztx+6ahqwAAqCI45Qmbem271yuxtc/QicMb\n9mk2b+3mwWOvdedyfoDyc/j6Ve0bYzgmAIBm5TIe06T+rC9GWeYHfP/zDa0GzwNUvOiUZO0b\na7NmGABAdVZO93eYeXl5EKUGBSWUz/kBuNIxIiBPAABoUk55QhQVpcOHP4CKhyklAAD0qBzy\nhDTr6V+zV50UEtXp2LHk8AqASkGn/glkDwAAzTiNxwzb2KP7xpdFt8kFOTlCORGRy7ANi3ry\nuJwfoPzomCcQKAAANOGUJxiJoKCgoOg2nrl9nUaN2vWb8OXyrwahdwIqLZ2mlECaAADQjFOe\naL4iSLRCX5UAVKgCodDQJQAAVB1YvwOqKbFUhym0kzIycMkDAEADTv0TyuT5yTHx6Xyydalb\n383eVF+nBSgPDMumZGVq314oEcsZxswUv9gAAKpx759gMwJ3zuzTqIaTu7dv69a+3u5ONRr1\nnbX7fpYeqgMoHwzD6HqITLuVzQEAqieueSL97Gedes7dGxBVwFjV9PLx8appxRREXf/18x6d\nZl/I0EuJAHonFIt1PaQMEQQAoPrglieEl7/57LdomYnXiI3XE/IyYl6+jMnISwjYOMKLJ329\n55Ml/hjxBpWSWCrV9RDMfwUAoAGnPCH1P3gslajBnINHF/d2N1dsNHfvtT/z5yQAACAASURB\nVPjowbkNiFKO/n1VhyFvABXm53+P6XqIHP0TAADqccoTCWFhfCLnASO7WRfdYd1tRH9nooKw\nMKzfAZVLZFIiEYXGxuh64NqDB/RfDQBAVcEpTyiuKFtaWpbcZWlpQbjkDJVPZGIilek382lk\nRDmUAwBQRXDKEx4NGlgQJd8MCC/2x5kJv3ErhciyYUN3LucH0DuJTJrL50vlOl+Ji09Pe/Y6\nsjxKAgCoAjjlCasBY4bYET3bMGXxhbjC8fLiuAuLp6x/RuQwdEx/K+4lAugRXyhsM2tGWFyc\nrgeKpdI8gaA8SgqLi7tw/x7uRwUAo8bt/g6n8Rt/6O9Iwoc/D/Op37LH0DFjhvRoWd9n2M8P\nReQ0cOPGsTX0VCaAnnC5TaOchmQevxUwd/tmvkhUHicHAKgYHOfHNGky58xdh2/mLNt7Iz74\n9sVgxVarur1m/bB7wyRvzOYNlUpKVub3B/8q8+HlNB7oZWxs+Z0cAKBicJ9v27r5lG0Bk9cn\nvAh6EZ0u4Nm41G/ZvqWnLaIEVD4yOSOU6DyT1bvDGf1fkmBZ9m5oMBElZKQ72dvr/fwAABVD\nT+t3mNh6tu7p2Vo/JwMoJzqtUV5SdHLyi+io5vW8TPW3kEfhsImDV69s+HSWvk4LAFDBytCN\nwNxe079Ll+6LL/LVteBfXNy9S5f+a+6gAxcqFYbhlCc2HDk4cuXS7IICvRQjEIsPX/OPSHwz\nR0tCRvrIlUsz83L1cnIAgAqme55gbv694+r9R5ZNOtiqa2LbobHFo/tXdx5CoIDK4+bzZ8St\nf0IhIzf30sMH3M/zw5FDy/bvfRD2UvFQLJG8iI4qw0TgAACVge554sG5c6lEXUaMcFHfxmXE\niM5EKWfPPuJQGoBenb13l9XHGhzJWZmHrl2RcH7jj0pOIqUxGYobTzj2oAAAGIrOeUISHZ1E\nZOnl5aqplZuXlyVRQnQ0PmxBZSGWSvTyXs0XCUUSSQ6f61WPPAGfiGJTUxUPFTejMiz69ADA\nKOmcJ1KSk1kiN/dSJr50d3clYpOTU8taGICeSaQyvfRP3A0NefgqjPuFCcUNogVCofLDhbt+\niU1N4XhmAICKp3v/hERCRCYmpRyoaCAWl/3ePADF2l16ERITcz8s9POtP3E/1ZWgh0Q0e8tP\nQol48a+7ynyexIwMIkrJylQ8VPRMBEWE5/LVjnQGAKi0dL5f1MXFhYiSYmKk1NFcbStJTExy\nYWMADe69DF13qPgcU39+s8zJzj4hPd3b3UPXE0plssJBjkRkaWHRoUnT2NSUnIKCHH3cmiGX\nM0T0Kj7uaWRkeEJ8mc8jEIuI6N7LUMXD+LR0xRcLd+84t3aDlYUF50oBACqOznnCsUMHb7oU\nefWcv3DcEGs1jQRXzl4VETXp2NGBY31Q5WXl5b2Ijiq2MeDpE/8nQRZmZr1btyl5yNGAaxN6\n91V3wlw+f8qGtYUPPV1c/DduLsMCYOooTiVnmINXr2Tn51+4f4+IPGrVat3IW6fzFLv0ohhO\nQUSvkxKlMhnyBAAYF93v72g1dlxToszDy75/qGa9AcG9NcuOZBHPZ9zYFhzLg6osMy+318L5\nq/7cX3JXVHLyhfv3wuJVr9r1OjlJw2mLDWmUyRm+UKjH1bZEEoniC6lMFp+eNnf75rnbN++/\ndCEpM0On82gYzJGVn8epRACACqd7nuC1XrB2ggtJn34/qN/Cw8HZRf52M1nPD37Vd8iG5zKq\nPXndgpY8vRUKVY9ULo9LS1U5g1NGXi4RKd5wd509te/iuSIHylR0NgjE4rnbNz98FVZsxa+s\n/LzZW3/ac+6MvsoujCaBL0MKN8ampeo67kFDnijZYaNoP33jep2eAgCgwpRlvm2XsXuPL40c\ntj7o7ubJLXd95duxU/N6NW1IkBEX+vBhSJqIiBw6rfx39/s19V0tVCms+rkWjly/SkQsyxDR\npYcPrC0tPx0yXLErOiU5Lk3FfUNiieTC/Xv92rYvNlhYIpU+fBWmz7rfyldavjw4Jjoo/FUt\nR0cXR20X1dVwq4nKhUwZhgmJjdatRACAilK29Tvse667db/FivnLdl2NTQu5dS5EaZ91g/5z\n1m9fO8HHSj8VQpVV6lwLLEv7L13IFwgszN+N/X0ZG3svNORM4J2R73VTbvwwPIyIGIa5/Ohh\neVSrmVwuX/Xnfnsbm1Fdu2t7jPr+CZVdFwzLiqXSO8EviMjTxcWrjsYpYAAAKlaZ1wOzbj75\nJ//xKyICr1+/HxKXmi3g2TjVrufbpU/fLo1r6GmVMajaXpV2cwTLsj8dP2pmYupsb58n4DvY\n2BJRdkG+QCyOSy3eRZGRm0tET19HJmdmllPBmrEsK5ZKC4RCO+siA5V/OX3C/3HQqdXfl2yv\n7lRRycklNzIsmy8QKIaazho+csnED/VRNQCAfnB75zdzatxjTOMeY/RUDFQv4rcDG9XJzMsT\nS6USkmbk5Xb7cu6LvX8QUWp2Nqm6IqCYD+rg1SvlUqt20nKygyJe9WpV5J6UrLx8lVNUabje\ncTTg2sKx44ttDAx91w/I6GNiLgAAPUJPApTd/ZehB/z/U97iWav20knvPjdP/eH7ku98SyZ+\n2KJ+A9LiTTG7IP/NF/kFijGYO8+cUkwAVbgs54X795rUrevt7lEZ3mCjkpN9vRoU28iwTC6f\n32bWDMXDYyu/s7O2+uX0SQ39EyqvBKXlZL9roGqAhUrBMdGK7zYAQLlCnqj6jgZcG9uzt2lp\nU5qWQXx6mmL2hUI+9eop54lbL56XPCorL49l2eCY6BitJ5aWM4ycYZ5HvWZYRjFBdWhsjGLX\no/BXNlZW3u4ebCVY+eLk7ZvDOnVR3pJdkJ+em8OybOHdH2KpRCyVHL7mr+E8it6X5MzMWVs2\nbZo1t4mnJxW9PpKYkf7sdaQ2M16k5WQTIU8AQLlDnqj6lv++b0SXrjZWqgfI/vTP0cy3sx18\nPGhIYw9P7c9c8vO18vKYKm9SICKGZSUy2ciVS3V4IpZhWTY6JZll2YsP7ys/tZxhdpw5+c/N\n65Vk2YsCkVD54ZWgR8Ui15+XL33Uf6Dmk/BFolN3brVu5P0iOoovKrLAh8Klhw8SMzLOrCn9\n9tHv/vrTzsq6k08zbV8AERGJJBJLc3MeDzd8A4C2kCeqPpZh9l44Z2leZHr0qQMH21haEtGZ\nwDuFt18O6tBRpzxRstdd8RlaJpfzRSKZmikpGYbRdV0uiUzGsCzLsoV5pbA3gmWZR+VzO2jZ\nfL17R5dmvpFvL8e8KjElF8OypV7okUilgaEhLRs0JKInkRECkYiIwt+e8+15Su+PEYhEsakp\nWfn52tevsPHo4UXjJqjLoAAAJSFPGL1LDx/cfPGs8GFTz7rTBg4ufMiyLMOyW078U+yo97v3\nVOQJ5bclRv2EECqVjAX5QiER3Xj29NOfN6o7SiAWsZoGI6qgmEIqJjVl99nTb5+aFu7ece1J\nkLC0QZ0VTM4w154ELdu/V12DyMSEjzYUv9ejpDwBX7Fg2Jq//1T3RKWeRFHGyj/2rT/8t/J2\nbw+P377+RsOBcoZJyMiwtrSwt7apYWdX6hMBACBPGL2nryOUL8a3aeQ9smu39JycoPBXpP4m\ngsKuBeUMERwT5eZc06dePc3P+Dgi/MrjR6Q0iKGQSCKm0gZaypky3p0gkkgKVwlnWCZPwK+c\nS3FqXitEKBbzi14TUenSwweXHj7QfJ4X0VHN6nmZmZqqa6MIixm5uURFJiG1La3jgWGZmZs3\npmRlzRg87H8TJpVaLQAA8oTRK/beLJHJGIa5HfxC3edaBZH0zcd65a6Cn48fexEd9euCxZqf\n8VlUZGE/QTGKdKL5ozPLshqmctJAec5svlCUkJ5ehpNUgN0a5/bW1yLssampI1cuDdy+y9XJ\nuXDjoWv+y992jcwbPUZdhxPDsjeeP41KTnoSGaHyasjrxEQzM1OGZbW5qgJQNYgkkojEBDNT\n02b1vAxdi1HSLU+I0yIj0tSsAqaKVe3G3rUtdSwJOJHJZdq8DbyKj0/KyHgRHaW4XaJQqV0H\n3/3159PXEer2Kp73fliohjMUCIUCsVjzs5QquyC/8G7SyqYiJ9QqNme58ogWvlBU2J1TDMMw\niRkZEYmJ91++VL4NVZl7zVpMWXuSKoM1f//Z2MNzYp9+hi6k4uTy+bn8AiJysLEVS6ViqcTB\nxhaXq7QXm5oycuVSJzv7x7v3GboWo6RbnojYObrl6pDS273l++2LYD+sMVqhUrKyGYbVsDSG\nAssyl4MeHrjyX/HtpR147MZ1Dd31iref4zcDNJxh5R+/PYlUm0hAJ8WCo3LP0O//XVB3VFxa\n6o7TJ3u0bKWhJyk9N0fOMLqOdFHILsi3srCwtij7Z4n49LS52zYXPpw+aMiY7j11OkN0SrKd\ntU2ZCzBGv108v/3Uv0T02dDhz6Oj7r8MnTNy9OLxuFylLalcTtqNTAKVdMsT9t7dBw0qNv5f\nnvzk2vM0xtTOrbGPtwulRYZFJhfITWq36tvWrZG3vR5rBZWK/cXPE/BZlin1n4RiRoeS20vt\n2NA8k5JEJt199rSotDGSuYLKOO7BGO08c8rB1nbOiNESmfT+y5fBMVotGCaWSpMyM/IEAnX3\n4NDbRVxLzZeqqzp9qkPTpoM6dCrDsQpCsUR5kdXU7CxdzyBnmOp2saZwfDTLvvl3arzdS+rs\nv3QhPTfn8xGjFLPv68X/ft3FF4uISDEeSyAWzd2+WbnBvFFjcAVEG7rlCa8puy9NUd4gfbF1\nSN/L5q1n/nF408RmivCQH3pwwcQZ+5N4ww7+MgU/AwOY/uOG3IICzW3Sc3PyVL2ph8TEKP4t\njXyvm8r3A81/o6Uy2Q9HD5Va4dXHQaW2AW0cuuZPRFP7D4pMSiz2R7BUFx/cK7VN2fon5Ayj\n/QyeKhU7vNTrLnKGScwoMp5GIBaXLQxxsfHo4di0lK1z5msYJFt+Cv9tyt/ekl31EtXha/6R\nSYlT+g3QY5649OiB8lrBMrm82Jwx43r2QZ7QBrfxmLE7Z359NcN35Y2dE5sV/vOxb/7h7kMv\n77X8fuHs3e/fnqfDdAagHyXnPCjpu79Uj9ZMz81R/FtqVs+rWJ6ISk7652aA4tZNqFQYlsl6\nOymZfqXn5Cj3E5Tk6uxccol2db1f2hBLpeEJ8dEpRVZES87MjElJru/qpu6orPy8XgvnF9tY\nv46r5uL1xatOHcXb2+3g5y+iozbNmmOgPMG+/eLNyJfCLesO/XXsxnUiGtO914nbNxQbN3/+\nRZ82bSu+TiI6fM2/V+s27jVradn+aMC1Z1GviSg1R/XyPWUTHBPNsqyiK06DqOSkBq5uXnXq\n6OVJqzBOeSLz0rl7cnLt3a95sX87Zi0G9HH9PuTO2YtZ8z5zVn0wVHIlPtrFpqaqu60DDEvO\nMClZOl8R0MaZwDtnAu9oaPDNhMmzR4wqfBgUEb5w1y9Z+Xnn7wcO6djZVPe3VcV9K8U2Hrx6\nJTY19a8ly9UdpfJOluM3AzQP5dGXXxcsHtC+A73tD6j4fhEiyuXzC68ziiQSxXukSCxW9OFn\n5OUpvsjhFxTeZZ2Vn6d8x7WFuVnJIS9iqVTl5UtTE5Ni6+jq5OSdW03q1lOZJwqEQhMer9hc\nandCXpwNvFv4kGPvV6H3v12uzQekNX//ee9lSKk3vgGnPJGVlUVEUlVjyCUSCRFlZGQSIU+U\ns/K5Rlqyp1TXSS2hwhy86h8WH2uQp5YX/T0RikWK6VYLhEIZw5TME3/8d1EkkUzo3dfJXvXg\nKnVvFZrXZzHs6i3yorO5lO0iEUdtZ39S+C/0aMA1xRd/+V/+y/+ycrOwuHe/J4v27FTeNaX/\nwDXTPyl22v2Xzm88erjk03m7e1zZ+HOZq1VMdqty1werV3jUctm/aInyxmIdEoeuX3W2U/37\nM6X/QO2DjvZ/04z4TqcKxClPuNWrZ06PM6+cuSvs31X5Jyi4ffpKFpG5l5c7x/rAUEr++6l6\n12KrjF/Pa5rxolwVe/tX/ruvslN6y4l/cvn8Xq3bqMsT6rqyNXdxG/bP/bvZ4RRXGSq8f0Iu\nl2v51vgyTm3uVPlJXS5X/W2XcRwfw7LqfqAyOVOykmJdPnvPn1V35qGdu2ifJ7T/tdFXj0jV\nxilP2A2f+r7zqWNROyaPd9uzdf6AhrYmxPBfX97yxcyd0UQ1x0wdprchM6BO7zZtne0d9H7a\nDk2aFtvi7e7xzYTJen8iMGodmvooP6xfx63wl8Rc1cWOL0Z/IJXJajk6qjthbacaKn/N3GrW\n1FCGvbWNAX85m9Stq/ji40FDsvPzLYqulVMReDzuL99H1ZDDLs19VZ7ZwZbT3/YP+w2oW7u2\nyl2fDhlmb1P8Rt9R3bq3athImzNrP06TZdn/aX0zbb3aGDxROh7HTuy0i/N6j9nxUkREPEtn\nV2c2KyVbzBKRVbO5JwN+Gaz6NwYAAACqEq55gogKQo9vXLv10MX7r3OkRGTu1Kjz4Mlfrvxm\nbDN0TgAAAFQLesgTb7AyYW6ugGfj6GBtxtPPKQEAAMAo6G89MB7xzC1tzC0RJgAAAKobE85n\nkCdc3Ti9V5OalhbWdvYd14USkeD21s8+/fTLAzqs9AEAAADGi2v/RPyhSZ0/+ieZsbC3M6e3\n057YNDQL++O325ftJn20pQu6KwAAAKo6bv0T+ScXzfknmfWeeTH+zteN3213Hzv2PaL448fv\ncywPAAAAjACnPCH+7+jpXDLrv3jd4NomPOWOiDre3nZEiU+epKs9GAAAAKoKTnki4fVrMZFH\nq1YlJ5pxdXWltxNyAwAAQNXGKU/wlPokeEX6JygtLY2I7OzsuJwfAAAAjAKnPOHRqJEVUXJE\nBL9YnmBDA+/lETk2a6Z2fWEAAACoMjjlCcv+owfbkeTyvt+jlddKkb7+bcnOUKJa73/Qk/v9\nqAAAAFDZcZwfUx76U+9Oi25L6/UdWS/0+O0U90Hzxjve/eufx5ms46Bfn176rL6+CgUAAIBK\ni/t821m310+dvPp8vFhpm2XdYWv+ObS4s/5XvQQAAIDKRz/rd0iSH50/6//oVXI+Y+3SqF3/\n0SPf87TifloAAAAwCvpbDwwAAACqK07jJRNPrZg+fc6W29lqdk1fcSqRy/kBAADAKHDKE9lP\nT/35564Fg/p87V98Hszsp6f+/PPPU09VRA0AAACoYvRwP6dznbzfhvecdSaRKb0tAAAAVEF6\nyBNuUw9fXevx79geU49Ey7ifDgAAAIyNXuabsm2/6PyNbS2uftRj/G9hEn2cEQAAAIyIvuav\ntPSd/e+t/d2DZvccufWZUE8nBQAAAKNgpr9TmXt/dOiW7Wf9J/UZXHBxg/7OCwAAAJWcHvME\nEZnUG7Pv5inbAR/0n1DLXK9nBgAAgMpL7+t18VyHbLtx8Ys6ObhTFAAAoLrglCfqjt10+PDh\nTWPrFtvu3GvdVf9dC2bNmtDBmcv5AcDYXfqigUfn756pb/BsXVcPj5nnK64iACgXnK53OLYY\nPLGF6l0OnWb/3InLuQGgsmNyQk7s2X3kv8DQuAyBib1rw9Y9R0ybO2OAl7WhKwOACqf36x0A\nUC3IIo98MmDokovMoBV/BzyPCA+6tHtum/j9n/QfvuxamqGLA4AKp1v/RNKZb1edSfQY+d3q\nke6FDzW0L2wJAFWK7OmGGd9cZkf9eXJbfyfFJo82Ixb91bLWxIHLP5/nc/XYVE/DVggAFUu3\nPJH1+N/ffgvx9fxKkRIUDzW0L2wJAFVJ7ult+19T9w0r3oaJN8zqT1sxbe+wnZv3PpiyupPK\n7s+CFwfXrt59/nGC2NG729QVP/Qp0SD0+Jaff79wLyyxwMSpfvtBM5YsndqmhmJf/O7hXdZY\nbXj8U409yzefePA6u9+e2F2Dy+EFAoCudMsTDabuu9K9wK5hA+WHGtoXtgSAKkQeGHBLTC36\n9a1dYhevTd8+zjt/DwgIo07NS+xlovZP/2BVeNtFW6981NG54MXRNYt3iOVKDQQP1o+ZuKeg\n//L1p0e3r8PG+m//+n/vj0s/df7r1hZv2/AD1iyv1fPrAzfq5Z+9pekPEABUIN3yhG3DLv0b\nqn0IANVDdmKigEw8PFR2Pnp6ehKFJyYQlcgTkoCtmwJp4LZfv+rrSEQ1u87eJpzfaSpR6zcN\nIvct3RFS/4v/dn7WwoyIyHf0d7viHvbcvPH4jIOT33aFPI/yurJvfHMeEU0ZXR4vDwDKAOMx\nAUBXLMuWspNHPBW7Xty6lUudBw50LNxi03dQz3dz3yVcvRLG1h88osW7Dzq8Bj26eUgC7z1+\ndxbPAQObqzo7ABiSfufHBIDqwNnT05oeJCYmEXmU2KnY7FFyB1F2djZZ1nKxV9rEc3GpWfgg\nPS2dKH774Hq/KGLJ2/+IGmcLid7churq6qbXVwMA+qBbnojYNnTItnDt2zeZf/HC/MY6lgQA\nlZxpl17dLU9cv3otbeZHxYZQsE+vXc+k2u/39lFxnJOTE4kz0vOJCiMFm56eWbjf2dmZyGHJ\nrctzvTQ8u7k5PgcBVD66/bsUZ8W9fv1a+/ZWWWId6wEAI1Bj1PzpW67t2fr91eFb+ynd4iGL\n/XPtnzG2733/qcqbO1p07+7w639XruRNHOOg2CK8fvmWtHC/V78BjddvvnA6Yvb8xqbl+woA\nQM90Gz/RYtUzqS6erVIzfSYAGDXzdkv3r+vHnpz1/ty918LSBVJpftKzcz9N/WDVU88pO3dO\nLz4Jv4Jlny+/7sL8993sbdejsgpyYgP3fnlA2EZpogqfmRvnNQv7edrsHZdDErNFotzE8KAL\ne775YOkZSQW9MAAoIx37DU1MzTCCEwCIzJtM+cO/9fE9u4+umfxTXKaIZ1enYZue0/Ze+WRQ\nAxt1B5k0+vSP45Zr1+ye1/9niWOjrpNXbJwT+P75a4UNbDstOXWp5Y5tv6+fsj0mS2br5tW4\nTd9x/5s/2ELdKQGgcuBpHKgNAAAAUDq9jGsSJjwKuPk4MiVHIGGKpBOX7p9+0r2WPp4CAAAA\nKi+ueUIed2bB+E923s+Qq9rr++1w5AkAAIAqj9toCDZs45jx2+9n1Oz66az+bkTkNvjrb+eN\naGpDpm4DFm/evnwoFu8AAACo+rjlibt7tgeJqc6Mv/33zuvmTETOnaf7bT/z8MrXjVP8DwV5\nDurkrJ8yAQAAoBLjlCeSgoKSiRyGftD/zbx1b6ays++6YuEAXuLfS7YFc68QAAAAKjtOeYLP\n5xNRHVdXHpG5uTkRicWKCaxqtGjhSfTq1q10PdQIAAAAlRunPKGYoj8jI4OIatWqRUSpKSmK\nXQzDEFFeXh7XAgEAAKDS45QnbFq2bEiU/epVOlHNtm3rEhXcuHRbRCQPueSfQGTaoIHqWfIA\nAACgKuE2HrPthx82J7p94EA0UcdpM1qaUsyu0R0HDuncf30wkcuEjzCpHQAAQDXAdX7MlMBD\n50IE7t2mDG1mJQn7Y9r7XxwJKyAi01qdF/59euOgOnqqEwAAACovfc+3zfKTQl8miB0a+DZx\nsdTniQEAAKDSwvodAAAAwJVe1u8gYiV5KfGJWUJ50XBiVbuxd210UwAAAFRxnNfvSL75y7fL\nf/zrbqKIKbnX99sXwX4tOD4FAAAAVHLc8oTw1sKe/bZFysjc2budb30XG9Oi++t723M6PwAA\nABgDTnmCf3Lz7kgZOQ/bHXhsVhMbfdUEYGgsEc/QNQBULUzwke+OhxHPzMbZq/3QD/o1Ks/3\nDGGk/7ELzwWuPd4f29G1xLwI6vemBfyyMyBD8XWT0Ssmt3n7HikN+WfTP/l9Fs3oYleOZRs3\nTvNPJMbESIhqjV+AMAFViSAz1tAlAFQ5TPCRtSfCiZVkvtj/cfsRu2LK76nCtg3qOudYeHzQ\n1jGdZl3I1WFvWsAvBx6JSpyQf2fl8h2nf9l/r6D8ijZ+nPonatepQ0SWlhhxCVVKdlSgTU0v\ndFEA6JlZq8l+fsOJshrFuN14Qp/XJwo95ncslGdu79Zi4PiRLWvwqODB/r9FfXrm+//3yqzN\nB1P6eJkTkTDyv4PnXvJ8hvYwu/C49vyJrUyIhNEBx88/SjXx6jPug/YuJiR49MeWhO7LRnvT\nvd0bc768enN5UxIN/KTejwe/Hzqn9rsiNO8lsnFwtLZzathj9PtvOyeyL/1vi+Oa77oPWVZx\n3ypjxKl/osbQUb3MKSkwEJ/moCqR5KfLxHxDVwFQ5cieH/LzW/XNzA9+jJ02sce77awo5cHW\nUX1XPZMTFTzYv+LT6UtPRSSH7Z/c65u7MqLobYPf++JkeNLLIws+mrnqyHOGSPzIr1ufJZdf\nZ2YE/zG52ycXlXsZBOHheR06NiUisurUyffVq9ek9d7aveeNbSLKS7z506iWw3+NZYko5dDi\nwx03LWmJ2Z5Lw208ptunu7ec7DZ33cdrux5b1q0Wt8m7ASoJlpHLRPlmlrhOCqB3jCAjMsmy\n28JOtYiIqHGPoS2zb4en1n2v5Y1T5yLXtHYksh68/uS2riYU59p1/N0Ecjv+Z9qsEzfWtiAS\nnZri8gcR0fXd2wTtZjdytCDq2Nlh66GAfUNGdZi+rAMREcuyhX2LihmW3gzdIIs2k5cNKLFX\nWe3e8/x6ExH5jf+yyexDkTNn3JvlF1V38u+r/RLu5sU5bPq1w6aZ3cv1G2S8ON4vauYz59TF\n5O49V/bw+qWJb1NPB/MiuxtM3bd3an1uTwFQ0WSi/PykECtHN0MXAlC1mLWa7PfdcGI/39Kz\n75Lz/X4fZnX7f90/DhkwqmNtO5Zls7OziRyJataubUJEZGNjI8ooPpiBJSJiGYZMTd98hG04\n/MsmXnKit/cX2jZtav/Tw1c0sCmJHj4MbdKuEdG7TnRVe1VTPJPHgMk9M/T2HajSOOaJvNsr\nBw5d+1hMRKmvHqa+KrbbtzsGr4DxkYpyJQXphq4CoKriNZqzbnjTXSl4mgAAIABJREFUL7eH\nD/1awBeZmJubsgWxkUliaqKyeYOx02p3GjMs6/1mguCbd0zrDjMhXt/P51tPuBTh1qeRs6UJ\n1fb0tFAaP9Fl1tcOPUcMSxnu9vjY5RG7f6ptUmOi0kxIxfcS5d3bvzuj9/+GN8y4ufuXayms\nNDfi6hF/n9UPvcm18Ty/vkREdK9gX5jnInROqMcpT7DPNn2y9n4+1ez65ffLPuhYcv4Jy1oN\nuZwfwCBYhpFLhIauAqBqMWkxcYXFm8Rg0WP53un/xqeYDdx8bfPBs6GyuqPnfDTkWLonkV2n\nGbNFzkREZNNh6mxBLaJa8y/ebXLoXBjPZ2zjvAkhHiZEVh397tzof+LcvddpBbISi0Y0W3D5\nbotjF54Luv+7ZmznGjrtJSIT27p9/nd266gOLspjsj0HLpzhgIugGnBavyNxaw/Pr25bDN6X\nfPETZz0WBWBQQfsm1fLp59X9U0MXAgBEREn+2369nSnOennxaMiIc8/WdNTTShGgV5x+KjKZ\njIg8W7VCmICqRJKfxjIyQ1cBAG+495/v15+IaP02Q5cC6nG6JcOzffvaRMnx8XJ9lQNgaBJ+\nFiOTsHLkCQAAHXDKE6a9FqzoZS88tW13tIrFwACMUVbkbakwF/0TAAA64XS9Iz8iscFnCweE\nr5nfqduzBdN6Nyt+v6hDk149m2BJMDAmwux4ImIZdLoBAOiAU56IPTRnxOoQIiK6t3f5vb0l\nGmC9cjA6+YkviEgmxq3OAAA64JQnnDtMmDUrUUMDjw4YqQlGpiAljIgkBZjBBgBAB5zyhPvw\nlbuH66sSAMPjp0UqZp5g5VJD1wIAYEy4rVd+asW0aVOXnUzQVzUAhpWfHCKXCokI4zEBAHTC\nKU/wQy8cOPDXmWc5+qoGwLBY9s2dSiyDW5YAAHTAKU+4eXqaEKWkpOirGgADeztdbGGwAAAA\nbXDKE/ZDxgywpsyAK09wrRmqhsLp51mWkUuRKgAAtMQpT5DLtF17J3hFbZ7y6e9Pc8q+DgiA\nweXGP018eJilwl9jVi7my0T5hqwJAMB4cLq/4/Wvkyb8GmnuIA89MKPtofmuDRp6FJ3Qynvm\n0SMz1S4uD1B5xN/9XSbKNzGzUDyU8LMKUl/ZYIFcAADtcMoTwqTgoKCQNw9kBSkRz4uNpBAl\nYdFnMA5SYW5u3OPChzJRnkyYh7s8AAC0xClP+Cy6kTxb09AJc7taXM4PUHGKDpWQCnJEuYlW\nTp5Wjm6GqggAwIhwyhNmdjVd7fRVCYDBsIy8ZFcEP+21g2cbg9QDAGB0OOUJZfL85Jj4dD7Z\nutSt72Zvqq/TAlSA/KQQqSC32EZRTgIrlxikHgAAo8Pt/g4iIjYjcOfMPo1qOLl7+7Zu7evt\n7lSjUd9Zu+9n6aE6gAoRc3O3YllRZYLMWAkfv8YAAFrhmifSz37WqefcvQFRBYxVTS8fH6+a\nVkxB1PVfP+/RafYFrKgERoCRS3JjH5XczjIyjMcEANAStzwhvPzNZ79Fy0y8Rmy8npCXEfPy\nZUxGXkLAxhFePOnrPZ8s8cftHVDpibITVM6uzTIMK0eeAADQCqc8IfU/eCyVqMGcg0cX93Z/\nM/OEuXuvxUcPzm1AlHL076v4cwyVnSAjhlG1migjE+fGPxVmY7k7AIDSccoTCWFhfCLnASO7\nWRfdYd1tRH9nooKwMPwthspOJs4nUjG7KyMTJz85kRV5q+JLAgAwOpzyBMMwRGRpaVlyl6Wl\nRWEDgMpM80UNlpFXWCUAAMaLU57waNDAgij5ZkB4sdjAhN+4lUJk2bChO5fzA1QAcX6ahr1Y\nEgwAQBuc8oTVgDFD7IiebZiy+EKc+O1WcdyFxVPWPyNyGDqmvxX3EgHKlYSfqWk3i4XuAABK\nx20+K6fxG3/4LWCu/8Ofh/kcatG+bWMXNj3iaVBwipDIaeDGjWNr6KlMgPKj+aoc8gQAgBY4\nzo9p0mTOmbsO38xZtvdGfPDti8GKrVZ1e836YfeGSd7cZ8sCKG+ar2hgCgoAAG1wn2/buvmU\nbQGT1ye8CHoRnS7g2bjUb9m+pactogQYCc15Qi7FJCoAAKXT0/odJraerXt6ttbPyQAqlMY7\nOFROdQUAAMWgGwGqO1bjCAmpILvCKgEAMF4690+8+rl/v5/DtGzss/Cq/8Kmuj4FQEXSPEIC\n94sCAGhD5zwhzUtJTEzUsnGNPBXTGANUKqLcZE27cX8HAIAWdM4TPguvxn+qKSXIUu78unr5\nlnOvhUSmpqYcagOoCLiDAwCAO53zhJlDHU8HNfvyw47/uHTZz6ci+GTq0vGT5T+s/rwZx/oA\nyl0pVzTQPwEAUDo93d8hTQrY7ffNmv0P0uVk23jkirUb/je+mb1+zg1QrjSPxwQAAG1wzhNs\nXvCxH5Ys23I+SkBmdbrOWrnRb2Y3V3N91AZQITT3TyBtAABogUuekMRf27Xqf2sPBGUwZNds\njN+69V+PbmKnt9IAKgTu4AAA4K5seYLNfnZ4/ZLl2y/FiMjco+fcbzeumtG5NsZegjES5Wi6\nvwO9EwAA2tA5T0jir25bvnjdwSfZDM+xxfhV69d9NbyRdXmUBlD+GLlULuFraoHrHQAAWtA5\nT4T/9uXiv0KIrBsNnb94Woeaoifnjz9R19ix+cABzdXdDQJgeHKxxjBBhB4KAABtlHn8hPD1\nhR9mXyilke+3L4L9WpT1KQDKHSMTl9IC/RMAAFrQOU84+PQbNcpby8ZePuicgEpNmBWruQGL\n/gkAAC3onCfqTdx6amJ5VAJgAAUpr0ppgf4JAAAtYH1RqNbkUpHmBvz0yIqpBADAqCFPQPVW\n2uQTMmFuxRQCAGDUkCegWit1eIRcIqyYSgAAjBryBFRvpQ2PYOSaVtMFAAAF5AkAjTAeEwBA\nC8gTUK2xjKyUBqy8YioBADBqyBNQrYnz0zU3YBl5qfeAAAAA8gRUa6XPj0kkFWRXQCUAAEYN\neQKqNVaL4ZbatAEAqOaQJ6Bak/AzS23DYkgmAEBpkCegWpMUlJ4nSp3zCgAAkCegWpOJ8g1d\nAgBAVYA8AVAKFv0TAAClQZ6A6g1jIwAA9AF5Aqo55AkAAD1AngAohTAz1tAlAABUdsgTUK1p\ncy9oVtRdASIFAIBGyBNQzZWeJ/ipEXnxTyqgFAAA44U8AdWbFv0TopzEgrSICqgFAMB4IU8A\nlIJl5DJBjqGrAACo1JAnoFrTZvyEhJ8lyk2ugGIAAIwX8gRUZywjFWrTjGXl5V4LAIAxQ56A\n6oufHqXl3JeMtPRlzQEAqjPkCai+hFlxWrYU5SSWayUAAMYOeQKqr9QX57VsyTK43gEAoAny\nBFRfrFyqbUssCQYAoBHyBFRf2qcEVi7FSh8AABogTwCUjpFL5RiSCQCgHvIEVGO6LFaeE32/\n/AoBADB2yBNQfek0KuL/7J13YBTV9sfPlmx67733AiQkpBBKOlUUQVFBeSrWh/oTRUUFfTZE\neCBPFEUFKYI0pUPokAQIhPRGCqRvyqbsZvvO/P64OAybQsJGEXI+fyVn79y9c+bu7pl7z5wv\nplAgCIL0A8YTyHBmEOsT+IgHgiBIP2A8gSAD4vrpbyRNpfd6FAiCIP9QMJ5AhjGDyZ9Qy8Uq\nWedfNxYEQZD7GownEGRAqGSdGqX0Xo8CQRDkHwrGE8jwhabUA29MqRWAKZkIgiB9gPEEMnzR\nqOSDaj8QcXMEQZDhCcYTyPBlsI9s4H4HgiBIX2A8gQxfBhsfXDv8+V80EgRBkPsdjCeQ4ctg\n1ycolayrvuAvGgyCIMh9DcYTyDBmkPkQNE01Fxz6i8aCIAhyX4PxBDJMoSnNXZTQbik9oVF2\n/xXjQRAEua/BeAIZpqgVYlqjHOxRtEalknX9FeNBEAS5r8F4Ahmm5Pz4lEYpG+xRyu42jULy\nV4wHQRDkvoZ/rweAIPcAtUKi6BJSasVdHItCowiCID3B9QlkOEKplXcXTACAGlU8EARBeoDx\nBDIsoe9efFzSfG0IB4IgCPJggPsdyLCj48ZleXvdXR8+2KoVCIIgwwGMJ5BhBNnjaC48rOwW\n3XUnstbrQzYgBEGQBwWMJ5BhhLih2Njet+HKTl06oTSqoRoPgiDIAwPmTyDDCJpS6b5bgc93\nIAiC9ATjCWQY0Vmbm/fLczp2Iszff9fPhiAIgjyoYDyBDCPUcolEWKZ7P7jlgSAIogXGE8iw\nYnACYH2hlouHpB8EQZAHBownkGHEUKU+SFurh6QfBEGQBwaMJ5DhxCAFyvvuB1MyEQRBbgPj\nCWS4UJP5c2ft1SHpCh/xQBAE0QLrTyDDhYbs7fLOxqHpa6jWORAEQR4UcH0CGRao5WKNSj5U\nvdEYTyAIgtwOxhPIsKDj+iWVtH2oeutuqRiqrhAEQR4MMJ5AhgVDu6LQUX1xCHtDEAR5AMB4\nAhkeDGk8gRKjCIIgWmA8gQwLaBjKJzLUCskQ9oYgCPIAgPEE8uDT3VLZXpE5hB3K2q4PYW8I\ngiAPABhPIA8+XfX5HTcuD2GHGpVcJesYwg4RBEHudzCeQB585KJaWXvt0PapFLeQP2rObyja\ntWioKmUhCILcp2A8gTzgUBqluLF4yLslKZkalby17GRL8TFJY8mQvwWCIMh9BMYTyP3EHXcZ\nNMpuSq1QdAkZi1ouFlUOZfIEoe3a2fKDn3TV5nbVFwJAR82VIX8LBEGQ+wist43cPdLWakWX\nUGBiY2zn8/e8Y0vJcbvgNL6+SV8NhIWHDS1cpKIbNv4T9U3tgLUxMbTUZ+9QSloarvx2c2DF\n6TSl5nDxA4UgyDAFv/6Qu4Smqeun1zUXH7ULTgua+aUuXck7GgQm1pRawTcwu9O7Un3Vfuiq\nL9AoZcK8/RweX2BsrZGLHUY+JDCxbS46osvY+kIp0Q5TaEqD8QSCIMMW/PpD7hJVt6i5+CgA\n0LSuxZ2qT611GPkwl6dn7jaq/5Y0TTOVqcSNJQC0wNhK38xBIW5pvLKLotSdtVcFJjZKcUtz\n0RGJsPwvCiZ6RdxYbO56h/EjCII8qGA8gQyaztrcgl9fhT81u7tqc3XtkabzNj9nYu+nZ2zF\nNzALfvQr5hWVtEPPyIL8LRPVAE21lp3UMzS39puQs2EOTVPOkXN8J71bn/1rY+5e0kwpaVVK\nWgFArRDrOrDB0FWXf//GE5KmUhOHgHs9CgRB7mMwHxMZHMKCg9Unv1bLu5gakWpFN6VW6NIn\nTVMAIBGWt1ddEFVkdDffUtuqPL6K+bu7uUIpbinbv6xk77vyzgZylDB//8W1kxsu7+jZrUYp\n02VUg6W19MRf0W13S+Vf0a0W4sbi+6XoJ6VR3eshIAjSCxhPIIND0lSqVRtKo5SqpLoVd2KJ\na2iU3TfOrmf+lXfUM393t1R0t1YDAE1R9J8/KmqFRNZep5b3shRBa5Q6jWqQSJrKdOyBpjQA\nt+mMKCUt7ZWZ7X+9/Fhjzu4rPzz+V7/LkCBvr7vXQ0AQpBdwvwMZBN0tldLWqp72qz/P43B5\n5O/Il/Zy+fp37KqlOF0quuE+9jl5Z4Na2c1+SVSV2d1cYWznoxA3KyWtlEbF5ekBQG3mRnIP\nTVNqcUPRHd9C9vf+8GhU8p5GtVzMNzAdYA/yjnqBiTVPYMxYSva+2159yTYwydJzzNCMsjcy\nVyWopO00RZ3+OMw2KIW93/QPpKu+wMjG816P4r6EfKAEJrb6prb3eizIAwjGE8gt2quygMOz\n9IxSSlp4+iY8PUOtBs0FB9uunet5oLyzkfmbpjQD+RHtbq1sKTrqPva5Kz/MUUnb2S+p5eLq\nU18LjG0kzeXS1mqluMXAwglYmuM0TV1nrWH0ha6rJoOGptQKrVhK3tloMrB4QlSZWbZ/maKr\nCQAs3CM4PD24mXMKrWWnz3851nHkDO+URUM+aEqtIOkmBGlLZWdNjpGtl56hxZC/l45c/n62\npKnUJiDBYcT0ez2W+wxKo6RU8ppzP9Rf3uEW96xX4mv3ekTIAwjGE8OX9qosfTMH9q0epVZK\nmq9xeXy+gSlNg1Y8QWmU9b2lKWhBUxqJsFzfzJ7LF+ib2tE0RamVPD0DrWbdzddkotq8LQvU\n8q6enbSWnQYA8ttcsP3V8Ge3lf7xPqW+tQAgE9UM4lT/Lkg8IarMvHboE2JxGPVIc+Eh28Ak\nU6eQvo7qqsuXNJdLWypJMAEAHTduq45FU2qSsKKWd/ENzGoyfqRUCo8JLw/JmMW3V/bsbqls\nKUnX5Mn8pn7I4fwj9kPLD/6HlEsnF727uUIlbdczsrzX4/pr0SilnTU5Vj5jh6S3+ku/Vqav\n/LPn7v4bI8jdgfHE8EWt6Na7PY+SpjSKzsbWstOGVi6mTqHMomjbtbNNefvIwsMdu73wdRql\nVlJqhYlDQMD0jyvTVwKHO+Ip7eUESqWgNMr2qgv9dEWSJLqbK8oPfNxSfGxwp3cvUHQ18w3M\nNAoJs9XSVn6mqy6Pr2/STzzRUnys9sIv7G2OXqFpqrX8jJG1Z3PhEZ7A6O7iiZrMnzlcnmv0\nPObBmfaqLK02dRe3AoDv5CUc3j8inuiqy5MIy5l/ZaKapvz9BmYOJo6BhpaukqZScWOx46hH\n7uEI/wrU8q7moqNDEk+U/L5EmL+f+VfaWqXsFgmMrXTvGUHYYDwxjKGp6lNrtfYvLDyiOq5f\nIn+PmPsD2baXtlYP/OeciTm6heVXN86n1HIuT6+56IhdcBrThjzKcecB/vlIqrDgwADf/d4i\nbioxsvFkR11ddXkA0FZxzm3sc30dRU7zjneNzQUHu2qvKsQtGqXU1CmYWdfh6hlweYI7jq29\n6gJNU1XH/8vl6TmOekQmqiHxRFd9fq/tc3580i3uX+yrdq/oWcGs8thXAOA8+jGbgMTW8tOi\nivP3MJ64fnpd27WzbItP6tvmbuF30VVnzdXSP5aQv2lKY+E+mlIr2srPAIC1/0SSRaSFWi7m\n8gX9ZCxJhOVkEjK0V1+SNJVaecfexQgfSKStVRwu39DKDQAojVKjkDLPqCODAuOJYYpMVCMT\n1bRXX9KyM8EEsL7Hmd/1QUHTFPmN1FCa2qxfajM3Oo2eTb73O25kU3/vwxd/D8L8/eYuI3o+\njtFXTc/arI2uMc+wH2/pB0qjkolqybUQNxSd//LmnavflPedImZrNVaIW3h6+nwDM5pSa5RS\nmtLkbVnA9NOYs7ul5Li0tRr6jmMkTaWdN67oEk9IhGU8gbGhpctd9wAAKlknpe59qtRf3kE2\n4PTNHABALRdTarnApPdMw9qsjbaBKSQRZ2jpqs/XEpzrqs/n6hmYOAQMdsNIWHCAnURM05RK\n1lm0axEAxL55utcVhQtrUp1Gz1Z0NlGU2j5ksk1AAvOSRiWnNcpuYVnPzcGinf/H4fL9py61\nDUoe1AgfPDQqWVd9Ad/A/M/wurC15Ljf1A/v9bjuSzCeGKaUH/y4ZzChxY1z3zde3QMA5IdH\nF8QNhQDQcHmHuL4QALrqC3Ts8J+JuKG45PclWreDACBpLL24drKBpcuIp75njA05u+ov/crT\nN5E0Xxtg/70Gdr0GK0U7/8/SI9IzYWFb+dnC317XerXhyk6lpFWjlPb/dp21VxtydrEtVl4x\nBhbOAxxtV12+vL3O1CmYbTSwcDK28+2ZtSqqzGD+tQ+dyiTc1GVtuqPWvFLcXLZ/maGli6gq\na+S8H3ttI2+vbyk9bmzrM7T35TRN9VSbq0xfBQBjF2f2IzTTK1qJya1lpzprcm6+EaUBgKb8\n/W3lZwIe+qQhezuJEjQqmajivER4DYCWd9SLKs4DgL65g3v8goojX5DPb0/IpW/K+0PcWOQx\n/qWBPJB1r6DUCpqmeHqGsvZaZuWPyxMMlWZQW9np0j8+YFu4PEFz8bGRT/9kYu83JG8xfMB4\n4r6nZ25aS8lxRqeqL8SNpXfsmfkuGyrEjSXiB1rXWy3v6hlMAAClUZIiGTUZPxnb+Zq5hNIa\ndVdtnryzsfzAxzq+af2lX1vLTmkZpS2VlEpu4hDQ1VDY85ABZrNKhOVaw/Ob8r6l122/x3x9\nYzL9aJqSdzSwX1JKWmsyf+7ZbeSLe7R+DLpvfyNr33FMPDGQtTGaplpLT/INTFWyDmYZxsY/\nQd/MngmFxY3F9Zd3GFq6hvVI5WGjZ2QxwCCgqy6v+vQ30PfwCrcv5PBufsFaece6xjwzgBO5\nbaVKo5QyMV/+1hddY5+RNBS3FB9zi50vLDwkabr5EWaSS8QNReQ5ahN7P31Tu/bqO2wptl07\n23btrGvsM3cRT2iUUhK7GNl46f7TTtMU0DSHy1NKWiiNysD81jKSStohbbtuYOFcfuAj5hbI\nwMI5euFhHd+0uehIR/UlaY/PAqVRUhqlWtapY//DEIwn7nvaKs4b23qzLd0tFQPJTkD+ZlSy\nzqoTq+1Dp4gqznW3VPa6HX4XSNuuS9uu97RLhGUtJcd76pbpwo2z68sPfsK2OIRNC5jxKQBo\nFN0X104eSCc0pda23P6rTOp6kdBTMTB5WJWsg2jZMzPf0NKt4/qllpLj7Gay9tr+B+mV9IZb\n7HwAkHfUq2Sd+qZ2AhObXlsqJa39f8rYZd+I1O2d6Ts66W6+Jqo4T37krmyY0383armkpfSE\nVnjXFxkrxvlN+cApYhYAVB5fVZu5EQBMHYMint/ez1EKcTPZiLELTus/ecXAwomkJvSDUtxM\nqZWGVm4le98TNxSNXXxryYemNHmbn9dq39cG4u1t1K2lJ/tp0Hh1b89kZPbhd3wLRAsOPbC9\nW+QfS2X6ytqsTWyLbWCS1tco8s+BJzDicLjA4fb6oOw/HA6He3fJNGz0Te24/NsSSGlKwy5h\nEr3wiMDU9uynETq+0V3gmbDQfexzAFC8663m4qMeE172GPdiry1bitOLdr05wG7tQ6cGPvzZ\nHZsVbP83yb78m2FWZVTSDlIyjsvT0zez7+cQSqNSdAkH0rlb7HyvpDe0jFc3PqMUNzP/ak2A\nO8Lh8gzMHQMfWW7mHNpXG5WsI2PFuIH3qUXYE99a+cTd9eHDE4wn7nsUXUJldxvbwtc3uV+0\nGIYtQ/LD/KBibOfD4fKZ9fy/E31TW5LRSXbr+yklqZZ3DbwAq56h+UDyTmSimgfvkyswtu4Z\nmkiE5bovABhaufWzOUVTGonw7kvg99850isYTyAIgiAIoiv/iHo1CIIgCILc12A8gSAIgiCI\nrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCI\nrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCI\nrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCI\nrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCI\nrmA8gSAIgiCIrmA8gSAIgiCIrmA8gSAIgiCIrvDv9QAQBEH+Fop+Wfh16YT3PnvEvZcX5Q3Z\nRw6eyimra5NSAnMHr7CYtMkTfc1Zd1y0uDrj2LHMvIp6kZQ2tLS1dw+NTUyI8jLjA6jaK3Oy\nMjIzrpS1yke9sP6F8Nv6VjZc2LPj8OWqVrnA2jMi9bGZcS76AACgyPzfwk0F7KZ+j694c6LZ\nn6PNUAAAh29kbu3sHzP5oZQgax4AAJW97qUNeVonwBv94jfPj+IAgKw28489J/JvCLtoM+fg\n+IdnpwWYcW6eg7yl7HJmZmZWTmW7YMKbq+b4/Xn4HfoEAJUw59DvRy+W1ot5Vh6jkmY+HO9h\nxBmA25FhA8YTCIIMd7pLflvxzVlO5Oyn3o7wtOJ3Nxad3vnrqk9L5i56OdaOCwBAtWSuX7m5\n0ibp0bkPh3hYCRStVVfSd+1Y0WLw7fPh0HFh5+5y+9i0J8ykXx/W7lxy9ZdVm66H/uvNF0Za\niAt2r/th1c+CZS9Gmt18mT/mlW/+FdZzTMHzvl4/DwCAUoqbSo9u3rB2jfjND2b76AFwI19e\nH3mrpbpo49tfX/ENC+QAAN1y6tsVu7rGzH/52VAbZU3Glu/XrlG/vWSaOxcAoP7Ur4dbfGJn\nzjL8/vvs296uvz4BgGo89b8vf5fHzn1xWaiTgaz28uHzV9o84m3u2uXIAwjudyAIMrxRFP72\n4wmR/5w3nh7nZ2esx9e3cA2fsfDF8fyCrRtPtQEAQPPJn7bmG0957fWZ0b52Jnp8gYlDwPi5\ni9+e6WcIAGARv+DtZx8aG2hr0POGXXjmQLY64pE5kY5Geob24Y8/Gg1X959qGPjwuAJTp7CZ\nj8SZCC9mV/c2/PzMnG7D8LhwAwCAG6ePlvGjHn1stIuJvoGVb+K8NI+6owevyElbl0kL35g3\neYy3pd4dXHJbnwAd57ftqvJ67JVZo93M9fn6Fp5xc57CYALRAuMJBEGGNfKrZy6JLcdOjjFj\nW/neacl+msozmfUAUHv+VBWEJCe53r6gq+ecND7wDr2LS0vqwDsoSPBnvwGBfhxhSUn74AbJ\n4XG5QPf2iiwnK09hFhkXogcAoGlpaQcbeztmoBb29gJVSfG1Xo/ti9v7BBBdziqH4JhIs/4P\nQ4Y5Q77foW7MOXoyr80ifHLaCBveUPeOIAgytNRcv06Bu7ub9s2VhbubORyvvi4Di+rqVnAc\n62lwF723traCUZjVrUP1rK1MoaKlFcASAAA0uRvffFWhNLCwcwmKnzZtnLeZ1kBoVXdzxYm9\nmQrvyRN8e/TflZ1ZqLJNjvMlR/GsrMyhQtishpshRYdQqARlS0sngMVAx6zVJ0BtbS3YeWnO\n//zZ0av1UoGVo0/UlEcnhdngfjnCRsf5IDu2OOW9E9xJXx39zwQDAE3BmpQJb5wS0QAcu6k/\nZO571hvzdRAE+Scjk8rA0Mys5w6Aubk5gEwqBalMBmBoaHh/esXNAAAgAElEQVQ3vSuVStDT\nY3euJxCAUqkEAABDl9jHJiSO9LEVSGpzD237dcUXta9+ODfkz+hDcf7rhZuLAAAEjtFzX57o\n2OP7tO1i1jXKaXqcx5+veI6b6HHuwO4dwRYPhdqoajK2HKviAvN+A6JHn5RcpoLGE9uyEv/1\n+ucBZrLKk5t++GaV5P+WzvHXH6Q7kAcZ3fY7mrYsX33+SlNwcpwBAIB499Ilp3hRc196NMyU\nbj6w+JNjiiEZJIIgyF+FoZEhyLq6VD1e6OzsBDA0MgIjQ0MAmUx2N70LBAJQqdidq5RKEAgE\nAAD6o2Y8PTHExdJA39jGJ27u85Nc2jKOXOhimuqPXbh+/bf/W/XJolmu1Rs///pMs9a2hTAr\nqwq842LsGQvXKfXV1x/2aNi/dsn/Lfp0a4nH49P9BxcM9dKnQMAD2mLsYzNHOBjrG9kETX0q\n0bnt7Kn8QfoCecDRKZ7QnD9xVgmCsQlj9QAAFOm/H1amfnn4l3U7970fCdD22/bjg9q0QxAE\n+btxdffgQl1NDaVl76yp6QQ7D3dDMPbwsIHGqmr5XfRuY2MD0jbRrUNVbSIx2Nj2kszIcXR1\n0aNb20S3m7l6xra+4+Y/GqEuO5FZd9tLNVkX6nlBcWMsb2tv6pv09FsfLV/zzf++WvbqNDNJ\nA1i4uJoOdMC99Wlrawtga2/P/F7Y2dtxqLbWgfaJDA90iifqbtxQA7h6eJBeCi5dkkckJ1sC\ngHtcnAuA9MaNlqEYJIIgyF+FYfj4KNP2c4cviNlWddWR9HKe1/hYFwBwi5/gRRemH69T33ak\nqv74mZI79G4aEOgClcXFf+43qEtLymn7wEDLnk3pxto6Fcfa2qq3fii1mga4bb+DvpZ1oUU/\nLG50P7GCqjjrcpfdmGjPOwyz/z6dg4LMoUUoZEKuZmEzzbe1HWCnyDBBp3iCz+cDgFxOQu+O\nnJwq6xEjnAEAbi7nSSQSXQeIIAjyl6IfMnv+RMvSX1f/cvZaq1StVnTUXt239rvTyuA5zySQ\ndQS7xPlPhEgOrl69+0JFS7dKrZIIS89sXv7l7vI7boLYj58ayb+y59fsRqlKLszZvusCjJo6\n0QkAoPH4jxuP599oESuU3a0VGZt/OFxnFZs6xgwAoHT/ut/Ol9S3datU8s6G4vQNu69yvcZH\nu9zqWFOadandOCJuxO05DOrcvT+dKG4UK1XdwsLD3228YJg8d7LrABPZ+uiT45s8LVB+bsfu\n/KZuhbS15MCWEw32iUkjBtYpMlzg0PTdb0lo9j5p9Mg25ejPq7Lf8RTveNjhKeq7tj/mmgHQ\nu+cYPrpdNfNX2a7HBXfuCEEQ5K+GqTjJwjb1vU8ecQcAWd3FI4fP5JTViaSUwMzeMzQmbUqC\nnwXrjosSV2YcSc/Mr6gXycDQwsbBIyw2cWKUlzkfoOq3xctPdNzWc8CTK98YZwIAAMqGrD07\nDmdXtikEVp4RqbMfHeuqDwBAS+supx86deVanUhlYGnrGhw/efLYm893aLqqMo8eOZdf1diu\nMjCzcvQdPSEtMdzx1nMiytyf3vq2LHbR54/53n5bqBRePrDr0IUyoVzfznvkhOkzxnsaMy/m\n/vTKtxdvW2Xhjn7h2+fD79AnAMhqz+7+LT2nWqQxsnb2j5s+MznAAusNIGx0iieg+485LjO2\ndxj6TZ4zuu3AtoLwjdcPP20LACVLg4M+Lg75pLRgif+QjRVBEKQv9u7du3Pnznvy1tu2bbsn\n7/sPoS/PD3O3DEN0iy+Np3369WPuAln5oZ+2XdGb+t9Vc8mGWu7WbcUA4U/MwWACQZC/hc2b\nN9/Dd6+url66dKmW8ejRo1u3bmVblErlihUrysrK2Mb6+volS5ZoNBq28ezZsxs2bGBb1Gr1\nN998c/nyZbaxpaXlvffeUyhuW3bJzs5eu3Yt20JR1M8//3zmzBm2saOjY8mSJWLxbXkjBQUF\nK1asYFtomt6xY8ehQ4fYRolEsnTp0paWFrjXnkf+OehYf4LrNXd7eeqH2SUd5r4jgp2MySad\nzGDc0s0BxoGpXkMwQgRBkH8uGo1m2bJly5cvl8vlISEhs2bNAoBr164tWbJk586dZmZmiYmJ\nDg4OAHD8+PHXXnutuLg4PT392LFjAKBUKr/99tsPPvhALBa7u7svWLAAAOrq6t57770tW7YY\nGBgkJSV5eHgAwMmTJ9944438/PyYmJiMjAwOh6NSqdatW7ds2bKOjg5zc/PFixcDQGNj47Jl\nyzZs2MDhcCZMmBAaGgoA2dnZr732WlZWVlBQUG5urp6eHkVRW7Zseeutt5qbmymK+vzzzwFA\nJBJ99NFH69at02g00dHR8fHxAJCTk/P666+fO3fO1dW1tLTUyMiIpuldu3YtWrSopqZGKBR+\n991398rzyD8N3fY7EARB/hk88sgjBgZ3U8FSF2pray9dukSqRfH5/LCwME9Pzxs3buTk5FAU\nBQAmJiYxMTECgaCwsPDGjRvkKEdHx5iYmJaWlry8vK6uLgDg8XjBwcF+fn719fWXLl0iaxVG\nRkZjxowxMTEpLi6urKwkx9rZ2cXExHR2dubl5bW3twMAl8v19/cPCQkRCoWZmZlqtRoADAwM\nIiMjraysysrKSktLybHW1taxsbEymSwvL48sLXA4HG9v71GjRlVVVRUVFZHkeoFAEBER4eDg\ncO3ataKiIvIbYWFhERsbq9Fo8vLympqaSIceHh6RkZGZmZkeHh5ubm5azsH9juHGkMQTkspT\n+w5llda3S22TF72Z4qAWNzd1KvXM7O17qTmHIMhwYe/KlbLxzz8x2gwAhOd+3HLVeebCNI++\nWisrD234vURmG//svKgB14Zm+PvjCYqidu7cSeKGYU5ISAhZC2GD8cRwQ+f6621nP5g56/Mz\nzWTrL9j4mTdTHKjMJSPTNogn/dh06F+9PGWNIMjwoL68XBJ6szakvKW6vILX3WdbRdnun09W\nSMXd0uBB1Ia+l3C5XC8vr4qKCvKvjY3NyJEjDQ0Ns7KyWltbAYDD4Xh5eYWEhIhEokuXLpEs\nB319/eDgYC8vr4KCgvLycnJHZ2lpOXLkSHNz8wsXLpC7fw6H4+7uHhoaKpFILly4QKpz6unp\nBQUF+fj4lJWVMSsH5ubmI0aMsLGxyc7Orq2tJYNxdXUNCwtTKBQXLlwgz+3z+fyAgAA/P7/q\n6ur8/HyyBGJiYjJixAhHR8fMzMyGhpuip46OjiNGjACArKyszs5OAODxeL6+voGBgfX19Tk5\nOWQJxMjIKCwszNXVNTMz09zc/G90PPIPRcfnfep/fOqhT84020x47e2ptx6MFiQ//bgTKNO3\n7sJyVgiCDAB5ya5N2Y6zpt1JrvMfRmRkZEpKSnR0NAC0trbGxcXt27dPKBRu2rTJ1taWpunq\n6uolS5acPXu2oaFh4cKFPB5PoVB0dnZu2bKltLT08uXLY8eOBYD29vawsLC9e/c2NDRs2rTJ\nwcGBpunr16//+9//PnnyZENDw+LFi/X19VUqVVNT008//VRQUFBcXDxp0iQA6Ozs9PDw2L17\nd01Nzb59+0i+RW1t7b/+9a/09HShULh06VIDAwO1Wl1fX79+/fqcnJzS0lKS5yGRSKysrHbu\n3FlfX5+enh4UFAQAjY2NM2fOPHz4cHNz8+rVq01NTTUazY0bN/773/9evHixurp67ty5HA5H\nKpVyudwdO3ZERUX13OxAhiG6xRO5335+pAMcn992ePXcCFZ8yh0TE8UF9bkTZ+6TGw0EQe4h\nsqIdm7KdHn1yzH14l2ttbZ2RkbFp06aIiIi33noLALhc7rx588rKyhYvXvziiy9GRkYCgJWV\n1Zo1awoKClJTU//73//q6+sDQHh4+NmzZ3/77bdRo0Z98MEHAMDhcObNm1dRUbF06dInnngi\nOTkZACwsLL744ov8/PypU6d++eWXZmZmABAQEHDo0KF9+/aNGDGCJFQCwLRp04qLi7/44ovp\n06c/8sgjAGBkZLRs2bLy8vK5c+cuW7bMxsYGAHx8fH777bfjx4+Hh4d/9dVX5NikpKTc3NzV\nq1ePHz9+/vz5ACAQCF577bXS0tIFCxa8++677u7uAODi4vLLL79kZWXFxMSsXr2aw0HNR+Qm\nOuVPCL+Z6PDqacvnjoh+SC36KDRkWWHw0oLCZSEAcP51l/g19VFfVl18a6CFXhEEecD43wsv\nSGZ+9k6KNQDc2PP+Z6d9Fn79TLB2K2neTx/91DV12evxljk/vLC+6eH/fJBmN+j32rx58+HD\nh4di1INmmCcK9OX5Ye6WYYhO+RMdHR0AYGNrCwBaUSqXywUArSeqEQRBtBBf3bY5137m0rE6\n5lrNnTt37ty5QzMmZDCg5xGCTvsdVlZWANDW1tbjFUV1dQMAx8Fh8HcZCIIMH8Q5W7fkOT48\nN94al80R5P5Gp/UJ2+hoLzhZdfp0ASTz2esTwi0bj2gARsaPNdN1gAiCPMDUlZRK9Cyu/LIq\nBwAAxI0AynM/riwKSHvz4R4bIwiC/IPR7XnRUc+/Grfy/zK+fmlp8ic3n8FWt5fs+WjuG8fl\nYDp54XyfIRgigiAPLO5Jr74ZeauAQ/mBlftFgQnTolwc7uGgEAS5C3SsP+Hx2i8/nB0///eP\nJyZwOABQ9kW0zUcKCoDvMefn7+fjbgeCIP1hZO/jZ3/rX4kpQLeNp58ffncgyP2GrvWsuF5z\nd1/xXf/JJ9/tOlnQKFMrFAKbwAmzX/lo2UvRtkMyQgRBHhSURbtXrjxy63+90JkLUzzu2XAQ\nBBlChlC/g1KKOyWUobm5AW+IekQQBEEQ5L5Ax/qYt3UlMLW0wmAC0Zm9e/c+0QcA0NXVtWfP\nHnZ7mUy2Y8cOdmSsUqm2bdvGFlbQaDRbt25lP8BM0/TWrVtVKhW7q23btmlJP+/YsUMqlbIt\nO3fuJAWM2QMmz04z7Nu3T+u5p0OHDgmFQrbl2LFj9fX1bMvJkycZySjCuXPnmHLOhAsXLpSU\nlLAtOTk5+fn5bEtBQYGWqnV5eXlGRgbbcv369VOnTrEtjY2NR48eZVtEItG+ffvYFrFYvHv3\nbrZFLpff0fkURW3dupUUaSb06vxff/2V6FEx9HT+rl27tPS1ezp///79pNw1w+HDh7Wcn56e\nruX8U6dOXb9+HQDeeeedvuYeAJw4caKmpqZ/n/Q6hpqamhMnTmg1++2337q7bytBnpeXl5OT\nw7b06queE0Mul2/fvl3r/vDw4cOMdhehubn54MGDWsPYs2cPqavNUFZWlpmZqdWs5wW6ePFi\ncXEx9O00xm/IsIAeAtRdtYUXzhw/euSwFueviYeif2R4MWbMmDl9QNP0okWLAGD8+PFXr14l\n7ZctWwYAo0ePzsjIIBZS8o8UHyQWoqpMSgoSy9atWwHA19d3//79xPLHH38AgKur66ZNm4jl\n5MmTAODs7Lxp0yaKomiavnz5MpfLdXR0XL9+vUajoWm6qKiIz+dbWVmtXr1arVbTNF1ZWamv\nr29pafnFF18oFAqapuvq6oyNjY2NjZcuXSqXy2mabmtrs7a2NjIyWrp0qUwmo2m6q6vL0dHR\n0NBw8eLFXV1dNE1LpVJ3d3c9Pb2FCxd2dHTQNK1UKv39/fl8/oIFC1paWmia1mg0ERERXC53\n7ty5QqGQpmmKouLj4zkczty5cxsbG8mJpKWlAcDUqVOrq6uJhRRPnDp1amVlJbHMmzcPAJKS\nkgoLC4nllVdeAYCEhIS8vDxiIaLYY8aMuXDhArF8/PHHxPnnz58nlv/+978AMGrUqDNnzhDL\nDz/8AAD+/v6M83/99Vfi/N9++41Y9u/fr+X88+fPczgctvOvXLnCOJ+4uri4WE9Pj+38qqoq\nAwMDUlCSOL+pqcnMzKwv50ulUsb5AoFg4cKFjo6Ofc29zs5OBwcH9jWiafrtt9/W8knPMdA0\nPXv2bOLwiooKYsnMzORwOE5OTsxcoihqzJgxWtduy5YtAODn58dMVKlU6uHhwZ4YNE1/8skn\nABAREcFcCPaJkzlG0/Rzzz0HAImJiQUFBcRSWFjI5/Otra0ZH9I0nZqaSkZ7/fr1vi6QSqUK\nDg4ms9HBwYFxVHx8vJbfkGGCrvsdspLNrz+7eGNWY6+FtZlymQgycJydncePH9/TXlNT4+jo\neOLECSLTzOFwQkJCvLy8MjMzGfHlwMBAHx+f7OzsxsZGcpS/v7+fn19ubi4jleTj4+Pv719c\nXFxdXU0snp6eAQEBFRUV165dIxY3N7fAwMCamhpmMcDZ2TkoKKixsbGwsJBYHBwcQkJCWltb\nyc8tANja2oaEhIjF4itXrhCLmZlZSkpKbW1tdnY2uWU3NTVNSkpqbm6+cOECWS8xNjZOTEzs\n6OjIyMggFkNDw4SEBKlUeu7cOUZ+evz48RqN5syZM+RWVV9fPz4+nsfjnTp1iuhlCwSCuLg4\nAwODU6dOkftIPp8fExNjbm5+4sQJIijF4/HGjBljbW198uRJcmfM4/FGjx5tb29/6tQpcuvP\n5XJHjRrl4uJy+vRpctvK4XBGjBjh7u5+7tw5kUjUq/MBgEhVaTnf19c3Ly+Pcb63t3dAQADb\n+R4eHoGBgWznu7q6BgUF1dbWkntfAHBycgoODm5qaiooKCAWe3t7orOVm5vLdr5EIrl8+TLj\n/OTkZKIAzqiHJycnazk/ISGhq6vr/PnzxMLhcHx9fSMiIrTmnpOTk1AoZK6RkZHRxIkTDQ0N\nT548qeWThoYGrTFoNJr09HTicC6XGxER4erqSjS9SP+Ojo7BwcF6enonT54kK2R6enoxMTF2\ndnaXL18mCyfMRJXL5VoTw8zM7MyZM83NzaQZ+RS0trZevHiROfHExEQej3f8+HGikM7lckeM\nGOHh4VFVVZWfn09Ga2dnFxISYmRkxEwYPp8fFRXl6OiYn5+vdYEoijpz5gyZe0T5LCoqCgDO\nnTsXHx/PuA6rZA4fdIsnuo8/75+yoZ4GU6+x46N87I21Njucp3/80XQn3UaIDDv6iicKCgqY\n33LkvsAnIoJfV1AqVAKAkXNIoJ3k2tXrXX02N3YNDbATyOoLi5sUfTb6e0hMTLSzu/WQCc69\ngRAbG0uCTownhic8slZ8d0j3vP3Ez8Ua64d+KTi3+sXZM6ZrM9HfdOiGigwXVq1aRTQStVAo\nFNHR0Y6Ojl1dXeTmicPhhIWFhYWFicViZqM9KCgoPDxcIpEwO9N+fn5RUVFSqZTZevfy8oqJ\niZHJZOReDQDc3Nzi4+OJ9iOxODk5TZgwQaVSMRY7O7ukpCS1Ws1s2FtbW6ekpNA03d7eTkJz\nCwuL1NRUDocjEomIxcrKKjk5mcfjMRZzc/OUlBR9fX2RSMQsWiQnJxsZGTEWIyOj5ORkExMT\nkUjELFokJiZaWlqKRCJyb6qvr5+QkGBjY8NY9PT0JkyYYG9vLxKJyDIGn8+Pj493cXFpb28n\nt5I8Hi8uLs7d3b2jo4PcCnO53OjoaG9v787OTrKwweFwIiMjAwICOjs7GVePGjUqJCSkq6uL\nuJrD4YSGho4cOVIsFjOuDgwMHD16tEQiISkmUdOmefBlrgERUqmU5xKZMIJfe/W6tadnXFyc\nXC5nnO/q6jpu3DiunX/8SC9XZ5PO0hKOlWNCQoJKpWJcbWtrS270yeoUcayW883NzdPS0jgc\nDmOxtLRMSUnh8/mM88mKEdv5xsbGKSkpxsbGbW1txEJkwQUCgdbcCwwMFAgE7GuUkpLi7u7O\n9lJYWFhsbCwAsMcwadIkOzu79vZ2xuFRUVERERHsE7SxsZkyZYqZmRlz7fT09MaNGxcUFKQ1\nUZOTkw0MDNra2piJkZSU5OXlpXUhxo0bx+Vy29vbyWjNzMzS0tKcnJw6OjqYqxwREREVFUVR\nFNk3IaOdMmWKlZWV1oQh6ufMZ8HR0TEtLY3MWDL3eDxeUFCQoaFhTU0NUQ4jzJw5s+dnGXkw\n0WWzpPyzcACwfiFdl04QRIuIiIi+8ifUarWXlxeZuhEREefOnaMoKiTk5p5aaGjoiRMnaJoe\nM2YMsTAJE4mJicTi7e29d+9emqZnzJhBLG5ubtu3b6co6qmnniIWJycnsmf/4osvEoudnd33\n33+vVqtJ9gYAWFlZrV27VqVSffTRR8Ribm6+YsUKhUKxcuVKYjExMfn0009lMtn69euJxdDQ\n8IMPPpBIJMxNm76+/ttvv93V1cVkPurp6b3++usikYhkbwAAj8d78cUXW1paLl26RIRyuFzu\n/PnzyeYL0crhcDhPPPFEbW1tVVWVnp4eOXDmzJlVVVUNDQ2GhobEMnXq1LKyMpFIRDQqASA5\nObmwsFAikVhbWxPLuHHjcnJyFAqFi4sLsURHR1+8eFGtVvv43CxRR3JTKIoKCwsjlpCQkPT0\ndJqmya8pAPj7+3+wYMHnR1tTUlIAwDbq8QXPTlu/ezf9Z/YGALi6upLkzXlPz5/2zOzIqJQF\nCx7/5NufNRoNyd4AAFtb2++++06tVpPsDQCwtLRcs2aNUqn8z3/+QyxmZmZffvmlQqFYvXo1\nsRgbG//nP/+RSqUbNmxgnP/+++9LJJLt27cTi0AgeOuttzo7O5kURT6fb2xsPHPmzF7zJ37/\n/Xf2NWpvb1er1d7e3oxPzp07R9N0zzF0dHRYWFgQY2JiItlfeO2114jFxsZm3bp1arX62rVr\nfP7NZ/hnzZpFchceeughYnFzc9uxYwdFUUwiLTMxKIoaOXIkMQYHB5ML8dNPPxGLgYHBe++9\nR2Jue/ub1T7i4uKys7Npmv7www+JxcLCYuXKlUqlsr6+3sDAgBinTZtWXl5O/5leAwCOjo4b\nN27UaDSXL19mZqORkdGMGTOIo2JiYjB/YniiUzzR/M1EAHD7v8yhGg2C0DT98MMP95OPuXPn\nThsbG3bu2PHjx0lGnkqlIpbMzEySDkny72iavnr1qpmZGTsx7dq1axYWFosXLxaLb2YN19TU\nWFhYLFy4sLOzk1iEQqGNjQ07600kEjk4OCxYsKC5uZlYuru73d3d586d29TURCwKhcLPz2/W\nrFk3btwgFrVaHRoaOnXq1KqqKmKhKCoqKoqdnUfTdHJyclJSEpMoR9P01KlTJ06cSFIECHPm\nzImMjMzMvPWhe/7550loxVjeeOONwMDAI0eOMJYPP/zQz8/vwIEDjGX58uVEtJqxfPPNNy4u\nLkzyI03Tv/zyCztbkKbpPXv2aCXunThxQsv5WVlZjPPXLljw+dHWwsJCc3Pz9954acG/N5BU\nz4qKitucTwl3/ufVOeO8F7760oIFSw8LbzmfyTylabqzs9PJyYnJPKX/zExkZy8qFAp/f3/m\nx5g4PywsTMv5cXFxSUlJRUVFzOmnpKSQX3onJ6e+5h5N0+PHj2enrNI0vXv3bi2f9BwDTdMr\nV65kJzPSNN3W1mZvb88+QZqm//3vf7NTWWmaLikpMTc3Z09UmqanTZs2YcIE9sQ4ceKEpaUl\n+0KQXF12yi1N0z/88AM7xZWmabFY7OzszJ7ANE2///77/v7+Bw8eZCy1tbVanw6app988kky\nG/tyGsYTwwrd8icqvgz3W5w3cZ3wxEs2d98LgtxGP8LT5LZeIpGYmJiw7cPH0t3dbWRkxJbz\nlUqlBgYGZJWCIJPJBAIBj3crnUkul/N4PGbdAgCYFE7GotFolEols5IBADRNS6VSY2Pjux42\no1cukUjajn3B1itnHUU1Hl3x2XnPF/5vckj1r2y98nvi6hdffJHZXGDT19wbYP8qlYqiKH19\n/f6b9byaA+x/gM0oipLL5UZGRv03k8vlfD6fWSzpqxkzG/tyGmD+xHBCx+c7ZFnvxyZ82vbo\nzoyfH3XVtdYmgiAPFkw8AQA39rzPjicYNPWHvvg8w+e1Dx/z1YecH9jxBIIg9xE6xQBNx9bt\nlkdGO/20ZVZodmLquGBXM73bGjimLHozBXV9EATpA3XtgZ8Odsf+e4av/p0bIwjyD0aneKI1\n8+eVK8nz051lJ34r0y7+BsEmz2A8gSBIn1ScO11n6hVw7dj+awAAjQ0AkpKT+1VeI6dFud7r\nwSEIMgh0iidcH/3q14COfhpYhOA3AoIgfWMTkjrVVHXndgiC/NPRKZ4wD0l7HKtfIghy19iE\npU0Lu/VvTtOBK02BCdMwfwJB7juGMIdSLe+W03pGhoIh1BhDEOQBQl17cf/+qlv/893GTBqB\noQOCPBDorleuqTux8v2PN+y/UCFS0kSwQ3p+zWsbC4zGvbFmXvCdO0AQBEEQ5D5H1/WJ2m1z\nxszd2UgJTE30QHlTFMzIi1+68cfzx0zmzF0dzem/AwTRZu/evTt37uz1pZ7Psl+9enXUqFHv\nvPMOoyK9bds2qVRaW1vr7+/PNFMqleXl5UwlTQDQaDSFhYUjRoxgLDRN5+bmjho1qmf/d2EZ\nOXIku0pEbm5uWFgYu65Afn5+cHAwu0pEYWGhv78/u0pEcXGxt7c3u2hBWVmZq6sru37AtWvX\nHBwcTE1v1bavrq62sLCwtLRkLLW1tYaGhjY2t8rEEMkuR0dHxtLa2iqTyVxdb+U8dXZ2tra2\nMvUfAUAmk924cSMgIICxqFSqsrIytmMpiiooKPiLHJubmztixIghcayXlxdTBRJYjmXPJTZk\n7pWWlrq7u7OrdPT0AAAUFBQEBgay6zd0dHS0t7d7enpqnaDWPKmrqxMIBGzpkF790POiUxSV\nn5/PVMlkTtzPz49dZaS7u7uuro790YDefCgUCjUajZPTbepLPYdRXV1taWlpYWHRl9MA608M\nJ3TbmxDvXfTyzkbaZ8Hh2ow3fW/ZnR59NAagdteuizoODxmObN68eYAt//jjj/Dw8GnTpm3c\nuJFtX758eWho6AsvvNDa2kosa9asGTFixLx58xgNxg0bNowaNWrevHlNTU3Esn37dtIbI+d4\n4MABYqmqurlIn5GRERERkZyczChDXr16dfTo0QkJCdOhp6AAACAASURBVPn5+cRSUVERHR09\nZsyYrKwsYqmtrY2Li4uMjDx//jyxNDc3jxs3LiQk5MiRI8TS1dWVnJwcHBx84MABYpHJZJMn\nT/b19f3ll1+IRaVSPfTQQ35+fr/88gtZWaQo6oknnvDy8lqzZg1RcwCAZ555xsfHZ82aNURY\nAQBeeOEFX1/f5cuXE/0IAHjjjTd8fX2XLVtGpBwA4L333vP393/nnXeI9AYAfPzxx4GBga+9\n9hpTqqinY7/++mstx/7444+jRo2aPXs2oym6Y8cOLcemp6cTS2VlJbFkZmZGRESQ0pPEkpub\nGxkZOXHixLy8vL4cW1dXN3bs2MjIyHPnzhFLS0vL+PHjQ0JCmJJoYrE4JSUlODiYCVJlMtmU\nKVOIG4lFrVbPnDmTuLqf6adUKh966CHSjFnZXbt2rZYHRCLRxIkTQ0JCDh06xBy7dOnSgIAA\ntjN7ng4AvPzyy97e3suWLWOu1MmTJ8PDw5OTkxm1VYqi5syZ4+/v//333zMX/eeffw4PD589\nezbzuy6RSFJTU9knDgBffPFFaGjoa6+9xuiG9PQhACxatEhreuTk5GjNcwB45plnvL2916xZ\nM/DPLPIAo9N+h2LX4+azdmiS1zcdW9D0UWjIskJGoPzgfNOpGyUJ65pPvGQ7dKNFhgXR0dGM\nSAcbohnNtrA1ml1cXGJjY+vr6+3t7Q8dOkS+B4mot7m5+ZEjR4hUEhH1tra2Tk9PJ+JGenp6\n0dHRdnZ2p0+fbmtrAwAej0c0mjMyMoRCIbFEREQ4OztnZ2fX1dXBn6Lerq6uubm55GeSaEF5\neHgUFxcTZWcOhxMUFOTl5VVRUcHongcEBPj4+NTU1DDfy76+vr6+vkKh8MqVK8RCxJaKi4sv\nXrwZktvZ2aWlpVVWVmZkZBCLjY1NampqfX396dOniYXoY7W2thIRE/hTeKyrq+vYsWPEQoTH\nFArF4cOHGSXrhIQEADh06BCJP4gSt0AgOHjwICmjSUSxTUxMDh06RISv9PX1x44da2FhcfTo\nURJ/CASC2NhYGxsbxrFELb1Xx2ZmZpIwjoh3u7i4XL58mcQfXC535MiRbm5ueXl5RNOcCI95\nenqWlJSUl5ezHVtZWcn8yhJt9NraWib+cHZ2TktLy83NZRzr4OCQlpZWWlp64cIFxrGpqanV\n1dVMqMflckNDQ4OCgnrOvYaGBi3/KxSKI0eOEA/o6enFxsba2tpWVFTk5uaSZk5OTmlpaSKR\n6ODBg0Tly9DQcPz48aamprm5uVrzRK1WHz16lK1rz+fzMzIyGhoayMDCw8NdXV1FItGZM2eY\ni56amqpWq9PT00mIwOfzo6Oj7e3ta2pqsrOzSTN7e/u0tDSxWKz10bCwsCguLtbyIQAcPnyY\nreduYGBw6dKlGzduAEu8XiKRnDx5kswrLpcbFBQUGhoKAHV1dYzyC+D6xHBCp3iicnmkzzuX\n3d/Muv5VdNHt8cSVd31Hf1Ex8pPSq0v879gPgrC5a71yBwcHGxubB0NXmsvlkt8VBg5H+9Pa\n04JowePxmDt4wgAdm5qayt4w6nXuDdD/PcdwT+h54kPOhAkTHB0dUa982KKTXnlH5oavjzdY\nxD73eopL69nv1p1utpvw8ssT7AAg79cPtl5V+Ex/51/RZkM2WGR40L9eeTALsg3MSFGbmpo6\nOjqS5Q25XM5oKwcGBsbExLCVx319fePj49nK4x4eHhMmTKAoipGZdnFxSUxMZItf29vbp6am\ncrlc5h2tra0nT56sp6fHaFibmZlNmzaNqFozd3jTp08nys7EYmBgMH36dFNTU0brWSAQTJs2\nzcLCgq1VPXnyZFtbW0Y2ms/np6amOjk5MbLXPB4vOTnZ1dWVUaDmcrkTJ0709PRkK2iPHz/e\n19eXrTMeGxsbFBTElnSPiooKDQ3t7u5mNjvCw8NHjRrFVsoODQ2Niopii1b3dKyPj098fDyj\nKp62YEGEidR31FiK0nOJn/GQhzi3os3J2TkpKamnY42cQ1MSYsaM9tRcK1EaW/XqWKIzTtxo\nZGR0R8fq6+sT9W22YydNmtS/Yzkcjq+vLzu7gpl71tbWbCHvlJSUoKAgLQ8kJiZqXdwpU6Z4\ne3tLpVKxWEyajRgxIj4+XmuePPTQQ87OzkQIlAwjJiYmKiqKw+EwU47ImtvZ2bEvekJCQlhY\nGFvM3d3dPSUlxdzcvL29nVGunzRpkr+/P/ujERQUNH78eENDQy0furm5sSdDZGRkTEwMkX0n\nl8Pc3Hz69OkODg7MTAMALy8vExMT1CsfvugiJibf9bgBgGD6LxKaLv44DACClxbQNE1TRR8E\nAYD580c1uvSPDE/60SvXajlx4kQyjcka7GOPPTZnzhyZTMakszk7O2/evFmpVDJfcHZ2dj/8\n8INarWYWtBnlcUbl3MzMbMWKFUqlMikpiViI8LRMJmOEtg0MDJYsWSKRSJ5++mliEQgEixYt\n6uzs/Pe//00sfD7/1VdfbWtre++994iFy+U+99xzQqHwiy++YEb+1FNP1dfXr1u3jvlUPvro\no9evX2fvSU+dOrW8vPyPP/5gLCSHg9E0B4D4+PicnJzs7Gwmv2/MmDEXLlwoKSlhUu1Gjhx5\n5syZGzduMMmJQUFBx44da2lpYXI8fXx8/vjjj66uLkZi293dfceOHXK53MHhZrlbIumuUqmY\nBEM7O7v169er1WomM9HKyurNBQs+OyyMjo4GMA6bsWDB1BHLv/xSoVAQBXO2Yx+d/eSfeuWP\nvrH4PYlEMn/+fMaxb775ZkdHByPwzefzX3nllba2tvfff59x7LPPPtvU1PTll18yjn3yySfr\n6uq+++47xkUzZ86srq7eunUrY5kyZUpZWdn+/fsZi76+/uTJk3ude3v27GGapaSkFBcXq9Vq\nZm+OeECj0axatYoZw5w5c2prazs6OhiBeE9Pz127dtE0zZ4n5HTKy8uZNNKwsDCymzBt2jRi\nMTQ0/PDDD7u7u0+cuFWKeNy4cVevXmULxxMxd5VKxQi1A8DDDz9cWVkplUptbW9uQJOPBkVR\nbB+SyVlfX8+kAAcEBBw+fJim6aeeeoq5HG+//XZXV9elS5eYmSYQCFJSUlCvfJijUzxBd2yf\nYQJgMG5tlYYVTygrfphmDQA2zxyRDdEwkeFE/3rlbNLT0zkczqxZs+zt7dltPvnkE4FAsHDh\nwq6uLtJyzZo1enp6bOXxn3/+mc/ns5XHd+/ezeVy2cLNJ06cIP0zwtNZWVkcDoctfl1UVMTn\n89ni15WVlQYGBiRzjVgaGxvNzMyioqKysrKIRSQSWVtbR0REnD9/nljEYrGTkxNbZ1yhUPj6\n+rJ1xonuOVtnnKKoMWPGaOmMp6SkODo6snXGZ8yYQVTFGU3tefPmEVVxhUJBLK+++qqxsfHS\npUsZkfd3333XyMiIrZT92WefEccyotVr167VcuymTZuIG5ubm4le+Z49ezgck/nPLVjw1SFl\nr46lhD998MrsaMdps+YuWPA+0SsvKirS09Nji4NXVVUZGBiQDE1iaWpq6ulYGxsbtoB7d3e3\nm5vbHR0bFhbm5ua2adOmfvTKVSpVSEiIls77//73Py0PdHZ2Ojg4hIeHnz17lmn21ltvGRkZ\nLV26VCaTac0T5nRomn700Ue1rlRGRgaZctXV1cxFj4qK0rromzdvJm5ni7m7u7sHBAQcOnSI\n6f/jjz/W+mj09CFN088++6yFhQURnSeWwsJCMs+Li4uZZsxMQ71yhNZVrxw0xSsnRC06r3JL\nmO5WvOt8k1Pqq7PNMzfvzGmjzVO/zz3yvIcOnSPDlDvqlbMpLS0NCAhgyyVv27ZNLpc3NTWx\nN03UavX169d9fHwYC0VR165d03pwjvT2F1n8/f3ZTwaWlZX5+fmxLeXl5d7e3uwHHSsqKtzd\n3dkPOlZVVTk7O7OfIL1x44adnR37Cca6ujoLCwu2tHRjY6OhoSGz3gAALS0tXC7X2tqasZDV\nfmYFAgAkEklHRwc7t06hUDQ0NLAfelSr1dXV1SSJj0DTdHl5OXEsoy9aWppd8/uG3SZP/+/V\nWD1tF93UK3/4sbAE5Tm2vuhdO9bX15f99OO1a9e8vLz6d2x1dbWTk5O+vn7/euVVVVUuLi7s\nJzB7egAAysvLfXx82GPo6uoSi8XOzs79n05TU5O+vj47daPXZj0vOtvt7BP39PRkP7Yqk8mE\nQqHWfmJPH7a2ttI0zSxm9DWM2tpaS0tLExMT1CtHYCjqWYnOfz7viY8O1ipYNn3XKf/Zue2t\nMZg6gSDDGZZeefuxL95hxxMMqFeOIA8Gutfbthr77oGKZy4f3H/8clmjmDK09Q5PmjE9xsXg\nzociCDLMQb1yBHlQGBr9DoHj6IcXjH54SPpCEGT4gHrlCPKgMIR6YAiCIIME9coR5EHhbuKJ\npswtv+dLHOPmPhRqDACV3z/+2PcVWm30wxft/v5xh94ORxAEuQnqlSPIg8JdxBP12xbNezPL\n+c2sFx8CAABZQ+GVK0Xara68+GpC0q7HbXocjiDIcEWYe2h/G0s0yH7UlCgX3USEEAT5hzD4\neKLlwP6LNNhPmz7mNrPH0z/9/MyfT5HVbvnXvB/3fL6u7PEPsdo2ggxboqZOVXqTZxoNvcdO\nndp+pwMcI6ZOlfgY/+UDQxBkyBl8PHH5UjYFEBEZebsQubFH5IQJjGKv0zMrflyav2t3+Yfv\n+ek+SGR4MSi9cpFIZGVlpaVXztjZLbu6uoyMjNjP4vfarKdFo9FIJBJzc/P+m0kkEn19fXZJ\ng16btbe3W1hYsJ/1p2m6o6NDq+RAzwOlUimXy2UXge61WUdHh7m5Obv/AZ6mQqHQaDRsJfRe\nm3V2dpqYmPB4PC2JaubSsA+J+rO2o0gk9R47zZvVj1wupyjK6PbFiU4j7ylTTLm3G3uOYeDG\nXq9dr1eKuS7965W3t7drXakBXj6FQqFWq42NjftvJhaLDQwM7jiL7voqq9VqqVTKFOvsq5lU\nKuXxeOwyJ7026+zsNDU15XK5g/rMIg8qg15qlAiF3QBmDg6G/bXyGznSCKAwK0ty90NDhisD\n1z6+cuWKm5vbO++8wwhPE4RCobe397x584g6KOGVV14JDQ1lV8rKyspydXVlK3RTFJWamqol\nyvz9998TUWZG1am9vT0gIICtDQ0Ab7/9tpY2dG5urrOzM1uiGgBmzJgRHR3N6FsCwLZt27y9\nvZcvX05UIQBAIpGEhoZOmzaNCGwSPv74Y7bENgBUVla6ubmx1cMB4IknnmALowPA/v373dzc\n2BLYSqUyOjqaLYENAF999ZWWEnd9fb2HhwdbiRsAXnjhhdDQ0KNHj/Z6mSorK93d3bWGtG/f\nPq0BAMCKFSvYwuuE5557LiIighHPBAC1Wj1+/Pjk5GS2HFdzc7OPjw9bax4ALly4QK4mo5EB\nABs2bNAScweAxYsXa10pYF2XfqYfTdPTp0+Pjo5mdF8BYPv27VqXTy6XR0REsNXYAWDVqlV+\nfn7ff/89I8rV3NxMZin7LBYtWhQSEsJo1gNAcXGxi4uL1iyaM2dOVFQUI3YKAOnp6cTJjMJ4\nr65bv369lkO6urpCQkJmz55N5EMJy5YtI5OBsVRVVZEr29LSwhiZ64V65QjA4PU7KpdHAoDv\nu1cYS+3ud5588sl3dteym2Uv9gaAyOWVd125Exm2jBkzZrD6HTweLzY2lmnz0ksvEbuFhcWq\nVauUSmVeXh5zM0e0DGiaHjt2LLG4ublt376doihG2YGR3mCrgTAVlBcvXkwsJiYmn332mUwm\nq6ioYMovTpo0qaSkhKbpSZMmEYujo+PGjRspimJ0IohWQlNTk0qlYooVBgcHE2HxTz/9lFgY\nyYbGxkbmZpEpOP3YY48Ri62t7XfffadWqxntciILUldXR1FUcHAwMfr6+u7bt4+m6bVr1xKL\nvr7+22+/3dnZ2d7ebmpqSoyxsbHZ2dk0TT///PPEYmlpuXr1aqVSeeXKFcaNhoaG06dP17o0\njz/+uNaQKIpipFL8/Pz2799P07RIJGLKd8bFxV2+fJmm6UuXLjHfS0wpbkaEQk9P7/XXXycS\nYgsXLiRGc3PzFStWkJLhjCatq6vrtm3bKIqSy+X29vbEOHLkyNOnT9M0XVlZyaxRMVeKfV2M\njIyYiu/x8fHsE2T0O3g8HqnUrlarmWqhQUFBR48epWma0e8wMDB49913xWIxW78jJibm0qVL\nNEu/gzkLtn4HERahWfodDg4OP/30k0ajYfQ7OBzOvHnz6uvraZpm9DuI/EqvrmPrd4SHh5OS\n5Ix+B9FSkUql9fX1TAHQxMTEgoICmqXfYW1tvW7dOrVazb5eVlZW7MmA9baHJ4Ouj1n1ZZT3\n4myiUd5Ps6w33WNX1YxZUX1hkceg+keQvvTKa2pqnJyc2BapVJqfn19bW0v+9fT0dHJycnJy\nUiqVxcXFlZWV5F7QysoqMTGxrq4uPz+fKGry+fz4+HgjI6O8vLy6ujrm8IiIiMrKyqKiInKv\naWRklJCQIJfLc3NzyQ03ER7z9PQsKSkpLy9ntBaTkpKEQmFeXh65OebxeHFxcWZmZgUFBcxt\nn6ur65gxY6qrqwsLC8mdur6+fkJCgkajyc3NZdYAQkJC/Pz8SktLy8rKyE2kiYlJSkpKW1tb\nbm4uUYbkcrnR0dFEnJ3IOgCAg4NDfHz8jRs3CgoKiOSjQCCYMGECl8vNy8trbGwk/fv7+4eE\nhJSXl5eUlBABTFNT08TExM7OztzcXCJQyeFwIiMjHR0di4qKSOwFADY2NhMnTqytrc3PzycC\nmFwuNzEx0cbGhrk0CoWi55DEYjF7AAEBAaGhoa2trVevXiX6ruTt3N3d2Z3r6emNGzfOwsKi\noqKiuLiYiGQaGxsnJiZyudzi4uKKigrif0tLy8TERJlMxp4MHh4ekZGR7e3tzLUDgJEjR/r6\n+jY2NrKvVGxsrL29Pfu68Pn8tLQ0U1NTRnqbnCBN01VVVUVFRaQZUd9Qq9V5eXnMShiRvS0v\nLy8tLSXuNTMzS0hI0HLv6NGjnZ2de54FGRtZMCOzyNzcvKCg4Pr166R/FxeX6OhocpXJUoRA\nIJg4cSKXy83NzWWc7OfnFxoa2tN1UqmUPZlHjhzp4eFRWlpaXl5OJpuZmVlycnJLS0tubi5Z\nEeFyuTExMdbW1uwra29vP27cOK3JkJSUxK7gTsD9juHDoOMJ8capZvMP6j28tWvPE32XwJRt\nfdj8qd9VUzeK9z9t0mcrBOmNvuKJgoIC9sptT6ysrJycnPpvgwwtsbGx7u7ud7w09x1JSUm2\ntrZMPKHLCXI4ussa3B/ExcW5ublpGTGeGD4MOh/TNDIyAA6Wnj6YLn1imlEfjbqPHTitAgiK\nisJgAhkyzM3NZ82a9f/t3XlAE2cWAPAXAgmHgByBACIqCAKKobBYlCqooK0GELHg2WqtbXVb\nV1vrvd0elqqtdltbLd619aIWxaMY5JRDVCQipyKHmJiEGwWSQJj9YyAMgQ2WYMH6fn+R+TKT\nL/M9Jm8m3+RRl6SnpwsEAvJvFos1adKkysrKefPm3blzp7CwkFxO1pYkCCIxMZE8UWMymeSJ\nb2pqqvJ8zsXFxd7evrq6+tq1a+T5oqGhob+/PwDweDzl+aKHh8ewYcPu3r2r/HSxsbFxcXHR\n0dG5cuUKeVWDwWD4+PiYmppmZmYqT5cdHR0dHR0bGhpSU1PJ7ZPniwwGIy4uTnnVgcPh2NnZ\nlZWVZWdnkyuy2WxXV1d9fX0ej0eeFuvo6Hh7e1tYWNy6daukpIR8mr29vZOTk1QqTU5OJk80\n9fT0/Pz8DAwMEhISqqurAYBGo40bN27kyJFCoZD8igEAhg4dGhAQIJfLL1++TF7V0NbW9vLy\nsrKyys3NLSoqIrdvZ2fn7OysUCgSExPJ024ajebg4EAWglcOTVZWlkqXVDrg5uZGfr/D4/GU\nLzdhwgQ2m11fX5+UlERuXFdXd8qUKUZGRhUVFZmZmWRXjYyMAgICaDRaXl6ecuaHra2ti4uL\nkZFRWlqaUCgkFzo7Ozs4ODAYDB6PR16H0NLS8vDwsLW1lUql1JGaNGmSmZmZRCJRjguNRhsz\nZoxKKSzyDYpEorS0NPJpQ4YMmTZtGjnu5FUWcviGDx/+8OHDrKwsss8mJiYBAQEymaz77s3P\nzy8oKKC+Cy0trfj4eJUounHjhvISl6Oj4+jRo5uamlJSUshR1tfX9/Pz09fXT0xMVF51IEdZ\nLBar7Lq2trbuwVxcXKycLWRtbe3q6spkMuPi4pTBNnHiRBaLxefzi4vbf2do1KhRTk5OLS0t\nyvEig6F7MoFeLH/+K5LcrS40AB3OlhvSnp/QfG3DOG0A2thP8v781hHy8PB4yvkThw8fBgAP\nDw9zc3Pqc8iq4tS60gqFwsvLS6WudGRkJAB4enqmpaWRS2pra1kslkqlZvJ77tmzZ5NX/omO\nGuU2NjbUgtG+vr4qBdBPnDgBAO7u7snJyeQSsoS0St3qzZs3AwC1MDdZyVql5jiXyyULUj96\n9IhcwuPxAMDJyenixYvkEplM5ujoqKent379emVB6q+++goAfH19+Xw+uYQsUa1SF3vhwoUq\nxdnJStkqu9HDw4PcjWw2u/vQxMXFqXSJIIiIiAigTPsgzZ8/n3y58vJycglZNJzcuLIeenV1\ntampqUpp9aysLC0tLVtb26NHjyo3ePDgQTIYlCXgCYJYvXo1OXbFxcXKhf7+/sqK6t3HxcrK\nSvmmvL29qW+QrClPVnVXDt/WrVuBMs+AIIiKigoDAwMzMzPq7l28eDHZE+Xu7fFdTJ48WSWK\nyHmj3eutk6OsLCW/c+dOAJgyZYpylHvcdatWrVIJ5vz8fB0dHZVgnjVrFrmLRCIRuYSctNG9\nyDs5XhwO5yn/Z9HfWF8uxFWdme8SerKSZvbKur0/bZrrbNx5k0hbXd5vX6x495v0WrBcFJ1/\nLFj1hi6EevX09crb2trOnTsXFBS0cuVKar1yAIiJiZkxYwb1hrecnBxdXV1HR0eV1YODg6n3\n3SUnJ7u4uFBPT2tqavh8/tSpU6kvffHiRfK8ULmkoKBAoVCMHau8axoIgoiOjg4ODqbWrU5P\nTx8xYgR1IsiTJ0+uXr2qnLxJio2N9fHxodYcLykpqaure+mll6hPO3v27OzZs6k3wV6/ft3S\n0pK8bECSyWQ8Hk85rY905coVT09PagVzgUDw4MEDb29v6tNiYmJmzpxJLc/N5/OHDBni4OCg\nUqJaOTTdu9RjByoqKgQCwcsvd5mGlZ2dbWRkZG9PvbEUEhISxo8fr/LF/Pnz5/39/al3z/Y4\nmrW1tbdu3Zo2bRp1XXJmA3WkgDIu6uuVp6amOjg4UKu6NzY2pqSkqAxfXFycl5cX9T5VoVBY\nVlY2ceJE9e+CvO6inMFKio6ODgwMpNZbz8zMtLa2trXtLHEil8tjY2MDAwOpK3bfddXV1Tk5\nOcqJzKRLly75+vpSg/n+/fsNDQ3u7u7Up509e3bWrFnUe1mV4/X0/7Pob6xvX+w1JK735e7I\nbgSg6bHHenk6DzPTg+bqhwU3rueKmwmAIR6bLiZum2zY/x1GCCGE0KDT54lCTXd+3vjBlsik\nCqlKg+5wv3e+/D5ioavaH6hACCGE0N+HZhOPW2qKMhKTrudViOuaaPpDLWxdvXz9vJ1MdXpf\nFSGEEEJ/Gy/KjUwIIYQQenawtB9CCCGENIX5BEIIIYQ0hfkEQgghhDSF+QRCCCGENIX5BEII\nIYQ0hfkEQgghhDSF+QRCCCGENIX5BEIIIYQ0hfkEQgghhDSF+QRCCCGENIX5BEIIIYQ0hfkE\nQgghhDSF+QRCCCGENIX5BEIIIYQ0hfkEQgghhDSF+QRCCCGENIX5BEIIIYQ0hfkEQgghhDSF\n+QRCCCGENIX5BEIIIYQ0hfkEQgghhDSF+QRCCCGENKU90B1AaIDV1NQUFxcDgKenp5YWZthI\nncrKytLSUgDw8vIa6L6gwU4sFpeXl9PpdA8Pj4Huy18B8wn0ort58+amTZsAICMjA/MJpF5G\nRsZnn31Gp9MzMzMHui9osEtOTt6xY4eBgUFycvJA9+WvgEdPhBBCCGkK8wmEEEIIaQrzCYQQ\nQghpikYQxED3AaGBJBaL79y5AwBTp07F+RNIPaFQmJ+fT6PRpk2bNtB9QYNdRUVFUVGRtra2\nr6/vQPflr4D5BEIIIYQ0hWdjCCGEENIU5hMIIYQQ0hT+/gQaJORVBenxV+LjU3NEzd7rYzZM\norQVH12x9owIAICuZ8KydZk8Z9G8STZMAABorS1MuXg+LiO/VNKsaz7c1ff1xXM82QzqlgVp\np4/9npRTXk9nOXoHvblkxughNADF1W1zdqr+hgDdZ+PvH3vTAKD40PK1ZyWUJsOZ235dOa7j\n0ZO7Fw8ePJd5v6pV33qsb/iKJT7szn8m2YOEo/t/u1ooamZaOk0KWb7Uf6SuslEhyfg18nh8\njqCRbjbSa/Zby4PGGGq68144f3m0AECLKDPqaFRSXkWNlGEyzOWVuUvCJ9koV1U7rH2OFqL2\nzukDhy9nldcTxsPd/Ze8He5uRuvfXfkCUBMtVdEfLTt8FwCArmtsbj365dkLw6fbG5CNvUaL\nourW78dOxWeVVLWZjPJ8bcnSwLEm7efo6qLlWR1bBj5aCIQGg6rYr9Z9fewyP+vnf3G5Ean/\n51ltsgYBP2rrIu4b32Q0kItSd3BDV+6+cPtRY/OTR/yofy/mLtqeXq9cQ/Hgwub5YWsPpBTX\nNMubq4ri9u6JfdTTpluydi/ghm5LbGx/fO/gW9w1UcKe+yGO3RwWsnr/dVGzrOE+L2Jp4LI9\nWc0djfXpOxcFvf1t8oMnssaHqXveC17wZXJNR6Ms56e3g5Z8fulenaxZknX4w9B5H58Xtv2J\nHYWIAYkWWW7kisCw9cf5jxqljeKcqE0LApftZKZ8wwAACWdJREFUudUx6GqHtc/Roig7sTok\nfOOJ3CqprKYgasuC4FU/32vtt734oniqaGlraRTnX/rqrcCwTy6L25f1Ei1VSRFvhK7azSuQ\nNMlldSVXj+w6c6+9SX20PJtjyyCIFswn0OAijFqj5hOC9ODk+9w5EaktBEEQRHbMyaKmzrbK\nsx9yA/8d2/7xQVRd2hgS+mlcba+v25y6Yx43bHemrGOBmv/5ttzIpdyl+/IV7Y8b4j+bE/jx\nhWry0cOT73PDdqRJ2xvlN75dwH33WBn5qC52c3Dwlj86+tN279By7qIf+AoC9cVfGC1F+5dy\nF+3N6Xx8eDl3/p7b5AN1w9r3aJGlfz2P+89TDzteUhy9hjs3IoXSf/RnPEW0PL60mcvdeKE9\nZ1AbLY2pO8KC1pzp+WRAbbQ8m2PLYIgWnD+Bnj90Oh2UtyVxuGGOep1tTCYTCEVbG/mo8mpi\nLvHS1MlDe9tkU3pCptRkynQPRm/PBICHt/lVum7uYzr+eww5HAeikJ8jAwCoz+GXgYs7h9ne\nqOPGGUsT8G9XAQC05t7OU4xwd+/oD82BwzGs5/PLnuJVUd/0U7RoaWlBl2vHBLkMoJdh7Xu0\nFN3mS1nu7jYdr2jB4VjL+fyCPu0G9DS0tOgA0BEv6qJFei0xvcnZb5pVj18oqIsWtZ7vaMF8\nAj1PiNYmSf75gxcq7OYE/aOHyT9tFbG8fLBxdTEmH5eWloC1RStv99plr899ffE7H0Ycvy5q\n6b5eXUr8LTnbb7orvcvi8jPrwkLmhi99b932XzIE0o7FYpEYWCyLzuOFqQWLTohEko7GISwL\nfWUjw4I1FMQiMQBAlUikAAsLVudLWLBYIBaJ8a7tZ6Bfo8V+RohbS9KvJ26LmuTNktzffr7y\n2Dn0NVcA6GVY+xwtUpGoHixYFpStWrCgSSR63B87B6lQSGvuJUT+nmsREDLZuId2lWh5UFqi\nGGqlnbV3w/Lw0HkLl6/+9EDyA+UhQl20kPr72DIoogXnY6LnxY1dcz9PagEAusWEtza97tz9\nSkJT/tHtx++bz/xPkB25QNHUJIeKmH0JgWs/O+hm0lhw/vudX2x5vG3PO+N0qStKkuJzFXYL\n/UdTTim0zN1DV0+b7j5iaKs49/KhHyPW5Ly9azt3GA0ImUwOOgwdygYYTAbIZDIAAJlMBjoM\naucYjC6NdB2GVtdGQiZrAXiaKyPoafV7tNBsZv7rgwdf7N664gQAANPO/4MtXDsyYNQNq06f\no0Umk0HXNRkMJkCdTAaAM3j7U9GhpevOVgMAzXjcgg1vePawd7tFS3NTE9THH/ht6vubfprA\nbilP2Lfjm01V2j9umGQE6qPl2RxbBkW04PUJ9Lz4x9ozMefOnDj49RqPqqPrN5woVnRplpfH\nbPsiun78ik9WuHck8HQmkw6Emf+KpV7DDHUN2O7hK7l2ktgL11u7rCpIiC+CMdOn2lAXjgpc\ntWTyGGtDXX0TO6/wDe/4aBWePn8HAIDGZDKgRU69zCGXyYHJZAKQF0Vb5HJqo7xLo6JF3ta1\nkcZkUg8gqB/0e7TUJG3/cGfOiPe+OXzqt9NHdv1zdOG3H22LIyfpqxvWvkcLk8mErmvK5TLo\nWBP1H6dlh2PORZ86+t9NM9rObfnoJ760a3tP0cJkMoHQ837zPd9Rxnr65mNmf/D6uMfpF1Jq\nAUB9tDybY8ugiBbMJ9DzhKZjwHKc/O47rw4pOXf5TudyhfDK9n8fLLV789ONs4ZRLrqxrawA\n2NbWyji3trGmKSSSaupG78cnlNPd/X3N1bywnv0oa6iXSGQAAJZsS6islHR+R1EjqVTQ2GyL\njsYnlZImZaNcUlkHlmxLAABzNpsOEkll54YllZVgybbEewCfgf6MFkH8mYwGr/kr/Uab6TF0\nTR2mrFo0sfnmmbgygF6Gtc/RostmG4OkknJfoURSCfpsNl6ceAZodD2TkRMWvR8y4tEfF65T\nPrP/T7SwrawAzG2slVcLjGysh4BEIgboJVpU9c+xZVBEC+YT6DnU2tIKQOv4DCaq0nZt3ZNn\nGf7pJyGjumbjIzjuJiASCpVnjkKBkNBmsympA5GXkCTS9fJ/pafvTJWk90uEYGzBYgIADBvP\nMZPmZBd2bPUxn19MG8MZxwQAMHbjjID8bL6svbElh59L2HDGmwMAaI8d70ovy86u63jxYv7t\nx8Yczoi+7AX0dPolWrRo3Y6VnVPs1A5r36PFaTxHtzI7W9DxghI+X8jgcJw13B9IjdaWVqAp\ng0VNtAzluI+AKoFQmXk0CIRPgM22AuglWlT107FlMEQL5hNo8Cs/+/V352/eF9XLWqT1glze\nf/fxnti+Op2c3VSf9cPWr28Yh3zyn/mjdVXXpLkGL+A0XY48fP3hY2mjmH/yx/MPbAKDvTqn\nXSpuxydXGU7yn9B13bbMw58eTcirqGqSN9eW3zgV8VNqm1PobLf2rQaFuzXyIg/dFEvlj0vj\nfvjlpklA2DQzclWbV+dP1kk7si+lolHeJEiLPJQE3uGz2r91NfYLe5WVc2LPH8X1cmll9s/7\nLtU5hwW74f9h/3k20cKeMHk0PfPEj4nF1c1yaU1x8g+/pBGjXvEeBgDqh7Xv0cKYEBoyShj9\nw8m8Gpm8tvDM91Glw4NCJ+qp9hv1WfWV73dGZdwV1Erl0seSopQD3519YD51RvttXmqjxfa1\nhZOZaUf2JpXWS5uqCy98d/qOydQ5U4wBeomWZ3RsGQzRgvXA0CBRtH/puvNdvoeA8e/98vmr\nRgDE49Kr56IupuWVVTbrGJtbO708I2TOVAdDGgA0/LF10d7bKtuyCN55YJlT+4Omkj8OHzyX\nXlTZamBh5zZ90ZvBbmbKfEJ2bdeSL3OmRxx627XrZ3pL1e3YqN+vZN0TNGgNZdmO8QmaP+dl\nm86DypO7Fw8cIH/Dzmqs3/wVi32sOudAyMoTjuyPulooljItHH1Cli8NGEX5fUxxxi+RxxPu\nCJ5omY70mv3W8mBnI8123QtoAKKFqM2/cPwEL7tUXCvTGWoxkuMftjBwnGlH2Kgd1j5HC1Gb\nc3r/kctZZXWE0XD3gCUrwl8yw+Tzz1ITLVJB5oWomOSc+48e0w3N2fYe0+eEBLia0gGeIlrk\nD68eO3AqOf+RlGk2zOWVsKVhEzp+PVNdtDyzY8uARwvmEwghhBDSFKa6CCGEENIU5hMIIYQQ\n0hTmEwghhBDS1P8Af82Fo20hBDcAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 283.4646,
       "width": 354.3307
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width = 150/25.4, repr.plot.height =120/25.4)\n",
    "p<-plotRegion(countdf = countdf,\n",
    "           ArchRProj = projHeme1_subset, base_size=12,\n",
    "           counts_group_colors = c('#205B56',\n",
    "                                   '#bf812d'))\n",
    "p\n",
    "ggsave(\"FigureF-1.pdf\", plot = p, \n",
    "       width = 10, \n",
    "       height = 5, \n",
    "       units = \"in\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "14258118-1d4c-474b-8794-12b3b7b44a1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAC9CAIAAAAyWCF/AAAABmJLR0QA/wD/AP+gvaeTAAAO\nFUlEQVR4nO3daXwTZQLH8SdtkVJKETkKpUiL0MpVhRbKTRXlKvdVQDyXRQVcvEDdVejhrrAC\nK6gsqy772f1wlMMK5RCoAgqCwC5gQSgqLUdbDsspd4/ZF2nTNG3SZDJ5cv2+r5JnZp55Mpn5\n53meSVqdoigCABzPx9kNgIMVFhzJWLF4/pzZ8xYv/WLbsUvFZQtyN89P/GTXpSo3KjmyKjEx\n9UiJEEKI6/uWJM7ekF119de+X5I4d9Np7dsND0TceLKLu2YPimjeffJnu7IvXTt/ZMP8Z6Lv\njxj40SEhhBC5m+cnWYibpCSjuEmyFDdJxA2s4+fsBsBRin58P77fzBvPrD2+cGBjX33ZvF9W\nTR6+8RchHra8rU+7MbNmlbTjwwiaIm481ZXUPyXvbTTl+wWGrBFC1Gw55rMvu52tsF7WV1t2\nHj59I6hV7yFDOjS0JmGuZn25ZuOhC+K+yEeGxxiV3z6wdPamgPGvx5xL27A3p6DR468+3TlQ\nCHEjZ+f6LXtzLusate07fFD7+8r2kZX2burNgW/089uesftYXmH9dn1GDGgTZPcrh8vi88tD\n3d2WvuV606EjY00/UHShoSGGJ2eWje48cm7GsbP5+z+d1LnNqKV5+vIKg6mKzqSObd9+3IKd\npwvOfL9wVMfBn5WPsm4fWJqUOH10p96vfnGk4E6JIoQQtw8uGNAqIj5lw/ELF7KWvxLbuv+H\nx4pK189Kezfpj8927/rUp9/+fP7MjvljH35o0sZrGh4EuBoFHunk3BghenyQa36NPdOaCp82\nM/bf1j+9u31KM9Fm5lFFURSlcMVIIYauKFQURVHOfthd1H0+Q1EURbmaPr6Bz8NvHyjdqODz\nMQ2ECJ6yU1EURbn8aT8havX5KKe4bB9F+95sqavR5c+ZpetfTEtoWKvvJ3n6Z188UVP4d5v/\nU5H+6dXVCUG6XovOa/Hy4ZLo3XgynU5neYWWw8bG1NQ/rBETEyWys83MCJfZk76+oMtzkzuU\nblR/xJRxTSqu0fOJJ8MMZ9X+lSt+8R8+4/X2pevfN3zKuHpbl6UVGFbvOGpsq9LBXlBMTCsl\nOzun2pcFd8XcjYdqEh7uL3ZnZwsRYmGtOnXqGB77+fmJwsJCi7Vezc39zSckpHF5SUhIxfpr\nNmxoNPuSl5cnAs9//ZfEH8pKrhzT6UOtgboWwJ3Ru/FQ9zw6pF/tvPTP9xaZLFByc/NV11o3\nNLROSX7+ufKS/HxLtQUHBwtfX1+jknu7TZw1Oa6B6hbAndG78VT3jk15Z2HsOxNf7puxYEDZ\nzam7J1ZNHrax/+F/j1JZa9fBg+ovWbLo4KSUDjWFEBfTPl5xVviaXb3LuPEtFm9vOPLNt9v7\nlxbdPrXzsK6Fyt3DvRE3Hsuv/YxNXxY9+dTIyK3dhvSLbhZwI2d3+rqDNeLem6q+0qAh7y0Y\n0f3pRzofGfdY89v/S8+4HNxEFJhd3S82OW1ewpCekRkD4juFBd7Kzdq753jTGWt7dlLfBLgv\n38TERGe3AQ6iCwjrNWHqpPjWQcrdu0pAs+iBv09eOG9ab8PkblDLrnHR99cybOAT3C6udxv9\nSEfXqG1cXLtG+rnme5p2iOsRUU8IIe6NGv3cqKg6d24pQRED3/ggOT7UP7RDXLdWdYUQQvg1\niYrr1bp+eRv8GnceN/nJHg1Kbt4sCmgeM+DFlHmv9A42LG7QuldcVBOjLwaFdozr1rKuI44G\nnE+n8BNNAFIwVQxAEuIGgCTEDQBJiBsAkhA3ACQhbgBIQtwAkIS4ASAJcQNAEuIGgCTEDQBJ\niBsAkhA3ACQhbgBIQtwAkIS4ASAJcQNAEuIGgCTEDQBJiBsAkhA3ACQhbgBIQtwAkIS4ASAJ\ncQNAEuKmGuETEsInJBg/AKAOcWMVggawH3EDQBLiBoAkxA0ASYgb2zBhDKhG3ACQhLhRgw4O\noAJxA0AS4gaAJMQNAEmIG5WYvgFsRdwAkIS4sYQuDKAh4gaAJMSNevR9AJsQNwAkIW4ASELc\nmMVYCdAWcQNAEuIGgCTETdUYSQGaI24ASELcAJCEuAEgCXEDQBLiBoAkxA0ASYibKlh/F5z7\n5YD1iBsAkhA3piR3WPg/efAexI0zETTwKsSNvdRFhkmnhtyBN/BzdgNci+rs0D/IWbpS0+YA\nHoXejauggwOPp1MUxdltsEvlqzRn6crwCQnWdDQccYVb2cExt2v6R/Bg7t27qfKirVwo8+4P\nd5oAc9w7biywfNk7OhS0qpzkgifx2LjRc+LlanLjSXXA0V2Cx3CtO1NVTsRYKLepTv2EjpXT\nOg5iU3BYubJzXxFgPc3iRn9tVD7vja8ZDa8Kez7wXbmzYO6euuVMIXHgFtTEjcnlanyiW1ik\njrbRIP8HCtK2tebIk0pwLjU3wq2/EvSDF5MSO+v0ZoajZ+fhInTgFI6du7F+9sGhzYActg6c\nzQ3AXZnm/XdXpvnX5WVPFbvjGeY6ZOay+75T7ttyDdl5EEzONK1mYJ1/Z4qujXzVfqfZ5Hae\nigrNbVjlmtXWbyVnRYw9d05V7Mhy5XYez2o3t2cG0PlxAzfi5R0H67PSXHegyvXtOZ6Wxzua\n785Ojp0qBuCpVMSW2/9EE4C78PAfMQBwHcQNAEmIGwCSEDcAJCFuAEhC3ACQhLgBIAlxA0AS\n4gaAJMQNAEmIGwCSEDcAJOEPUKD4Utb2Ldt/OHW5OKBJZNe+fTs1rVW+8Pq+JXO3NXrqzUEt\nKm9YkpmavNZvzMxRbeQ1Fu6M3o13u33sn09FNY+e+Ml3p65c//Xw2pn9H2j5eOL2i4YVru9b\nkjR7Q3aVG5dkpiYlrzkqq61we/RuvFlB2nOPTdzabtF/173Y2l8IIcScdzdP6RkfP7j2/p3T\n2/o6uXnwNMSN91IO/m3GivOx739TljVCCJ/g/u/PGbt6RMrMtOc/Hx1UeaNbOdvWpO87cyco\nrMeQYZWW3s3bt2HTd8d/LaoXETdsWKfGpafXpV2LF+5rMXFS2NG0LQdPXwofNmtMO4e9LLgu\nBlPe69j69SdE7BNPtKxYHDhswtDav23e8E1RpS0ub5ka3Xpwypcnfj17aOlzsV3/eshoYfGJ\n/4xr+0Dv6SszzxVkb5rZJ6JH0p6b+kWXdi1OmjUtPnpg8pafrxTy99y8Fr0b75WTkyP8h4Y3\nNi3XhYeHiZs5OReECDEuv7v73Rc+vvHsV0f+3qeOECLl2z9E9c4UXUqX/vLBhN+t9Hnp2wPz\ne9QRQqSMnxb1yOQFCQfeelAnhBDiekFkctaqhPqOfU1wafRuvJtOp6tcaOYPyh5en34yYsLk\nPnX0TwN7TX2mvWHhyTUrvi9+5OW3e5Qurd1z6jNtDy1L/bFsBf+BT44ha7wcvRvvFR4eLm5l\nZ58TokmFcuXkyVMiIDqskcn6ubm5IiTEqMNj/CQvL08E+OxdmHi+rKTo0C2Rk50thH6apn7D\nhlUkG7wKvRvv1XrQoBZi7/LlJyoW31i3bN2NwP6D40w/ikJDQ0V+fn55gfGT4OBg4etrfC/L\n7+Hxs6YPDCvRvN1wXwq814XVCY1FgwH/yLpdVlJyfuuUSF//2NmHi/QFZz/sLuo+n6EoinJn\n1yvNxf0vfnVNv+S3b14KF8I3YbX+adacTjUipuy8Wl773bO79/xUpCiK8vN7D4mmr+2R8ZLg\nyhhMebOGo5Z8tXjiqFc7Ppg6cmhsaI2LRzPSNp+Pemv9yhntKn/p5p7uby96YePwIdHZ4we1\n0R3ftDE3sHn50sjXUv91JH5wZOdHB3dvVa/43M8HdmeKict3dKlUD7wW/2cKxRePbtu844fT\nV4oDGkd26ds3NjSgfKHpjxhunvh6dfq+vMKgsB5DRwTuml3xRwxFv2Zu3fTN4fxbgaGRHeP6\ndW2m/0JP6fduXu0bKvWFwdUQNwAkYaoYgCTEDQBJiBsAkhA3NtiRHOXsJgBujLixDYkDqEbc\nAJCEuAEgCXEDQBLiBoAkxI21mCQG7ETcAJCEuAEgCXFjM0ZVgDrEjVWIGMB+xE31yBpAE8RN\nNcgaQCvEjRpkEKACcQNAEuIGgCTEDQBJiBsAkhA3ACQhbgBIQtwAkIS4ASAJcQNAEuIGgCTE\njSX8WAHQEHEDQBLiBoAkxI1KjLMAWxE3ACQhbgBIQtwAkIS4ASAJcQNAEuIGgCTEDQBJiBv1\n+OoNYBPiBoAkxA0ASYgbAJIQN2YxNQNoi7gBIAlxA0AS4sYuDLgA6xE3ACQhbgBIQtzYi/EU\nYCXiBoAkxA0ASYibqjFEAjRH3GiAbAKsQdxUgfgAHIG4MUXWAA5C3ACQhLgBIAlxA0AS4qYC\nJm4AxyFuytmTNeQUUC3iBoAkxA0ASfyc3QBXocloyFBJ3MxM+2sDPAy9G80wfQNY5lq9G0f0\nDvR1mlRoHA30RAA5XCVutLorZCE7zO2CXgkgh0sMpjS84PVV7UiOMjzQfBfWN8NZm5tU5cQ8\nde7e4Wrk9W6qHNQIM5eWNaMqNzqPdyRH2fpCzB0uW/erYW2q9+4NuEtgDcfGTeXLzNbz3uSU\ntWZDZ/VoLLREmE8cy43U9iS2nHpacfphV03d8XHf1yufTlEUx9Vu4Z2Im5mp+n2yZ1snUhE3\nVlZiUz2OThzLLbHQfq26cpZ3ZHkrc5vb1De3fu8uy/B2aNsvdmDcuGMiSGD8LqreXM9wKthz\nqC2fSVXWbNikyivcmsaYu27tOa0r79emvrCtqj3sTkkcfUzYeePV8ptuDzVxQ45AMuNLyPou\nhkkNVZa7/slc+QPGnhpUV0XcAHAaFQHk2LkbADBwie/dAPAGxA0ASf4PtIza7wpDRSEAAAAA\nSUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 94.4882,
       "width": 188.9764
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width = 80/25.4, repr.plot.height =40/25.4)\n",
    "\n",
    "p <- ggplot(data=countdf, aes(x=Locus, y=Counts)) +\n",
    "  geom_bar(stat=\"identity\", aes(fill=Groups)) +\n",
    "  theme_void() +\n",
    "  facet_wrap(~Groups, ncol=1) +\n",
    "  theme(legend.position=\"none\") +\n",
    "  scale_fill_manual(values = c(\"Older\" = '#bf812d', \n",
    "                               \"Children\" = '#205B56'))\n",
    "p\n",
    "ggsave(\"FigureF-2.pdf\", plot = p, \n",
    "       width = 80/25.4, \n",
    "       height = 40/25.4, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6e6aa03-6d4e-4920-82d7-bf9841f76ba1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R_BRI_Figures",
   "language": "R",
   "name": "r_bri_figures"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
